// @sphereon/kmp-mdoc-core/@sphereon/kmp-common.js
(function (factory) {
  if (typeof define === 'function' && define.amd)
    define(['exports', '../kotlin-kotlin-stdlib.js', '../kotlinx-serialization-kotlinx-serialization-json.js', '../Kotlin-DateTime-library-kotlinx-datetime.js', '../kotlinx-serialization-kotlinx-serialization-core.js', '../encoding-library-core.js', '../encoding-library-base64.js', '../Kermit-kermit-core.js', '../Kermit-kermit.js'], factory);
  else if (typeof exports === 'object')
    factory(module.exports, require('../kotlin-kotlin-stdlib.js'), require('../kotlinx-serialization-kotlinx-serialization-json.js'), require('../Kotlin-DateTime-library-kotlinx-datetime.js'), require('../kotlinx-serialization-kotlinx-serialization-core.js'), require('../encoding-library-core.js'), require('../encoding-library-base64.js'), require('../Kermit-kermit-core.js'), require('../Kermit-kermit.js'));
  else {
    if (typeof globalThis['kotlin-kotlin-stdlib'] === 'undefined') {
      throw new Error("Error loading module '@sphereon/kmp-common'. Its dependency 'kotlin-kotlin-stdlib' was not found. Please, check whether 'kotlin-kotlin-stdlib' is loaded prior to '@sphereon/kmp-common'.");
    }
    if (typeof globalThis['kotlinx-serialization-kotlinx-serialization-json'] === 'undefined') {
      throw new Error("Error loading module '@sphereon/kmp-common'. Its dependency 'kotlinx-serialization-kotlinx-serialization-json' was not found. Please, check whether 'kotlinx-serialization-kotlinx-serialization-json' is loaded prior to '@sphereon/kmp-common'.");
    }
    if (typeof globalThis['Kotlin-DateTime-library-kotlinx-datetime'] === 'undefined') {
      throw new Error("Error loading module '@sphereon/kmp-common'. Its dependency 'Kotlin-DateTime-library-kotlinx-datetime' was not found. Please, check whether 'Kotlin-DateTime-library-kotlinx-datetime' is loaded prior to '@sphereon/kmp-common'.");
    }
    if (typeof globalThis['kotlinx-serialization-kotlinx-serialization-core'] === 'undefined') {
      throw new Error("Error loading module '@sphereon/kmp-common'. Its dependency 'kotlinx-serialization-kotlinx-serialization-core' was not found. Please, check whether 'kotlinx-serialization-kotlinx-serialization-core' is loaded prior to '@sphereon/kmp-common'.");
    }
    if (typeof globalThis['encoding-library-core'] === 'undefined') {
      throw new Error("Error loading module '@sphereon/kmp-common'. Its dependency 'encoding-library-core' was not found. Please, check whether 'encoding-library-core' is loaded prior to '@sphereon/kmp-common'.");
    }
    if (typeof globalThis['encoding-library-base64'] === 'undefined') {
      throw new Error("Error loading module '@sphereon/kmp-common'. Its dependency 'encoding-library-base64' was not found. Please, check whether 'encoding-library-base64' is loaded prior to '@sphereon/kmp-common'.");
    }
    if (typeof globalThis['Kermit-kermit-core'] === 'undefined') {
      throw new Error("Error loading module '@sphereon/kmp-common'. Its dependency 'Kermit-kermit-core' was not found. Please, check whether 'Kermit-kermit-core' is loaded prior to '@sphereon/kmp-common'.");
    }
    if (typeof globalThis['Kermit-kermit'] === 'undefined') {
      throw new Error("Error loading module '@sphereon/kmp-common'. Its dependency 'Kermit-kermit' was not found. Please, check whether 'Kermit-kermit' is loaded prior to '@sphereon/kmp-common'.");
    }
    globalThis['@sphereon/kmp-common'] = factory(typeof globalThis['@sphereon/kmp-common'] === 'undefined' ? {} : globalThis['@sphereon/kmp-common'], globalThis['kotlin-kotlin-stdlib'], globalThis['kotlinx-serialization-kotlinx-serialization-json'], globalThis['Kotlin-DateTime-library-kotlinx-datetime'], globalThis['kotlinx-serialization-kotlinx-serialization-core'], globalThis['encoding-library-core'], globalThis['encoding-library-base64'], globalThis['Kermit-kermit-core'], globalThis['Kermit-kermit']);
  }
}(function (_, kotlin_kotlin, kotlin_org_jetbrains_kotlinx_kotlinx_serialization_json, kotlin_org_jetbrains_kotlinx_kotlinx_datetime, kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core, kotlin_io_matthewnelson_encoding_core, kotlin_io_matthewnelson_encoding_base64, kotlin_co_touchlab_kermit_core, kotlin_co_touchlab_kermit) {
  'use strict';
  //region block: imports
  var toList = kotlin_kotlin.$_$.y7;
  var toSet = kotlin_kotlin.$_$.f8;
  var LinkedHashMap_init_$Create$ = kotlin_kotlin.$_$.t;
  var toMutableMap = kotlin_kotlin.$_$.d8;
  var Pair = kotlin_kotlin.$_$.de;
  var mapOf = kotlin_kotlin.$_$.g7;
  var JsonObject = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_json.$_$.h;
  var protoOf = kotlin_kotlin.$_$.xa;
  var VOID = kotlin_kotlin.$_$.f;
  var Companion_getInstance = kotlin_org_jetbrains_kotlinx_kotlinx_datetime.$_$.f;
  var Companion_getInstance_0 = kotlin_org_jetbrains_kotlinx_kotlinx_datetime.$_$.d;
  var Companion_getInstance_1 = kotlin_org_jetbrains_kotlinx_kotlinx_datetime.$_$.g;
  var toLocalDateTime = kotlin_org_jetbrains_kotlinx_kotlinx_datetime.$_$.k;
  var _Char___init__impl__6a9atx = kotlin_kotlin.$_$.c2;
  var endsWith = kotlin_kotlin.$_$.yb;
  var initMetadataForObject = kotlin_kotlin.$_$.ba;
  var initMetadataForCompanion = kotlin_kotlin.$_$.x9;
  var LocalDateTime_init_$Create$ = kotlin_org_jetbrains_kotlinx_kotlinx_datetime.$_$.a;
  var THROW_CCE = kotlin_kotlin.$_$.ge;
  var toInstant = kotlin_org_jetbrains_kotlinx_kotlinx_datetime.$_$.j;
  var Comparable = kotlin_kotlin.$_$.qd;
  var initMetadataForClass = kotlin_kotlin.$_$.w9;
  var STRING_getInstance = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.j;
  var PrimitiveSerialDescriptor = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.h1;
  var KSerializer = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.r2;
  var defineProp = kotlin_kotlin.$_$.n9;
  var System_instance = kotlin_org_jetbrains_kotlinx_kotlinx_datetime.$_$.c;
  var lazy = kotlin_kotlin.$_$.ef;
  var toLong = kotlin_kotlin.$_$.za;
  var Formats_getInstance = kotlin_org_jetbrains_kotlinx_kotlinx_datetime.$_$.e;
  var format = kotlin_org_jetbrains_kotlinx_kotlinx_datetime.$_$.i;
  var THROW_IAE = kotlin_kotlin.$_$.he;
  var Unit_instance = kotlin_kotlin.$_$.q4;
  var Enum = kotlin_kotlin.$_$.ud;
  var Companion_getInstance_2 = kotlin_kotlin.$_$.g4;
  var hexToByteArray = kotlin_kotlin.$_$.ac;
  var toHexString = kotlin_kotlin.$_$.yc;
  var Companion_instance = kotlin_io_matthewnelson_encoding_core.$_$.c;
  var Companion_instance_0 = kotlin_io_matthewnelson_encoding_core.$_$.d;
  var encodeToByteArray = kotlin_kotlin.$_$.xb;
  var noWhenBranchMatchedException = kotlin_kotlin.$_$.ff;
  var decodeToString = kotlin_kotlin.$_$.wb;
  var isByteArray = kotlin_kotlin.$_$.ea;
  var Base64 = kotlin_io_matthewnelson_encoding_base64.$_$.c;
  var Severity_Info_getInstance = kotlin_co_touchlab_kermit_core.$_$.f;
  var Companion_getInstance_3 = kotlin_co_touchlab_kermit.$_$.a;
  var Severity_Debug_getInstance = kotlin_co_touchlab_kermit_core.$_$.d;
  var Severity_Warn_getInstance = kotlin_co_touchlab_kermit_core.$_$.g;
  var Severity_Error_getInstance = kotlin_co_touchlab_kermit_core.$_$.e;
  var toDouble = kotlin_kotlin.$_$.xc;
  var objectCreate = kotlin_kotlin.$_$.wa;
  var Number_0 = kotlin_kotlin.$_$.ce;
  var numberToLong = kotlin_kotlin.$_$.va;
  var _ULong___init__impl__c78o9k = kotlin_kotlin.$_$.c3;
  var _UInt___init__impl__l7qpdl = kotlin_kotlin.$_$.t2;
  var _ULong___get_data__impl__fggpzb = kotlin_kotlin.$_$.d3;
  var LONG_getInstance = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.i;
  var _UInt___get_data__impl__f0vqqw = kotlin_kotlin.$_$.u2;
  var Long = kotlin_kotlin.$_$.ae;
  var Companion_getInstance_4 = kotlin_kotlin.$_$.i4;
  //endregion
  //region block: pre-declaration
  initMetadataForObject(Static, 'Static');
  initMetadataForCompanion(Companion);
  initMetadataForClass(LocalDateTimeKMP, 'LocalDateTimeKMP', VOID, VOID, [Comparable], VOID, VOID, {0: LocalDateTimeIso8601SerializerKMP_getInstance});
  initMetadataForObject(LocalDateTimeIso8601SerializerKMP, 'LocalDateTimeIso8601SerializerKMP', VOID, VOID, [KSerializer]);
  initMetadataForObject(Static_0, 'Static');
  initMetadataForClass(DateTimeUtils, 'DateTimeUtils', DateTimeUtils);
  initMetadataForObject(DefaultDateTimeUtils, 'DefaultDateTimeUtils');
  initMetadataForClass(Encoding, 'Encoding', VOID, Enum);
  initMetadataForObject(Base64Serializer, 'Base64Serializer', VOID, VOID, [KSerializer]);
  initMetadataForObject(Base64UrlSerializer, 'Base64UrlSerializer', VOID, VOID, [KSerializer]);
  initMetadataForObject(Static_1, 'Static');
  initMetadataForClass(Logger, 'Logger', Logger);
  initMetadataForCompanion(Companion_0);
  initMetadataForClass(LongKMP, 'LongKMP', VOID, Number_0, [Number_0, Comparable], VOID, VOID, {0: LongKMPSerializer_getInstance});
  initMetadataForObject(LongKMPSerializer, 'LongKMPSerializer', VOID, VOID, [KSerializer]);
  initMetadataForObject(Uuid, 'Uuid');
  //endregion
  function kmpListOf(elements) {
    return toList(elements);
  }
  function kmpSetOf(elements) {
    return toSet(elements);
  }
  function kmpMapOf() {
    // Inline function 'kotlin.collections.mutableMapOf' call
    return LinkedHashMap_init_$Create$();
  }
  function mergeJsonElement(_this__u8e3s4, key, value) {
    var result = toMutableMap(_this__u8e3s4);
    result.t2(mapOf(new Pair(key, value)));
    return new JsonObject(result);
  }
  function Static() {
  }
  protoOf(Static).now = function () {
    return this.nowWithTimezone();
  };
  protoOf(Static).b2z = function (utils, timeZoneId) {
    return utils.dateTime(timeZoneId);
  };
  protoOf(Static).nowWithTimezone = function (utils, timeZoneId, $super) {
    utils = utils === VOID ? Static_getInstance_0().DEFAULT : utils;
    timeZoneId = timeZoneId === VOID ? null : timeZoneId;
    return this.b2z(utils, timeZoneId);
  };
  protoOf(Static).fromString = function (value) {
    var tmp;
    // Inline function 'kotlin.text.lowercase' call
    // Inline function 'kotlin.js.asDynamic' call
    var tmp$ret$1 = value.toLowerCase();
    if (endsWith(tmp$ret$1, _Char___init__impl__6a9atx(122))) {
      var instant = Companion_getInstance_0().i24(value);
      tmp = toLocalDateTime(instant, Companion_getInstance_1().f25(Static_getInstance_0().DEFAULT.timeZoneId));
    } else {
      tmp = Companion_getInstance().s24(value);
    }
    var datetime = tmp;
    return new LocalDateTimeKMP(datetime.t1p(), datetime.u1p(), datetime.v1p(), datetime.v1q(), datetime.z1q(), datetime.d1r(), datetime.b1r());
  };
  var Static_instance;
  function Static_getInstance() {
    return Static_instance;
  }
  function Companion() {
  }
  protoOf(Companion).k19 = function () {
    return LocalDateTimeIso8601SerializerKMP_getInstance();
  };
  var Companion_instance_1;
  function Companion_getInstance_5() {
    return Companion_instance_1;
  }
  function LocalDateTimeKMP(year, monthNumber, dayOfMonth, hour, minute, second, nanosecond) {
    second = second === VOID ? 0 : second;
    nanosecond = nanosecond === VOID ? 0 : nanosecond;
    this.year = year;
    this.monthNumber = monthNumber;
    this.dayOfMonth = dayOfMonth;
    this.hour = hour;
    this.minute = minute;
    this.second = second;
    this.nanosecond = nanosecond;
    this.delegate = LocalDateTime_init_$Create$(this.year, this.monthNumber, this.dayOfMonth, this.hour, this.minute, this.second, this.nanosecond);
  }
  protoOf(LocalDateTimeKMP).t1p = function () {
    return this.year;
  };
  protoOf(LocalDateTimeKMP).u1p = function () {
    return this.monthNumber;
  };
  protoOf(LocalDateTimeKMP).v1p = function () {
    return this.dayOfMonth;
  };
  protoOf(LocalDateTimeKMP).v1q = function () {
    return this.hour;
  };
  protoOf(LocalDateTimeKMP).z1q = function () {
    return this.minute;
  };
  protoOf(LocalDateTimeKMP).d1r = function () {
    return this.second;
  };
  protoOf(LocalDateTimeKMP).b1r = function () {
    return this.nanosecond;
  };
  protoOf(LocalDateTimeKMP).d2z = function () {
    return this.delegate;
  };
  protoOf(LocalDateTimeKMP).e2z = function (other) {
    return Static_instance.fromString(this.toString()).e2z(Static_instance.fromString(other.toString()));
  };
  protoOf(LocalDateTimeKMP).d = function (other) {
    return this.e2z(other instanceof LocalDateTimeKMP ? other : THROW_CCE());
  };
  protoOf(LocalDateTimeKMP).toString = function () {
    return this.delegate.toString();
  };
  protoOf(LocalDateTimeKMP).f2z = function (timeZoneId) {
    return fromLong(toInstant(this.delegate, Static_getInstance_0().DEFAULT.g2z(timeZoneId)).k1t());
  };
  protoOf(LocalDateTimeKMP).toEpochSeconds = function (timeZoneId, $super) {
    timeZoneId = timeZoneId === VOID ? Companion_getInstance_1().e25_1.i25() : timeZoneId;
    return this.f2z(timeZoneId);
  };
  function LocalDateTimeIso8601SerializerKMP() {
    LocalDateTimeIso8601SerializerKMP_instance = this;
    this.h2z_1 = PrimitiveSerialDescriptor('com.sphereon.kmp.datetime.LocalDateTime', STRING_getInstance());
  }
  protoOf(LocalDateTimeIso8601SerializerKMP).ln = function () {
    return this.h2z_1;
  };
  protoOf(LocalDateTimeIso8601SerializerKMP).zn = function (decoder) {
    var parsed = Companion_getInstance().s24(decoder.rq());
    return new LocalDateTimeKMP(parsed.t1p(), parsed.u1p(), parsed.v1p(), parsed.v1q(), parsed.z1q(), parsed.d1r(), parsed.b1r());
  };
  protoOf(LocalDateTimeIso8601SerializerKMP).i2z = function (encoder, value) {
    encoder.zr(value.toString());
  };
  protoOf(LocalDateTimeIso8601SerializerKMP).yn = function (encoder, value) {
    return this.i2z(encoder, value instanceof LocalDateTimeKMP ? value : THROW_CCE());
  };
  var LocalDateTimeIso8601SerializerKMP_instance;
  function LocalDateTimeIso8601SerializerKMP_getInstance() {
    if (LocalDateTimeIso8601SerializerKMP_instance == null)
      new LocalDateTimeIso8601SerializerKMP();
    return LocalDateTimeIso8601SerializerKMP_instance;
  }
  function Static_0() {
    Static_instance_0 = this;
    this.DEFAULT = new DateTimeUtils();
  }
  protoOf(Static_0).j2z = function () {
    return this.DEFAULT;
  };
  var Static_instance_0;
  function Static_getInstance_0() {
    if (Static_instance_0 == null)
      new Static_0();
    return Static_instance_0;
  }
  function now($this) {
    return $this.clock.k1p();
  }
  function DateTimeUtils$defaultTimeZone$lambda(this$0) {
    return function () {
      return Companion_getInstance_1().f25(this$0.timeZoneId);
    };
  }
  function DateTimeUtils(clock, timeZoneId) {
    clock = clock === VOID ? System_instance : clock;
    timeZoneId = timeZoneId === VOID ? Companion_getInstance_1().v25().i25() : timeZoneId;
    this.clock = clock;
    this.timeZoneId = timeZoneId;
    var tmp = this;
    tmp.c2z_1 = lazy(DateTimeUtils$defaultTimeZone$lambda(this));
  }
  protoOf(DateTimeUtils).k2z = function (_set____db54di) {
    this.clock = _set____db54di;
  };
  protoOf(DateTimeUtils).l2z = function () {
    return this.clock;
  };
  protoOf(DateTimeUtils).m2z = function (_set____db54di) {
    this.timeZoneId = _set____db54di;
  };
  protoOf(DateTimeUtils).n2z = function () {
    return this.timeZoneId;
  };
  protoOf(DateTimeUtils).epochSeconds = function () {
    return now(this).k1t().i1();
  };
  protoOf(DateTimeUtils).o2z = function (epochSeconds) {
    return this.dateTime(Companion_getInstance_1().e25_1.i25(), epochSeconds);
  };
  protoOf(DateTimeUtils).dateTimeUTC = function (epochSeconds, $super) {
    epochSeconds = epochSeconds === VOID ? this.epochSeconds() : epochSeconds;
    return this.o2z(epochSeconds);
  };
  protoOf(DateTimeUtils).p2z = function (epochSeconds) {
    return this.dateTime(this.timeZoneId, epochSeconds);
  };
  protoOf(DateTimeUtils).dateTimeLocal = function (epochSeconds, $super) {
    epochSeconds = epochSeconds === VOID ? this.epochSeconds() : epochSeconds;
    return this.p2z(epochSeconds);
  };
  protoOf(DateTimeUtils).q2z = function (timeZoneId, epochSeconds) {
    var tmp = Companion_getInstance_0();
    var tmp1_elvis_lhs = epochSeconds == null ? null : toLong(epochSeconds);
    return toKMP(toLocalDateTime(tmp.n25(tmp1_elvis_lhs == null ? toLong(this.epochSeconds()) : tmp1_elvis_lhs), this.g2z(timeZoneId)));
  };
  protoOf(DateTimeUtils).dateTime = function (timeZoneId, epochSeconds, $super) {
    timeZoneId = timeZoneId === VOID ? null : timeZoneId;
    epochSeconds = epochSeconds === VOID ? this.epochSeconds() : epochSeconds;
    return this.q2z(timeZoneId, epochSeconds);
  };
  protoOf(DateTimeUtils).r2z = function (epochSeconds) {
    return this.dateISO(this.timeZoneId, epochSeconds);
  };
  protoOf(DateTimeUtils).dateLocalISO = function (epochSeconds, $super) {
    epochSeconds = epochSeconds === VOID ? this.epochSeconds() : epochSeconds;
    return this.r2z(epochSeconds);
  };
  protoOf(DateTimeUtils).s2z = function (timeZoneId, epochSeconds) {
    return format(toKotlin(this.dateTime(timeZoneId == null ? this.timeZoneId : timeZoneId, epochSeconds)).j1x(), Formats_getInstance().q1p());
  };
  protoOf(DateTimeUtils).dateISO = function (timeZoneId, epochSeconds, $super) {
    epochSeconds = epochSeconds === VOID ? this.epochSeconds() : epochSeconds;
    return this.s2z(timeZoneId, epochSeconds);
  };
  protoOf(DateTimeUtils).g2z = function (timeZoneId) {
    var tmp;
    if (timeZoneId == null) {
      tmp = null;
    } else {
      // Inline function 'kotlin.let' call
      // Inline function 'kotlin.contracts.contract' call
      // Inline function 'com.sphereon.kmp.DateTimeUtils.timeZone.<anonymous>' call
      tmp = Companion_getInstance_1().f25(timeZoneId);
    }
    var tmp1_elvis_lhs = tmp;
    return tmp1_elvis_lhs == null ? this.c2z_1.d2() : tmp1_elvis_lhs;
  };
  function DefaultDateTimeUtils() {
    DefaultDateTimeUtils_instance = this;
    this.INSTANCE = new DateTimeUtils();
  }
  protoOf(DefaultDateTimeUtils).t2z = function () {
    return this.INSTANCE;
  };
  var DefaultDateTimeUtils_instance;
  function DefaultDateTimeUtils_getInstance() {
    if (DefaultDateTimeUtils_instance == null)
      new DefaultDateTimeUtils();
    return DefaultDateTimeUtils_instance;
  }
  function toKMP(_this__u8e3s4) {
    return Static_instance.fromString(_this__u8e3s4.toString());
  }
  function toKotlin(_this__u8e3s4) {
    return Companion_getInstance().s24(_this__u8e3s4.toString());
  }
  function getDateTime(dateTimeUtils) {
    dateTimeUtils = dateTimeUtils === VOID ? null : dateTimeUtils;
    return dateTimeUtils == null ? DefaultDateTimeUtils_getInstance().INSTANCE : dateTimeUtils;
  }
  function toLocalDateTimeKMP(_this__u8e3s4, dateTimeUtils) {
    dateTimeUtils = dateTimeUtils === VOID ? null : dateTimeUtils;
    return getDateTime(dateTimeUtils).dateTimeLocal(_this__u8e3s4.i1());
  }
  function get_base64UrlConf() {
    _init_properties_Encoding_kt__jty2h9();
    return base64UrlConf;
  }
  var base64UrlConf;
  function get_base64Conf() {
    _init_properties_Encoding_kt__jty2h9();
    return base64Conf;
  }
  var base64Conf;
  var Encoding_BASE64_instance;
  var Encoding_BASE64URL_instance;
  var Encoding_HEX_instance;
  var Encoding_UTF8_instance;
  function values() {
    return [Encoding_BASE64_getInstance(), Encoding_BASE64URL_getInstance(), Encoding_HEX_getInstance(), Encoding_UTF8_getInstance()];
  }
  function valueOf(value) {
    switch (value) {
      case 'BASE64':
        return Encoding_BASE64_getInstance();
      case 'BASE64URL':
        return Encoding_BASE64URL_getInstance();
      case 'HEX':
        return Encoding_HEX_getInstance();
      case 'UTF8':
        return Encoding_UTF8_getInstance();
      default:
        Encoding_initEntries();
        THROW_IAE('No enum constant value.');
        break;
    }
  }
  var Encoding_entriesInitialized;
  function Encoding_initEntries() {
    if (Encoding_entriesInitialized)
      return Unit_instance;
    Encoding_entriesInitialized = true;
    Encoding_BASE64_instance = new Encoding('BASE64', 0);
    Encoding_BASE64URL_instance = new Encoding('BASE64URL', 1);
    Encoding_HEX_instance = new Encoding('HEX', 2);
    Encoding_UTF8_instance = new Encoding('UTF8', 3);
  }
  function Encoding(name, ordinal) {
    Enum.call(this, name, ordinal);
  }
  function decodeFromHex(_this__u8e3s4) {
    _init_properties_Encoding_kt__jty2h9();
    return hexToByteArray(_this__u8e3s4, Companion_getInstance_2().vi_1);
  }
  function encodeToHex(_this__u8e3s4) {
    _init_properties_Encoding_kt__jty2h9();
    return toHexString(_this__u8e3s4, Companion_getInstance_2().vi_1);
  }
  function decodeFromBase64Url(_this__u8e3s4) {
    _init_properties_Encoding_kt__jty2h9();
    return Companion_instance.u2v(_this__u8e3s4, get_base64UrlConf());
  }
  function encodeToBase64Url(_this__u8e3s4) {
    _init_properties_Encoding_kt__jty2h9();
    return Companion_instance_0.f2w(_this__u8e3s4, get_base64UrlConf());
  }
  function encodeToBase64(_this__u8e3s4, urlSafe) {
    urlSafe = urlSafe === VOID ? false : urlSafe;
    _init_properties_Encoding_kt__jty2h9();
    return Companion_instance_0.f2w(_this__u8e3s4, urlSafe ? get_base64UrlConf() : get_base64Conf());
  }
  function decodeFromBase64(_this__u8e3s4, urlSafe) {
    urlSafe = urlSafe === VOID ? false : urlSafe;
    _init_properties_Encoding_kt__jty2h9();
    return Companion_instance.u2v(_this__u8e3s4, urlSafe ? get_base64UrlConf() : get_base64Conf());
  }
  function decodeFrom(_this__u8e3s4, encoding) {
    _init_properties_Encoding_kt__jty2h9();
    var tmp;
    switch (encoding.v2_1) {
      case 1:
        tmp = decodeFromBase64Url(_this__u8e3s4);
        break;
      case 0:
        tmp = decodeFromBase64(_this__u8e3s4, false);
        break;
      case 2:
        tmp = decodeFromHex(_this__u8e3s4);
        break;
      case 3:
        tmp = encodeToByteArray(_this__u8e3s4);
        break;
      default:
        noWhenBranchMatchedException();
        break;
    }
    return tmp;
  }
  function encodeTo(_this__u8e3s4, encoding) {
    _init_properties_Encoding_kt__jty2h9();
    var tmp;
    switch (encoding.v2_1) {
      case 1:
        tmp = encodeToBase64Url(_this__u8e3s4);
        break;
      case 0:
        tmp = encodeToBase64(_this__u8e3s4, false);
        break;
      case 2:
        tmp = encodeToHex(_this__u8e3s4);
        break;
      case 3:
        tmp = decodeToString(_this__u8e3s4);
        break;
      default:
        noWhenBranchMatchedException();
        break;
    }
    return tmp;
  }
  function Base64Serializer() {
    Base64Serializer_instance = this;
    this.w2z_1 = PrimitiveSerialDescriptor('Base64', STRING_getInstance());
  }
  protoOf(Base64Serializer).ln = function () {
    return this.w2z_1;
  };
  protoOf(Base64Serializer).x2z = function (encoder, value) {
    return encoder.zr(encodeToBase64(value, false));
  };
  protoOf(Base64Serializer).yn = function (encoder, value) {
    return this.x2z(encoder, (!(value == null) ? isByteArray(value) : false) ? value : THROW_CCE());
  };
  protoOf(Base64Serializer).zn = function (decoder) {
    var str = decoder.rq();
    return decodeFromBase64(str, false);
  };
  var Base64Serializer_instance;
  function Base64Serializer_getInstance() {
    if (Base64Serializer_instance == null)
      new Base64Serializer();
    return Base64Serializer_instance;
  }
  function Base64UrlSerializer() {
    Base64UrlSerializer_instance = this;
    this.y2z_1 = PrimitiveSerialDescriptor('Base64Url', STRING_getInstance());
  }
  protoOf(Base64UrlSerializer).ln = function () {
    return this.y2z_1;
  };
  protoOf(Base64UrlSerializer).x2z = function (encoder, value) {
    return encoder.zr(encodeToBase64(value, true));
  };
  protoOf(Base64UrlSerializer).yn = function (encoder, value) {
    return this.x2z(encoder, (!(value == null) ? isByteArray(value) : false) ? value : THROW_CCE());
  };
  protoOf(Base64UrlSerializer).zn = function (decoder) {
    var str = decoder.rq();
    return decodeFromBase64(str, true);
  };
  var Base64UrlSerializer_instance;
  function Base64UrlSerializer_getInstance() {
    if (Base64UrlSerializer_instance == null)
      new Base64UrlSerializer();
    return Base64UrlSerializer_instance;
  }
  function base64UrlConf$lambda($this$Base64) {
    _init_properties_Encoding_kt__jty2h9();
    $this$Base64.l2x_1 = true;
    $this$Base64.m2x_1 = 0;
    $this$Base64.n2x_1 = true;
    $this$Base64.o2x_1 = false;
    return Unit_instance;
  }
  function base64Conf$lambda($this$Base64) {
    _init_properties_Encoding_kt__jty2h9();
    $this$Base64.l2x_1 = true;
    $this$Base64.m2x_1 = 0;
    $this$Base64.n2x_1 = false;
    $this$Base64.o2x_1 = false;
    return Unit_instance;
  }
  function Encoding_BASE64_getInstance() {
    Encoding_initEntries();
    return Encoding_BASE64_instance;
  }
  function Encoding_BASE64URL_getInstance() {
    Encoding_initEntries();
    return Encoding_BASE64URL_instance;
  }
  function Encoding_HEX_getInstance() {
    Encoding_initEntries();
    return Encoding_HEX_instance;
  }
  function Encoding_UTF8_getInstance() {
    Encoding_initEntries();
    return Encoding_UTF8_instance;
  }
  var properties_initialized_Encoding_kt_vg5col;
  function _init_properties_Encoding_kt__jty2h9() {
    if (!properties_initialized_Encoding_kt_vg5col) {
      properties_initialized_Encoding_kt_vg5col = true;
      base64UrlConf = Base64(base64UrlConf$lambda);
      base64Conf = Base64(base64Conf$lambda);
    }
  }
  function get_DefaultLogger() {
    _init_properties_Logger_kt__mpqsw();
    return DefaultLogger;
  }
  var DefaultLogger;
  function Static_1() {
  }
  protoOf(Static_1).z2z = function (tag, severity) {
    // Inline function 'kotlin.also' call
    var this_0 = new Logger(tag);
    // Inline function 'kotlin.contracts.contract' call
    // Inline function 'com.sphereon.kmp.Static.tag.<anonymous>' call
    this_0.b2v(severity);
    return this_0;
  };
  protoOf(Static_1).b30 = function (tag, severity, $super) {
    tag = tag === VOID ? '' : tag;
    severity = severity === VOID ? Severity_Info_getInstance() : severity;
    return $super === VOID ? this.z2z(tag, severity) : $super.z2z.call(this, tag, severity);
  };
  var Static_instance_1;
  function Static_getInstance_1() {
    return Static_instance_1;
  }
  function Logger(tag) {
    tag = tag === VOID ? '' : tag;
    this.a30_1 = tag;
  }
  protoOf(Logger).c30 = function (message, tag) {
    // Inline function 'co.touchlab.kermit.Logger.d' call
    // Inline function 'co.touchlab.kermit.BaseLogger.log' call
    var this_0 = Companion_getInstance_3();
    var severity = Severity_Debug_getInstance();
    if (this_0.f2u().y2u().y2(severity) <= 0) {
      this_0.h2u(severity, tag, null, message);
    }
  };
  protoOf(Logger).d30 = function (message, tag, $super) {
    tag = tag === VOID ? this.a30_1 : tag;
    var tmp;
    if ($super === VOID) {
      this.c30(message, tag);
      tmp = Unit_instance;
    } else {
      tmp = $super.c30.call(this, message, tag);
    }
    return tmp;
  };
  protoOf(Logger).e30 = function (message, tag) {
    // Inline function 'co.touchlab.kermit.Logger.i' call
    // Inline function 'co.touchlab.kermit.BaseLogger.log' call
    var this_0 = Companion_getInstance_3();
    var severity = Severity_Info_getInstance();
    if (this_0.f2u().y2u().y2(severity) <= 0) {
      this_0.h2u(severity, tag, null, message);
    }
  };
  protoOf(Logger).f30 = function (message, tag, $super) {
    tag = tag === VOID ? this.a30_1 : tag;
    var tmp;
    if ($super === VOID) {
      this.e30(message, tag);
      tmp = Unit_instance;
    } else {
      tmp = $super.e30.call(this, message, tag);
    }
    return tmp;
  };
  protoOf(Logger).g30 = function (message, tag) {
    // Inline function 'co.touchlab.kermit.Logger.w' call
    // Inline function 'co.touchlab.kermit.BaseLogger.log' call
    var this_0 = Companion_getInstance_3();
    var severity = Severity_Warn_getInstance();
    if (this_0.f2u().y2u().y2(severity) <= 0) {
      this_0.h2u(severity, tag, null, message);
    }
  };
  protoOf(Logger).h30 = function (message, tag, $super) {
    tag = tag === VOID ? this.a30_1 : tag;
    var tmp;
    if ($super === VOID) {
      this.g30(message, tag);
      tmp = Unit_instance;
    } else {
      tmp = $super.g30.call(this, message, tag);
    }
    return tmp;
  };
  protoOf(Logger).i30 = function (message, throwable, tag) {
    // Inline function 'co.touchlab.kermit.Logger.e' call
    // Inline function 'co.touchlab.kermit.BaseLogger.log' call
    var this_0 = Companion_getInstance_3();
    var severity = Severity_Error_getInstance();
    if (this_0.f2u().y2u().y2(severity) <= 0) {
      this_0.h2u(severity, tag, throwable, message);
    }
  };
  protoOf(Logger).j30 = function (message, throwable, tag, $super) {
    throwable = throwable === VOID ? null : throwable;
    tag = tag === VOID ? this.a30_1 : tag;
    var tmp;
    if ($super === VOID) {
      this.i30(message, throwable, tag);
      tmp = Unit_instance;
    } else {
      tmp = $super.i30.call(this, message, throwable, tag);
    }
    return tmp;
  };
  protoOf(Logger).b2v = function (severity) {
    return Companion_getInstance_3().b2v(severity);
  };
  var properties_initialized_Logger_kt_6n8a9a;
  function _init_properties_Logger_kt__mpqsw() {
    if (!properties_initialized_Logger_kt_6n8a9a) {
      properties_initialized_Logger_kt_6n8a9a = true;
      DefaultLogger = new Logger('');
    }
  }
  function LongKMP_init_$Init$(value, $this) {
    LongKMP.call($this, toDouble(value));
    return $this;
  }
  function fromString(value) {
    return LongKMP_init_$Init$(value, objectCreate(protoOf(LongKMP)));
  }
  function LongKMP_init_$Init$_0(value, $this) {
    LongKMP.call($this, value.v3());
    return $this;
  }
  function fromLong(value) {
    return LongKMP_init_$Init$_0(value, objectCreate(protoOf(LongKMP)));
  }
  function LongKMP_init_$Init$_1(value, $this) {
    LongKMP.call($this, value);
    return $this;
  }
  function fromNumber(value) {
    return LongKMP_init_$Init$_1(value, objectCreate(protoOf(LongKMP)));
  }
  function Companion_0() {
  }
  protoOf(Companion_0).k19 = function () {
    return LongKMPSerializer_getInstance();
  };
  var Companion_instance_2;
  function Companion_getInstance_6() {
    return Companion_instance_2;
  }
  function LongKMP(value) {
    Number_0.call(this);
    this.k30_1 = numberToLong(value);
  }
  protoOf(LongKMP).s3 = function () {
    return this.k30_1.s3();
  };
  protoOf(LongKMP).v3 = function () {
    return this.k30_1.v3();
  };
  protoOf(LongKMP).u3 = function () {
    return this.k30_1.u3();
  };
  protoOf(LongKMP).i1 = function () {
    return this.k30_1.i1();
  };
  protoOf(LongKMP).m19 = function () {
    return this.k30_1;
  };
  protoOf(LongKMP).toULong = function () {
    // Inline function 'kotlin.toULong' call
    var this_0 = this.k30_1;
    return _ULong___init__impl__c78o9k(this_0);
  };
  protoOf(LongKMP).toUInt = function () {
    // Inline function 'kotlin.toUInt' call
    var this_0 = this.k30_1;
    return _UInt___init__impl__l7qpdl(this_0.i1());
  };
  protoOf(LongKMP).t3 = function () {
    return this.k30_1.t3();
  };
  protoOf(LongKMP).toString = function () {
    return this.k30_1.toString();
  };
  protoOf(LongKMP).l30 = function (other) {
    return this.k30_1.g1(other.k30_1);
  };
  protoOf(LongKMP).d = function (other) {
    return this.l30(other instanceof LongKMP ? other : THROW_CCE());
  };
  protoOf(LongKMP).equals = function (other) {
    if (this === other)
      return true;
    if (!(other instanceof LongKMP))
      return false;
    if (!this.k30_1.equals(other.k30_1))
      return false;
    return true;
  };
  protoOf(LongKMP).hashCode = function () {
    return this.k30_1.hashCode();
  };
  function numberToKmpLong(_this__u8e3s4) {
    return toKmpLong(_this__u8e3s4);
  }
  function ulongToKmpLong(_this__u8e3s4) {
    // Inline function 'kotlin.ULong.toLong' call
    var tmp$ret$0 = _ULong___get_data__impl__fggpzb(_this__u8e3s4);
    return fromLong(tmp$ret$0);
  }
  function byteToKmpLong(_this__u8e3s4) {
    return fromLong(toLong(_this__u8e3s4));
  }
  function stringToKmpLong(_this__u8e3s4) {
    return fromString(_this__u8e3s4);
  }
  function LongKMPSerializer() {
    LongKMPSerializer_instance = this;
    this.m30_1 = PrimitiveSerialDescriptor('LongKMP', LONG_getInstance());
  }
  protoOf(LongKMPSerializer).ln = function () {
    return this.m30_1;
  };
  protoOf(LongKMPSerializer).n30 = function (encoder, value) {
    encoder.vr(value.m19());
  };
  protoOf(LongKMPSerializer).yn = function (encoder, value) {
    return this.n30(encoder, value instanceof LongKMP ? value : THROW_CCE());
  };
  protoOf(LongKMPSerializer).zn = function (decoder) {
    return fromLong(decoder.nq());
  };
  var LongKMPSerializer_instance;
  function LongKMPSerializer_getInstance() {
    if (LongKMPSerializer_instance == null)
      new LongKMPSerializer();
    return LongKMPSerializer_instance;
  }
  function toKmpLong(_this__u8e3s4) {
    return fromLong(numberToLong(_this__u8e3s4));
  }
  function uintToKmpLong(_this__u8e3s4) {
    // Inline function 'kotlin.UInt.toLong' call
    // Inline function 'kotlin.uintToLong' call
    var value = _UInt___get_data__impl__f0vqqw(_this__u8e3s4);
    var tmp$ret$1 = toLong(value).p3(new Long(-1, 0));
    return fromLong(tmp$ret$1);
  }
  function Uuid() {
  }
  protoOf(Uuid).v4 = function () {
    return Companion_getInstance_4().jl();
  };
  protoOf(Uuid).v4String = function () {
    return this.v4().toString();
  };
  var Uuid_instance;
  function Uuid_getInstance() {
    return Uuid_instance;
  }
  //region block: post-declaration
  defineProp(protoOf(Encoding), 'name', protoOf(Encoding).w2);
  defineProp(protoOf(Encoding), 'ordinal', protoOf(Encoding).x2);
  //endregion
  //region block: init
  Static_instance = new Static();
  Companion_instance_1 = new Companion();
  Static_instance_1 = new Static_1();
  Companion_instance_2 = new Companion_0();
  Uuid_instance = new Uuid();
  //endregion
  //region block: exports
  function $jsExportAll$(_) {
    var $com = _.com || (_.com = {});
    var $com$sphereon = $com.sphereon || ($com.sphereon = {});
    var $com$sphereon$kmp = $com$sphereon.kmp || ($com$sphereon.kmp = {});
    $com$sphereon$kmp.kmpListOf = kmpListOf;
    $com$sphereon$kmp.kmpSetOf = kmpSetOf;
    $com$sphereon$kmp.kmpMapOf = kmpMapOf;
    var $com = _.com || (_.com = {});
    var $com$sphereon = $com.sphereon || ($com.sphereon = {});
    var $com$sphereon$kmp = $com$sphereon.kmp || ($com$sphereon.kmp = {});
    $com$sphereon$kmp.LocalDateTimeKMP = LocalDateTimeKMP;
    defineProp($com$sphereon$kmp.LocalDateTimeKMP, 'Static', Static_getInstance);
    defineProp($com$sphereon$kmp.LocalDateTimeKMP, 'Companion', Companion_getInstance_5);
    var $com = _.com || (_.com = {});
    var $com$sphereon = $com.sphereon || ($com.sphereon = {});
    var $com$sphereon$kmp = $com$sphereon.kmp || ($com$sphereon.kmp = {});
    $com$sphereon$kmp.DateTimeUtils = DateTimeUtils;
    defineProp($com$sphereon$kmp.DateTimeUtils, 'Static', Static_getInstance_0);
    defineProp($com$sphereon$kmp, 'DefaultDateTimeUtils', DefaultDateTimeUtils_getInstance);
    var $com = _.com || (_.com = {});
    var $com$sphereon = $com.sphereon || ($com.sphereon = {});
    var $com$sphereon$kmp = $com$sphereon.kmp || ($com$sphereon.kmp = {});
    $com$sphereon$kmp.Encoding = Encoding;
    $com$sphereon$kmp.Encoding.values = values;
    $com$sphereon$kmp.Encoding.valueOf = valueOf;
    defineProp($com$sphereon$kmp.Encoding, 'BASE64', Encoding_BASE64_getInstance);
    defineProp($com$sphereon$kmp.Encoding, 'BASE64URL', Encoding_BASE64URL_getInstance);
    defineProp($com$sphereon$kmp.Encoding, 'HEX', Encoding_HEX_getInstance);
    defineProp($com$sphereon$kmp.Encoding, 'UTF8', Encoding_UTF8_getInstance);
    $com$sphereon$kmp.decodeFromHex = decodeFromHex;
    $com$sphereon$kmp.encodeToHex = encodeToHex;
    $com$sphereon$kmp.decodeFromBase64Url = decodeFromBase64Url;
    $com$sphereon$kmp.encodeToBase64Url = encodeToBase64Url;
    $com$sphereon$kmp.encodeToBase64 = encodeToBase64;
    $com$sphereon$kmp.decodeFromBase64 = decodeFromBase64;
    $com$sphereon$kmp.decodeFrom = decodeFrom;
    $com$sphereon$kmp.encodeTo = encodeTo;
    var $com = _.com || (_.com = {});
    var $com$sphereon = $com.sphereon || ($com.sphereon = {});
    var $com$sphereon$kmp = $com$sphereon.kmp || ($com$sphereon.kmp = {});
    $com$sphereon$kmp.LongKMP = LongKMP;
    $com$sphereon$kmp.LongKMP.fromString = fromString;
    $com$sphereon$kmp.LongKMP.fromLong = fromLong;
    $com$sphereon$kmp.LongKMP.fromNumber = fromNumber;
    defineProp($com$sphereon$kmp.LongKMP, 'Companion', Companion_getInstance_6);
    $com$sphereon$kmp.numberToKmpLong = numberToKmpLong;
    $com$sphereon$kmp.ulongToKmpLong = ulongToKmpLong;
    $com$sphereon$kmp.byteToKmpLong = byteToKmpLong;
    $com$sphereon$kmp.stringToKmpLong = stringToKmpLong;
    var $com = _.com || (_.com = {});
    var $com$sphereon = $com.sphereon || ($com.sphereon = {});
    var $com$sphereon$kmp = $com$sphereon.kmp || ($com$sphereon.kmp = {});
    defineProp($com$sphereon$kmp, 'Uuid', Uuid_getInstance);
  }
  $jsExportAll$(_);
  _.$jsExportAll$ = $jsExportAll$;
  _.$_$ = _.$_$ || {};
  _.$_$.a = get_DefaultLogger;
  _.$_$.b = LongKMP;
  _.$_$.c = decodeFromBase64Url;
  _.$_$.d = decodeFromHex;
  _.$_$.e = decodeFrom;
  _.$_$.f = encodeToBase64Url;
  _.$_$.g = encodeTo;
  _.$_$.h = getDateTime;
  _.$_$.i = mergeJsonElement;
  _.$_$.j = numberToKmpLong;
  _.$_$.k = stringToKmpLong;
  _.$_$.l = toKmpLong;
  _.$_$.m = toKotlin;
  _.$_$.n = toLocalDateTimeKMP;
  _.$_$.o = uintToKmpLong;
  _.$_$.p = ulongToKmpLong;
  _.$_$.q = Encoding_BASE64URL_getInstance;
  _.$_$.r = Encoding_BASE64_getInstance;
  _.$_$.s = Encoding_HEX_getInstance;
  _.$_$.t = Encoding_UTF8_getInstance;
  _.$_$.u = fromNumber;
  _.$_$.v = fromLong;
  _.$_$.w = Base64Serializer_getInstance;
  _.$_$.x = Base64UrlSerializer_getInstance;
  _.$_$.y = Static_getInstance_0;
  _.$_$.z = DefaultDateTimeUtils_getInstance;
  _.$_$.a1 = Static_instance;
  _.$_$.b1 = Static_instance_1;
  _.$_$.c1 = LongKMPSerializer_getInstance;
  _.$_$.d1 = Uuid_instance;
  //endregion
  return _;
}));
