// @sphereon/kmp-mdoc-core/core-library-digest.js
(function (factory) {
  if (typeof define === 'function' && define.amd)
    define(['exports', './kotlin-kotlin-stdlib.js'], factory);
  else if (typeof exports === 'object')
    factory(module.exports, require('./kotlin-kotlin-stdlib.js'));
  else {
    if (typeof globalThis['kotlin-kotlin-stdlib'] === 'undefined') {
      throw new Error("Error loading module 'core-library-digest'. Its dependency 'kotlin-kotlin-stdlib' was not found. Please, check whether 'kotlin-kotlin-stdlib' is loaded prior to 'core-library-digest'.");
    }
    globalThis['core-library-digest'] = factory(typeof globalThis['core-library-digest'] === 'undefined' ? {} : globalThis['core-library-digest'], globalThis['kotlin-kotlin-stdlib']);
  }
}(function (_, kotlin_kotlin) {
  'use strict';
  //region block: imports
  var initMetadataForClass = kotlin_kotlin.$_$.w9;
  var VOID = kotlin_kotlin.$_$.f;
  var THROW_CCE = kotlin_kotlin.$_$.ge;
  var protoOf = kotlin_kotlin.$_$.xa;
  var isBlank = kotlin_kotlin.$_$.dc;
  var toString = kotlin_kotlin.$_$.bb;
  var IllegalArgumentException_init_$Create$ = kotlin_kotlin.$_$.n1;
  var Long = kotlin_kotlin.$_$.ae;
  var initMetadataForCompanion = kotlin_kotlin.$_$.x9;
  var Unit_instance = kotlin_kotlin.$_$.q4;
  var toLong = kotlin_kotlin.$_$.za;
  var fill = kotlin_kotlin.$_$.m6;
  var equals = kotlin_kotlin.$_$.p9;
  var hashCode = kotlin_kotlin.$_$.v9;
  //endregion
  //region block: pre-declaration
  initMetadataForClass(DigestState, 'DigestState');
  initMetadataForClass(RealState, 'RealState', VOID, DigestState);
  initMetadataForCompanion(Companion);
  initMetadataForClass(DigestDelegate, 'DigestDelegate');
  initMetadataForClass(Digest, 'Digest');
  //endregion
  function RealState() {
  }
  function Companion() {
  }
  protoOf(Companion).x1l = function (state, compress, digest, resetDigest) {
    var tmp = (state instanceof RealState ? state : THROW_CCE()).y1l_1;
    // Inline function 'kotlin.collections.copyOf' call
    // Inline function 'kotlin.js.asDynamic' call
    var tmp$ret$1 = state.e1m_1.slice();
    return new DigestDelegate(tmp, state.z1l_1, state.a1m_1, tmp$ret$1, state.f1m_1, state.g1m_1, compress, digest, resetDigest);
  };
  protoOf(Companion).h1m = function (algorithm, blockSize, digestLength, compress, digest, resetDigest) {
    // Inline function 'kotlin.require' call
    // Inline function 'kotlin.text.isNotBlank' call
    // Inline function 'kotlin.contracts.contract' call
    if (!!isBlank(algorithm)) {
      // Inline function 'org.kotlincrypto.core.digest.internal.Companion.instance.<anonymous>' call
      var message = 'algorithm cannot be blank';
      throw IllegalArgumentException_init_$Create$(toString(message));
    }
    // Inline function 'kotlin.require' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(blockSize > 0)) {
      // Inline function 'org.kotlincrypto.core.digest.internal.Companion.instance.<anonymous>' call
      var message_0 = 'blockSize must be greater than 0';
      throw IllegalArgumentException_init_$Create$(toString(message_0));
    }
    // Inline function 'kotlin.require' call
    // Inline function 'kotlin.contracts.contract' call
    if (!((blockSize % 8 | 0) === 0)) {
      // Inline function 'org.kotlincrypto.core.digest.internal.Companion.instance.<anonymous>' call
      var message_1 = 'blockSize must be a factor of 8';
      throw IllegalArgumentException_init_$Create$(toString(message_1));
    }
    // Inline function 'kotlin.require' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(digestLength >= 0)) {
      // Inline function 'org.kotlincrypto.core.digest.internal.Companion.instance.<anonymous>' call
      var message_2 = 'digestLength cannot be negative';
      throw IllegalArgumentException_init_$Create$(toString(message_2));
    }
    return new DigestDelegate(algorithm, blockSize, digestLength, new Int8Array(blockSize), 0, new Long(0, 0), compress, digest, resetDigest);
  };
  var Companion_instance;
  function Companion_getInstance() {
    return Companion_instance;
  }
  function DigestDelegate(algorithm, blockSize, digestLength, buffer, bufferOffs, compressCount, compress, digest, resetDigest) {
    this.i1m_1 = algorithm;
    this.j1m_1 = blockSize;
    this.k1m_1 = digestLength;
    this.l1m_1 = buffer;
    this.m1m_1 = bufferOffs;
    this.n1m_1 = compressCount;
    this.o1m_1 = compress;
    this.p1m_1 = digest;
    this.q1m_1 = resetDigest;
  }
  protoOf(DigestDelegate).r1m = function (input) {
    this.l1m_1[this.m1m_1] = input;
    this.m1m_1 = this.m1m_1 + 1 | 0;
    if (!(this.m1m_1 === this.j1m_1))
      return Unit_instance;
    this.o1m_1(this.l1m_1, 0);
    this.n1m_1 = this.n1m_1.i3();
    this.m1m_1 = 0;
  };
  protoOf(DigestDelegate).s1m = function (input, offset, len) {
    var i = offset;
    var remaining = len;
    while (!(this.m1m_1 === 0) && remaining > 0) {
      var tmp0 = i;
      i = tmp0 + 1 | 0;
      this.r1m(input[tmp0]);
      remaining = remaining - 1 | 0;
    }
    while (remaining >= this.j1m_1) {
      this.o1m_1(input, i);
      i = i + this.j1m_1 | 0;
      remaining = remaining - this.j1m_1 | 0;
      this.n1m_1 = this.n1m_1.i3();
    }
    $l$loop: while (true) {
      var tmp2 = remaining;
      remaining = tmp2 - 1 | 0;
      if (!(tmp2 > 0)) {
        break $l$loop;
      }
      var tmp3 = i;
      i = tmp3 + 1 | 0;
      this.r1m(input[tmp3]);
    }
  };
  protoOf(DigestDelegate).t1m = function () {
    // Inline function 'kotlin.Long.times' call
    // Inline function 'kotlin.Long.plus' call
    // Inline function 'kotlin.Long.times' call
    var this_0 = this.n1m_1;
    var other = this.j1m_1;
    var this_1 = this_0.f3(toLong(other));
    var other_0 = this.m1m_1;
    var bitLength = this_1.d3(toLong(other_0)).f3(toLong(8));
    var final = this.p1m_1(bitLength, this.m1m_1, this.l1m_1);
    this.u1m();
    return final;
  };
  protoOf(DigestDelegate).u1m = function () {
    fill(this.l1m_1, 0);
    this.m1m_1 = 0;
    this.n1m_1 = new Long(0, 0);
    this.q1m_1();
  };
  function DigestState() {
  }
  function Digest_init_$Init$(algorithm, blockSize, digestLength, $this) {
    Digest.call($this, algorithm, blockSize, digestLength, null);
    return $this;
  }
  function Digest$compress$ref($boundThis) {
    var l = function (p0, p1) {
      $boundThis.w1m(p0, p1);
      return Unit_instance;
    };
    l.callableName = 'compress';
    return l;
  }
  function Digest$digest$ref($boundThis) {
    var l = function (p0, p1, p2) {
      return $boundThis.x1m(p0, p1, p2);
    };
    l.callableName = 'digest';
    return l;
  }
  function Digest$resetDigest$ref($boundThis) {
    var l = function () {
      $boundThis.y1m();
      return Unit_instance;
    };
    l.callableName = 'resetDigest';
    return l;
  }
  function Digest$compress$ref_0($boundThis) {
    var l = function (p0, p1) {
      $boundThis.w1m(p0, p1);
      return Unit_instance;
    };
    l.callableName = 'compress';
    return l;
  }
  function Digest$digest$ref_0($boundThis) {
    var l = function (p0, p1, p2) {
      return $boundThis.x1m(p0, p1, p2);
    };
    l.callableName = 'digest';
    return l;
  }
  function Digest$resetDigest$ref_0($boundThis) {
    var l = function () {
      $boundThis.y1m();
      return Unit_instance;
    };
    l.callableName = 'resetDigest';
    return l;
  }
  function Digest(algorithm, blockSize, digestLength, state) {
    var tmp = this;
    var tmp_0;
    if (!(state == null)) {
      var tmp_1 = Companion_instance;
      var tmp_2 = Digest$compress$ref(this);
      var tmp_3 = Digest$digest$ref(this);
      tmp_0 = tmp_1.x1l(state, tmp_2, tmp_3, Digest$resetDigest$ref(this));
    } else {
      var tmp_4 = Companion_instance;
      var tmp_5 = Digest$compress$ref_0(this);
      var tmp_6 = Digest$digest$ref_0(this);
      tmp_0 = tmp_4.h1m(algorithm, blockSize, digestLength, tmp_5, tmp_6, Digest$resetDigest$ref_0(this));
    }
    tmp.v1m_1 = tmp_0;
  }
  protoOf(Digest).z1m = function () {
    return this.v1m_1.i1m_1;
  };
  protoOf(Digest).a1n = function (input) {
    this.b1n(input, 0, input.length);
  };
  protoOf(Digest).c1n = function () {
    return this.v1m_1.t1m();
  };
  protoOf(Digest).equals = function (other) {
    var tmp;
    if (other instanceof Digest) {
      tmp = equals(other.v1m_1, this.v1m_1);
    } else {
      tmp = false;
    }
    return tmp;
  };
  protoOf(Digest).hashCode = function () {
    return hashCode(this.v1m_1);
  };
  protoOf(Digest).toString = function () {
    // Inline function 'org.kotlincrypto.core.digest.internal.commonToString' call
    return 'Digest[' + this.z1m() + ']@' + this.hashCode();
  };
  protoOf(Digest).b1n = function (input, offset, len) {
    this.v1m_1.s1m(input, offset, len);
  };
  //region block: init
  Companion_instance = new Companion();
  //endregion
  //region block: exports
  _.$_$ = _.$_$ || {};
  _.$_$.a = Digest_init_$Init$;
  _.$_$.b = Digest;
  //endregion
  return _;
}));
