// @sphereon/kmp-mdoc-core/cryptography-kotlin-cryptography-serialization-pem.js
(function (factory) {
  if (typeof define === 'function' && define.amd)
    define(['exports', './kotlin-kotlin-stdlib.js'], factory);
  else if (typeof exports === 'object')
    factory(module.exports, require('./kotlin-kotlin-stdlib.js'));
  else {
    if (typeof globalThis['kotlin-kotlin-stdlib'] === 'undefined') {
      throw new Error("Error loading module 'cryptography-kotlin-cryptography-serialization-pem'. Its dependency 'kotlin-kotlin-stdlib' was not found. Please, check whether 'kotlin-kotlin-stdlib' is loaded prior to 'cryptography-kotlin-cryptography-serialization-pem'.");
    }
    globalThis['cryptography-kotlin-cryptography-serialization-pem'] = factory(typeof globalThis['cryptography-kotlin-cryptography-serialization-pem'] === 'undefined' ? {} : globalThis['cryptography-kotlin-cryptography-serialization-pem'], globalThis['kotlin-kotlin-stdlib']);
  }
}(function (_, kotlin_kotlin) {
  'use strict';
  //region block: imports
  var decodeToString = kotlin_kotlin.$_$.wb;
  var protoOf = kotlin_kotlin.$_$.xa;
  var split = kotlin_kotlin.$_$.mc;
  var startsWith = kotlin_kotlin.$_$.nc;
  var Unit_instance = kotlin_kotlin.$_$.q4;
  var toString = kotlin_kotlin.$_$.bb;
  var IllegalStateException_init_$Create$ = kotlin_kotlin.$_$.q1;
  var substringAfter = kotlin_kotlin.$_$.qc;
  var substringBefore = kotlin_kotlin.$_$.rc;
  var THROW_CCE = kotlin_kotlin.$_$.ge;
  var isCharSequence = kotlin_kotlin.$_$.ga;
  var trim = kotlin_kotlin.$_$.ld;
  var isBlank = kotlin_kotlin.$_$.dc;
  var joinToString = kotlin_kotlin.$_$.v6;
  var Default_getInstance = kotlin_kotlin.$_$.w3;
  var initMetadataForObject = kotlin_kotlin.$_$.ba;
  var initMetadataForClass = kotlin_kotlin.$_$.w9;
  var initMetadataForCompanion = kotlin_kotlin.$_$.x9;
  //endregion
  //region block: pre-declaration
  initMetadataForObject(Pem, 'Pem');
  initMetadataForClass(PemContent, 'PemContent');
  initMetadataForCompanion(Companion);
  //endregion
  function Pem() {
    this.f1b_1 = '-----BEGIN ';
    this.g1b_1 = '-----END ';
    this.h1b_1 = '-----';
  }
  protoOf(Pem).i1b = function (bytes) {
    return this.j1b(decodeToString(bytes));
  };
  protoOf(Pem).j1b = function (string) {
    var lines = split(string, ['\n']);
    var tmp$ret$1;
    $l$block: {
      // Inline function 'kotlin.collections.indexOfFirst' call
      var index = 0;
      var tmp0_iterator = lines.q();
      while (tmp0_iterator.r()) {
        var item = tmp0_iterator.s();
        // Inline function 'dev.whyoleg.cryptography.serialization.pem.Pem.decode.<anonymous>' call
        if (startsWith(item, '-----BEGIN ')) {
          tmp$ret$1 = index;
          break $l$block;
        }
        index = index + 1 | 0;
      }
      tmp$ret$1 = -1;
    }
    var beginLine = tmp$ret$1;
    // Inline function 'kotlin.check' call
    // Inline function 'kotlin.contracts.contract' call
    if (!!(beginLine === -1)) {
      // Inline function 'dev.whyoleg.cryptography.serialization.pem.Pem.decode.<anonymous>' call
      var message = 'Invalid PEM format: missing BEGIN label';
      throw IllegalStateException_init_$Create$(toString(message));
    }
    var tmp$ret$4;
    $l$block_0: {
      // Inline function 'kotlin.collections.indexOfFirst' call
      var index_0 = 0;
      var tmp0_iterator_0 = lines.q();
      while (tmp0_iterator_0.r()) {
        var item_0 = tmp0_iterator_0.s();
        // Inline function 'dev.whyoleg.cryptography.serialization.pem.Pem.decode.<anonymous>' call
        if (startsWith(item_0, '-----END ')) {
          tmp$ret$4 = index_0;
          break $l$block_0;
        }
        index_0 = index_0 + 1 | 0;
      }
      tmp$ret$4 = -1;
    }
    var endLine = tmp$ret$4;
    // Inline function 'kotlin.check' call
    // Inline function 'kotlin.contracts.contract' call
    if (!!(endLine === -1)) {
      // Inline function 'dev.whyoleg.cryptography.serialization.pem.Pem.decode.<anonymous>' call
      var message_0 = 'Invalid PEM format: missing END label';
      throw IllegalStateException_init_$Create$(toString(message_0));
    }
    // Inline function 'kotlin.text.trim' call
    var this_0 = substringBefore(substringAfter(lines.u(beginLine), '-----BEGIN '), '-----');
    var beginLabel = toString(trim(isCharSequence(this_0) ? this_0 : THROW_CCE()));
    // Inline function 'kotlin.check' call
    // Inline function 'kotlin.text.isNotBlank' call
    // Inline function 'kotlin.contracts.contract' call
    if (!!isBlank(beginLabel)) {
      // Inline function 'dev.whyoleg.cryptography.serialization.pem.Pem.decode.<anonymous>' call
      var message_1 = 'Invalid PEM format: BEGIN label is empty';
      throw IllegalStateException_init_$Create$(toString(message_1));
    }
    // Inline function 'kotlin.text.trim' call
    var this_1 = substringBefore(substringAfter(lines.u(endLine), '-----END '), '-----');
    var endLabel = toString(trim(isCharSequence(this_1) ? this_1 : THROW_CCE()));
    // Inline function 'kotlin.check' call
    // Inline function 'kotlin.text.isNotBlank' call
    // Inline function 'kotlin.contracts.contract' call
    if (!!isBlank(endLabel)) {
      // Inline function 'dev.whyoleg.cryptography.serialization.pem.Pem.decode.<anonymous>' call
      var message_2 = 'Invalid PEM format: BEGIN label is empty';
      throw IllegalStateException_init_$Create$(toString(message_2));
    }
    // Inline function 'kotlin.check' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(beginLabel === endLabel)) {
      // Inline function 'dev.whyoleg.cryptography.serialization.pem.Pem.decode.<anonymous>' call
      var message_3 = 'Invalid PEM format: BEGIN=`' + beginLabel + '`, END=`' + endLabel + '`';
      throw IllegalStateException_init_$Create$(toString(message_3));
    }
    var contentText = joinToString(lines.b2(beginLine + 1 | 0, endLine), '');
    return new PemContent(_PemLabel___init__impl__xifhtx(beginLabel), Default_getInstance().yh(contentText));
  };
  var Pem_instance;
  function Pem_getInstance() {
    return Pem_instance;
  }
  function PemContent(label, bytes) {
    this.k1b_1 = label;
    this.l1b_1 = bytes;
  }
  function ensurePemLabel(_this__u8e3s4, label) {
    // Inline function 'kotlin.check' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(_this__u8e3s4.k1b_1 === label)) {
      // Inline function 'dev.whyoleg.cryptography.serialization.pem.ensurePemLabel.<anonymous>' call
      var message = 'Wrong PEM label, expected ' + PemLabel__toString_impl_auwpt1(label) + ', actual ' + PemLabel__toString_impl_auwpt1(_this__u8e3s4.k1b_1);
      throw IllegalStateException_init_$Create$(toString(message));
    }
    return _this__u8e3s4;
  }
  function _PemLabel___init__impl__xifhtx(representation) {
    return representation;
  }
  function Companion() {
    Companion_instance = this;
    this.m1b_1 = _PemLabel___init__impl__xifhtx('PUBLIC KEY');
    this.n1b_1 = _PemLabel___init__impl__xifhtx('PRIVATE KEY');
    this.o1b_1 = _PemLabel___init__impl__xifhtx('RSA PUBLIC KEY');
    this.p1b_1 = _PemLabel___init__impl__xifhtx('RSA PRIVATE KEY');
    this.q1b_1 = _PemLabel___init__impl__xifhtx('EC PRIVATE KEY');
  }
  var Companion_instance;
  function Companion_getInstance() {
    if (Companion_instance == null)
      new Companion();
    return Companion_instance;
  }
  function PemLabel__toString_impl_auwpt1($this) {
    return 'PemLabel(representation=' + $this + ')';
  }
  //region block: init
  Pem_instance = new Pem();
  //endregion
  //region block: exports
  _.$_$ = _.$_$ || {};
  _.$_$.a = ensurePemLabel;
  _.$_$.b = Companion_getInstance;
  _.$_$.c = Pem_instance;
  //endregion
  return _;
}));
