// @sphereon/kmp-mdoc-core/cryptography-kotlin-cryptography-provider-webcrypto.js
(function (factory) {
  if (typeof define === 'function' && define.amd)
    define(['exports', './cryptography-kotlin-cryptography-core.js', './kotlin-kotlin-stdlib.js', './cryptography-kotlin-cryptography-serialization-pem.js', './cryptography-kotlin-cryptography-serialization-asn1.js', './cryptography-kotlin-cryptography-serialization-asn1-modules.js', './cryptography-kotlin-cryptography-provider-base.js'], factory);
  else if (typeof exports === 'object')
    factory(module.exports, require('./cryptography-kotlin-cryptography-core.js'), require('./kotlin-kotlin-stdlib.js'), require('./cryptography-kotlin-cryptography-serialization-pem.js'), require('./cryptography-kotlin-cryptography-serialization-asn1.js'), require('./cryptography-kotlin-cryptography-serialization-asn1-modules.js'), require('./cryptography-kotlin-cryptography-provider-base.js'));
  else {
    if (typeof globalThis['cryptography-kotlin-cryptography-core'] === 'undefined') {
      throw new Error("Error loading module 'cryptography-kotlin-cryptography-provider-webcrypto'. Its dependency 'cryptography-kotlin-cryptography-core' was not found. Please, check whether 'cryptography-kotlin-cryptography-core' is loaded prior to 'cryptography-kotlin-cryptography-provider-webcrypto'.");
    }
    if (typeof globalThis['kotlin-kotlin-stdlib'] === 'undefined') {
      throw new Error("Error loading module 'cryptography-kotlin-cryptography-provider-webcrypto'. Its dependency 'kotlin-kotlin-stdlib' was not found. Please, check whether 'kotlin-kotlin-stdlib' is loaded prior to 'cryptography-kotlin-cryptography-provider-webcrypto'.");
    }
    if (typeof globalThis['cryptography-kotlin-cryptography-serialization-pem'] === 'undefined') {
      throw new Error("Error loading module 'cryptography-kotlin-cryptography-provider-webcrypto'. Its dependency 'cryptography-kotlin-cryptography-serialization-pem' was not found. Please, check whether 'cryptography-kotlin-cryptography-serialization-pem' is loaded prior to 'cryptography-kotlin-cryptography-provider-webcrypto'.");
    }
    if (typeof globalThis['cryptography-kotlin-cryptography-serialization-asn1'] === 'undefined') {
      throw new Error("Error loading module 'cryptography-kotlin-cryptography-provider-webcrypto'. Its dependency 'cryptography-kotlin-cryptography-serialization-asn1' was not found. Please, check whether 'cryptography-kotlin-cryptography-serialization-asn1' is loaded prior to 'cryptography-kotlin-cryptography-provider-webcrypto'.");
    }
    if (typeof globalThis['cryptography-kotlin-cryptography-serialization-asn1-modules'] === 'undefined') {
      throw new Error("Error loading module 'cryptography-kotlin-cryptography-provider-webcrypto'. Its dependency 'cryptography-kotlin-cryptography-serialization-asn1-modules' was not found. Please, check whether 'cryptography-kotlin-cryptography-serialization-asn1-modules' is loaded prior to 'cryptography-kotlin-cryptography-provider-webcrypto'.");
    }
    if (typeof globalThis['cryptography-kotlin-cryptography-provider-base'] === 'undefined') {
      throw new Error("Error loading module 'cryptography-kotlin-cryptography-provider-webcrypto'. Its dependency 'cryptography-kotlin-cryptography-provider-base' was not found. Please, check whether 'cryptography-kotlin-cryptography-provider-base' is loaded prior to 'cryptography-kotlin-cryptography-provider-webcrypto'.");
    }
    globalThis['cryptography-kotlin-cryptography-provider-webcrypto'] = factory(typeof globalThis['cryptography-kotlin-cryptography-provider-webcrypto'] === 'undefined' ? {} : globalThis['cryptography-kotlin-cryptography-provider-webcrypto'], globalThis['cryptography-kotlin-cryptography-core'], globalThis['kotlin-kotlin-stdlib'], globalThis['cryptography-kotlin-cryptography-serialization-pem'], globalThis['cryptography-kotlin-cryptography-serialization-asn1'], globalThis['cryptography-kotlin-cryptography-serialization-asn1-modules'], globalThis['cryptography-kotlin-cryptography-provider-base']);
  }
}(function (_, kotlin_dev_whyoleg_cryptography_cryptography_core, kotlin_kotlin, kotlin_dev_whyoleg_cryptography_cryptography_serialization_pem, kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1, kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1_modules, kotlin_dev_whyoleg_cryptography_cryptography_provider_base) {
  'use strict';
  //region block: imports
  var CryptographyProvider = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.p;
  var Companion_getInstance = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.k1;
  var equals = kotlin_kotlin.$_$.p9;
  var Companion_getInstance_0 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.m1;
  var Companion_getInstance_1 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.i1;
  var Companion_getInstance_2 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.j1;
  var Companion_getInstance_3 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.o1;
  var Companion_getInstance_4 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.p1;
  var Companion_getInstance_5 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.n1;
  var Companion_getInstance_6 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.w;
  var Companion_getInstance_7 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.v;
  var Companion_getInstance_8 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.u;
  var Companion_getInstance_9 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.l1;
  var SHA512_getInstance = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.d2;
  var SHA384_getInstance = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.c2;
  var SHA256_getInstance = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.b2;
  var SHA1_getInstance = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.a2;
  var THROW_CCE = kotlin_kotlin.$_$.ge;
  var CryptographyAlgorithm = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.o;
  var isInterface = kotlin_kotlin.$_$.ka;
  var protoOf = kotlin_kotlin.$_$.xa;
  var initMetadataForObject = kotlin_kotlin.$_$.ba;
  var VOID = kotlin_kotlin.$_$.f;
  var lazy = kotlin_kotlin.$_$.ef;
  var initMetadataForClass = kotlin_kotlin.$_$.w9;
  var noWhenBranchMatchedException = kotlin_kotlin.$_$.ff;
  var Format = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.c;
  var _BinarySize___get_inBits__impl__p3mjjl = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.t;
  var IvAuthenticatedEncryptor = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.b;
  var IvAuthenticatedDecryptor = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.a;
  var initMetadataForCompanion = kotlin_kotlin.$_$.x9;
  var _Curve___get_name__impl__klrusc = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.s;
  var DER_getInstance = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.e1;
  var PEM_getInstance = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.g1;
  var RAW_getInstance = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.h1;
  var JWK_getInstance = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.f1;
  var Format_0 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.e;
  var Companion_getInstance_10 = kotlin_dev_whyoleg_cryptography_cryptography_serialization_pem.$_$.b;
  var Default_getInstance = kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1.$_$.i;
  var Companion_instance = kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1_modules.$_$.j;
  var EcParameters = kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1_modules.$_$.b;
  var toString = kotlin_kotlin.$_$.bb;
  var IllegalStateException_init_$Create$ = kotlin_kotlin.$_$.q1;
  var Unit_instance = kotlin_kotlin.$_$.q4;
  var EcKeyAlgorithmIdentifier = kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1_modules.$_$.a;
  var PrivateKeyInfo = kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1_modules.$_$.d;
  var Companion_getInstance_11 = kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1_modules.$_$.k;
  var RAW_getInstance_0 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.d1;
  var Generic_getInstance = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.y;
  var Generic_getInstance_0 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.b1;
  var SEC1_getInstance = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.z;
  var SEC1_getInstance_0 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.c1;
  var JWK_getInstance_0 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.a1;
  var Format_1 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.d;
  var Companion_instance_0 = kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1.$_$.k;
  var get_secp521r1 = kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1_modules.$_$.h;
  var Companion_instance_1 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.x;
  var get_secp384r1 = kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1_modules.$_$.g;
  var get_secp256r1 = kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1_modules.$_$.f;
  var _EcParameters___init__impl__i4ww9z = kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1_modules.$_$.i;
  var EcPrivateKey = kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1_modules.$_$.c;
  var PrivateKey = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.f;
  var PublicKey = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.g;
  var SharedSecretGenerator = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.m;
  var Companion_instance_2 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.e2;
  var CoroutineImpl = kotlin_kotlin.$_$.z8;
  var get_COROUTINE_SUSPENDED = kotlin_kotlin.$_$.k8;
  var initMetadataForCoroutine = kotlin_kotlin.$_$.y9;
  var SecretDerivation = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.l;
  var get_EmptyByteArray = kotlin_dev_whyoleg_cryptography_cryptography_provider_base.$_$.a;
  var Format_2 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.h;
  var Generic_getInstance_1 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.v1;
  var Generic_getInstance_2 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.y1;
  var PKCS1_getInstance = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.w1;
  var PKCS1_getInstance_0 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.z1;
  var JWK_getInstance_1 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.x1;
  var Format_3 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.j;
  var RsaKeyAlgorithmIdentifier_instance = kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1_modules.$_$.l;
  var Generic_getInstance_3 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.q1;
  var Generic_getInstance_4 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.t1;
  var PKCS1_getInstance_1 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.r1;
  var PKCS1_getInstance_2 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.u1;
  var JWK_getInstance_2 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.s1;
  var Format_4 = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.i;
  var Pem_instance = kotlin_dev_whyoleg_cryptography_cryptography_serialization_pem.$_$.c;
  var ensurePemLabel = kotlin_dev_whyoleg_cryptography_cryptography_serialization_pem.$_$.a;
  var Companion_getInstance_12 = kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1_modules.$_$.m;
  var BitArray = kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1.$_$.a;
  var SubjectPublicKeyInfo = kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1_modules.$_$.e;
  var KeyDecoder = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.k;
  var SignatureGenerator = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.n;
  var Registry_getInstance = kotlin_dev_whyoleg_cryptography_cryptography_core.$_$.g2;
  var defineProp = kotlin_kotlin.$_$.n9;
  var decodeToString = kotlin_kotlin.$_$.wb;
  var Companion_instance_3 = kotlin_kotlin.$_$.l4;
  var _Result___init__impl__xyqfz8 = kotlin_kotlin.$_$.g2;
  var createFailure = kotlin_kotlin.$_$.ye;
  var intercepted = kotlin_kotlin.$_$.m8;
  var SafeContinuation_init_$Create$ = kotlin_kotlin.$_$.z;
  var returnIfSuspended = kotlin_kotlin.$_$.h;
  //endregion
  //region block: pre-declaration
  initMetadataForObject(WebCryptoCryptographyProvider, 'WebCryptoCryptographyProvider', VOID, CryptographyProvider);
  initMetadataForClass(WebCryptoEncodableKey, 'WebCryptoEncodableKey', VOID, VOID, VOID, [1]);
  initMetadataForClass(AesKey, 'AesKey', VOID, WebCryptoEncodableKey, VOID, [1]);
  initMetadataForClass(WebCryptoAes, 'WebCryptoAes', VOID, VOID, [CryptographyAlgorithm]);
  initMetadataForClass(WebCryptoKeyProcessor, 'WebCryptoKeyProcessor');
  initMetadataForObject(AesKeyProcessor, 'AesKeyProcessor', VOID, WebCryptoKeyProcessor);
  initMetadataForClass(AesCbcKey, 'AesCbcKey', VOID, AesKey, VOID, [1]);
  initMetadataForObject(WebCryptoAesCbc, 'WebCryptoAesCbc', VOID, WebCryptoAes, [WebCryptoAes, CryptographyAlgorithm]);
  initMetadataForClass(AesCtrKey, 'AesCtrKey', VOID, AesKey, VOID, [1]);
  initMetadataForObject(WebCryptoAesCtr, 'WebCryptoAesCtr', VOID, WebCryptoAes, [WebCryptoAes, CryptographyAlgorithm]);
  initMetadataForClass(AesGcmKey, 'AesGcmKey', VOID, AesKey, VOID, [1]);
  initMetadataForObject(WebCryptoAesGcm, 'WebCryptoAesGcm', VOID, WebCryptoAes, [WebCryptoAes, CryptographyAlgorithm]);
  initMetadataForClass(AesGcmCipher, 'AesGcmCipher', VOID, VOID, [IvAuthenticatedEncryptor, IvAuthenticatedDecryptor], [2, 3, 1]);
  initMetadataForCompanion(Companion);
  initMetadataForClass(WebCryptoDigest, 'WebCryptoDigest', VOID, VOID, [CryptographyAlgorithm], [1]);
  initMetadataForClass(EcPublicKey, 'EcPublicKey', VOID, WebCryptoEncodableKey, VOID, [1]);
  initMetadataForClass(EcPrivateKey_0, 'EcPrivateKey', VOID, WebCryptoEncodableKey, VOID, [1]);
  initMetadataForClass(WebCryptoEc, 'WebCryptoEc', VOID, VOID, [CryptographyAlgorithm]);
  initMetadataForObject(EcPublicKeyProcessor, 'EcPublicKeyProcessor', VOID, WebCryptoKeyProcessor);
  initMetadataForObject(EcPrivateKeyProcessor, 'EcPrivateKeyProcessor', VOID, WebCryptoKeyProcessor);
  initMetadataForClass(EcdhKeyPair, 'EcdhKeyPair');
  initMetadataForClass(EcdhPublicKey, 'EcdhPublicKey', VOID, EcPublicKey, [EcPublicKey, PublicKey, SharedSecretGenerator], [1]);
  initMetadataForClass(EcdhPrivateKey, 'EcdhPrivateKey', VOID, EcPrivateKey_0, [EcPrivateKey_0, PrivateKey, SharedSecretGenerator], [1]);
  initMetadataForObject(WebCryptoEcdh, 'WebCryptoEcdh', VOID, WebCryptoEc, [WebCryptoEc, CryptographyAlgorithm]);
  initMetadataForClass(EcdsaKeyPair, 'EcdsaKeyPair');
  initMetadataForClass(EcdsaPublicKey, 'EcdsaPublicKey', VOID, EcPublicKey, VOID, [1]);
  initMetadataForClass(EcdsaPrivateKey, 'EcdsaPrivateKey', VOID, EcPrivateKey_0, VOID, [1]);
  initMetadataForObject(WebCryptoEcdsa, 'WebCryptoEcdsa', VOID, WebCryptoEc, [WebCryptoEc, CryptographyAlgorithm]);
  initMetadataForCoroutine($deriveSecretToByteArrayCOROUTINE$7, CoroutineImpl);
  initMetadataForClass(HkdfSecretDerivation, 'HkdfSecretDerivation', VOID, VOID, [SecretDerivation], [1]);
  initMetadataForObject(WebCryptoHkdf, 'WebCryptoHkdf', VOID, VOID, [CryptographyAlgorithm]);
  initMetadataForClass(HmacKey, 'HmacKey', VOID, WebCryptoEncodableKey, VOID, [1]);
  initMetadataForObject(WebCryptoHmac, 'WebCryptoHmac', VOID, VOID, [CryptographyAlgorithm]);
  initMetadataForObject(HmacKeyProcessor, 'HmacKeyProcessor', VOID, WebCryptoKeyProcessor);
  initMetadataForObject(WebCryptoPbkdf2, 'WebCryptoPbkdf2', VOID, VOID, [CryptographyAlgorithm]);
  initMetadataForClass(RsaPublicKey, 'RsaPublicKey', VOID, WebCryptoEncodableKey, VOID, [1]);
  initMetadataForClass(RsaPrivateKey, 'RsaPrivateKey', VOID, WebCryptoEncodableKey, VOID, [1]);
  initMetadataForClass(WebCryptoRsa, 'WebCryptoRsa', VOID, VOID, [CryptographyAlgorithm]);
  initMetadataForObject(RsaPublicKeyProcessor, 'RsaPublicKeyProcessor', VOID, WebCryptoKeyProcessor);
  initMetadataForObject(RsaPrivateKeyProcessor, 'RsaPrivateKeyProcessor', VOID, WebCryptoKeyProcessor);
  initMetadataForClass(RsaOaepKeyPair, 'RsaOaepKeyPair');
  initMetadataForClass(RsaOaepPublicKey, 'RsaOaepPublicKey', VOID, RsaPublicKey, VOID, [1]);
  initMetadataForClass(RsaOaepPrivateKey, 'RsaOaepPrivateKey', VOID, RsaPrivateKey, VOID, [1]);
  initMetadataForObject(WebCryptoRsaOaep, 'WebCryptoRsaOaep', VOID, WebCryptoRsa, [WebCryptoRsa, CryptographyAlgorithm]);
  initMetadataForClass(RsaPkcs1KeyPair, 'RsaPkcs1KeyPair');
  initMetadataForClass(RsaPkcs1PublicKey, 'RsaPkcs1PublicKey', VOID, RsaPublicKey, VOID, [1]);
  initMetadataForClass(RsaPkcs1PrivateKey, 'RsaPkcs1PrivateKey', VOID, RsaPrivateKey, VOID, [1]);
  initMetadataForObject(WebCryptoRsaPkcs1, 'WebCryptoRsaPkcs1', VOID, WebCryptoRsa, [WebCryptoRsa, CryptographyAlgorithm]);
  initMetadataForClass(RsaPssKeyPair, 'RsaPssKeyPair');
  initMetadataForClass(RsaPssPublicKey, 'RsaPssPublicKey', VOID, RsaPublicKey, VOID, [1]);
  initMetadataForClass(RsaPssPrivateKey, 'RsaPssPrivateKey', VOID, RsaPrivateKey, VOID, [1]);
  initMetadataForObject(WebCryptoRsaPss, 'WebCryptoRsaPss', VOID, WebCryptoRsa, [WebCryptoRsa, CryptographyAlgorithm]);
  initMetadataForCoroutine($decodeFromByteArrayCOROUTINE$11, CoroutineImpl);
  initMetadataForClass(WebCryptoKeyDecoder, 'WebCryptoKeyDecoder', VOID, VOID, [KeyDecoder], [2]);
  initMetadataForClass(WebCryptoKeyWrapper, 'WebCryptoKeyWrapper');
  initMetadataForClass(WebCryptoSignatureGenerator, 'WebCryptoSignatureGenerator', VOID, VOID, [SignatureGenerator], [1]);
  initMetadataForCoroutine($encryptCOROUTINE$16, CoroutineImpl);
  initMetadataForCoroutine($decryptCOROUTINE$17, CoroutineImpl);
  initMetadataForCoroutine($signCOROUTINE$18, CoroutineImpl);
  initMetadataForCoroutine($deriveBitsCOROUTINE$19, CoroutineImpl);
  initMetadataForObject(WebCrypto, 'WebCrypto', VOID, VOID, VOID, [2, 3, 4, 5]);
  initMetadataForCoroutine($awaitCOROUTINE$21, CoroutineImpl);
  //endregion
  function get_defaultProvider() {
    _init_properties_WebCryptoCryptographyProvider_kt__i22tls();
    return defaultProvider;
  }
  var defaultProvider;
  var WebCrypto$delegate;
  function WebCryptoCryptographyProvider() {
    WebCryptoCryptographyProvider_instance = this;
    CryptographyProvider.call(this);
  }
  protoOf(WebCryptoCryptographyProvider).j1a = function (identifier) {
    var tmp = equals(identifier, SHA1_getInstance()) ? Companion_getInstance_13().z1e_1 : equals(identifier, SHA256_getInstance()) ? Companion_getInstance_13().a1f_1 : equals(identifier, SHA384_getInstance()) ? Companion_getInstance_13().b1f_1 : equals(identifier, SHA512_getInstance()) ? Companion_getInstance_13().c1f_1 : equals(identifier, Companion_getInstance_9()) ? WebCryptoHmac_getInstance() : equals(identifier, Companion_getInstance_8()) ? WebCryptoAesCbc_getInstance() : equals(identifier, Companion_getInstance_7()) ? WebCryptoAesCtr_getInstance() : equals(identifier, Companion_getInstance_6()) ? WebCryptoAesGcm_getInstance() : equals(identifier, Companion_getInstance_5()) ? WebCryptoRsaOaep_getInstance() : equals(identifier, Companion_getInstance_4()) ? WebCryptoRsaPss_getInstance() : equals(identifier, Companion_getInstance_3()) ? WebCryptoRsaPkcs1_getInstance() : equals(identifier, Companion_getInstance_2()) ? WebCryptoEcdsa_getInstance() : equals(identifier, Companion_getInstance_1()) ? WebCryptoEcdh_getInstance() : equals(identifier, Companion_getInstance_0()) ? WebCryptoPbkdf2_instance : equals(identifier, Companion_getInstance()) ? WebCryptoHkdf_instance : null;
    return (tmp == null ? true : isInterface(tmp, CryptographyAlgorithm)) ? tmp : THROW_CCE();
  };
  var WebCryptoCryptographyProvider_instance;
  function WebCryptoCryptographyProvider_getInstance() {
    if (WebCryptoCryptographyProvider_instance == null)
      new WebCryptoCryptographyProvider();
    return WebCryptoCryptographyProvider_instance;
  }
  function defaultProvider$lambda() {
    _init_properties_WebCryptoCryptographyProvider_kt__i22tls();
    return WebCryptoCryptographyProvider_getInstance();
  }
  var properties_initialized_WebCryptoCryptographyProvider_kt_n89q9u;
  function _init_properties_WebCryptoCryptographyProvider_kt__i22tls() {
    if (!properties_initialized_WebCryptoCryptographyProvider_kt_n89q9u) {
      properties_initialized_WebCryptoCryptographyProvider_kt_n89q9u = true;
      defaultProvider = lazy(defaultProvider$lambda);
      WebCrypto$delegate = get_defaultProvider();
    }
  }
  function AesKey(key) {
    WebCryptoEncodableKey.call(this, key, AesKeyProcessor_getInstance());
    this.f1f_1 = key;
  }
  function WebCryptoAes(algorithmName, keyWrapper) {
    this.g1f_1 = algorithmName;
    this.h1f_1 = keyWrapper;
  }
  protoOf(WebCryptoAes).i1f = function () {
    return new WebCryptoKeyDecoder(Algorithm(this.g1f_1), AesKeyProcessor_getInstance(), this.h1f_1);
  };
  function AesKeyProcessor() {
    AesKeyProcessor_instance = this;
    WebCryptoKeyProcessor.call(this);
  }
  protoOf(AesKeyProcessor).j1f = function (format) {
    var tmp;
    switch (format.v2_1) {
      case 0:
        tmp = 'raw';
        break;
      case 1:
        tmp = 'jwk';
        break;
      default:
        noWhenBranchMatchedException();
        break;
    }
    return tmp;
  };
  protoOf(AesKeyProcessor).k1f = function (format) {
    return this.j1f(format instanceof Format ? format : THROW_CCE());
  };
  protoOf(AesKeyProcessor).l1f = function (algorithm, format, key) {
    return key;
  };
  protoOf(AesKeyProcessor).m1f = function (algorithm, format, key) {
    return this.l1f(algorithm, format instanceof Format ? format : THROW_CCE(), key);
  };
  var AesKeyProcessor_instance;
  function AesKeyProcessor_getInstance() {
    if (AesKeyProcessor_instance == null)
      new AesKeyProcessor();
    return AesKeyProcessor_instance;
  }
  function AesCbcKey(key) {
    AesKey.call(this, key);
  }
  function WebCryptoAesCbc$AesCbcKey$_init_$ref_lop2vf() {
    var l = function (p0) {
      return new AesCbcKey(p0);
    };
    l.callableName = '<init>';
    return l;
  }
  function WebCryptoAesCbc() {
    WebCryptoAesCbc_instance = this;
    // Inline function 'kotlin.arrayOf' call
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    var tmp = ['encrypt', 'decrypt'];
    WebCryptoAes.call(this, 'AES-CBC', new WebCryptoKeyWrapper(tmp, WebCryptoAesCbc$AesCbcKey$_init_$ref_lop2vf()));
  }
  var WebCryptoAesCbc_instance;
  function WebCryptoAesCbc_getInstance() {
    if (WebCryptoAesCbc_instance == null)
      new WebCryptoAesCbc();
    return WebCryptoAesCbc_instance;
  }
  function AesCtrKey(key) {
    AesKey.call(this, key);
  }
  function WebCryptoAesCtr$AesCtrKey$_init_$ref_5qzjvv() {
    var l = function (p0) {
      return new AesCtrKey(p0);
    };
    l.callableName = '<init>';
    return l;
  }
  function WebCryptoAesCtr() {
    WebCryptoAesCtr_instance = this;
    // Inline function 'kotlin.arrayOf' call
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    var tmp = ['encrypt', 'decrypt'];
    WebCryptoAes.call(this, 'AES-CTR', new WebCryptoKeyWrapper(tmp, WebCryptoAesCtr$AesCtrKey$_init_$ref_5qzjvv()));
  }
  var WebCryptoAesCtr_instance;
  function WebCryptoAesCtr_getInstance() {
    if (WebCryptoAesCtr_instance == null)
      new WebCryptoAesCtr();
    return WebCryptoAesCtr_instance;
  }
  function AesGcmKey(key) {
    AesKey.call(this, key);
  }
  protoOf(AesGcmKey).q1f = function (tagSize) {
    return new AesGcmCipher(this.f1f_1, _BinarySize___get_inBits__impl__p3mjjl(tagSize));
  };
  function WebCryptoAesGcm$AesGcmKey$_init_$ref_c6pr3f() {
    var l = function (p0) {
      return new AesGcmKey(p0);
    };
    l.callableName = '<init>';
    return l;
  }
  function WebCryptoAesGcm() {
    WebCryptoAesGcm_instance = this;
    // Inline function 'kotlin.arrayOf' call
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    var tmp = ['encrypt', 'decrypt'];
    WebCryptoAes.call(this, 'AES-GCM', new WebCryptoKeyWrapper(tmp, WebCryptoAesGcm$AesGcmKey$_init_$ref_c6pr3f()));
  }
  var WebCryptoAesGcm_instance;
  function WebCryptoAesGcm_getInstance() {
    if (WebCryptoAesGcm_instance == null)
      new WebCryptoAesGcm();
    return WebCryptoAesGcm_instance;
  }
  function AesGcmCipher(key, tagSizeBits) {
    this.r1f_1 = key;
    this.s1f_1 = tagSizeBits;
  }
  protoOf(AesGcmCipher).l1a = function (iv, plaintext, associatedData, $completion) {
    return WebCrypto_getInstance().u1f(AesGcmCipherAlgorithm(associatedData, iv, this.s1f_1), this.r1f_1, plaintext, $completion);
  };
  protoOf(AesGcmCipher).n1a = function (iv, ciphertext, associatedData, $completion) {
    return WebCrypto_getInstance().v1f(AesGcmCipherAlgorithm(associatedData, iv, this.s1f_1), this.r1f_1, ciphertext, $completion);
  };
  protoOf(AesGcmCipher).o1a = function (iv, ciphertext, associatedData) {
    nonBlocking();
  };
  protoOf(AesGcmCipher).m1a = function (iv, plaintext, associatedData) {
    nonBlocking();
  };
  function Companion() {
    Companion_instance_4 = this;
    this.z1e_1 = new WebCryptoDigest('SHA-1', SHA1_getInstance());
    this.a1f_1 = new WebCryptoDigest('SHA-256', SHA256_getInstance());
    this.b1f_1 = new WebCryptoDigest('SHA-384', SHA384_getInstance());
    this.c1f_1 = new WebCryptoDigest('SHA-512', SHA512_getInstance());
  }
  var Companion_instance_4;
  function Companion_getInstance_13() {
    if (Companion_instance_4 == null)
      new Companion();
    return Companion_instance_4;
  }
  function WebCryptoDigest(algorithm, id) {
    Companion_getInstance_13();
    this.w1f_1 = algorithm;
    this.x1f_1 = id;
  }
  function EcPublicKey(publicKey) {
    WebCryptoEncodableKey.call(this, publicKey, EcPublicKeyProcessor_getInstance());
    this.a1g_1 = publicKey;
  }
  function EcPrivateKey_0(privateKey) {
    WebCryptoEncodableKey.call(this, privateKey, EcPrivateKeyProcessor_getInstance());
    this.d1g_1 = privateKey;
  }
  function WebCryptoEc$keyPairWrapper$lambda($keyPairWrapper, this$0) {
    return function (it) {
      return $keyPairWrapper(this$0.h1g_1.f1g_1(it.publicKey), this$0.i1g_1.f1g_1(it.privateKey));
    };
  }
  function WebCryptoEc(algorithmName, publicKeyWrapper, privateKeyWrapper, keyPairWrapper) {
    this.g1g_1 = algorithmName;
    this.h1g_1 = publicKeyWrapper;
    this.i1g_1 = privateKeyWrapper;
    var tmp = this;
    // Inline function 'kotlin.collections.plus' call
    var this_0 = this.h1g_1.e1g_1;
    var elements = this.i1g_1.e1g_1;
    // Inline function 'kotlin.js.asDynamic' call
    tmp.j1g_1 = this_0.concat(elements);
    var tmp_0 = this;
    tmp_0.k1g_1 = WebCryptoEc$keyPairWrapper$lambda(keyPairWrapper, this);
  }
  protoOf(WebCryptoEc).l1g = function (curve) {
    return new WebCryptoKeyDecoder(EcKeyAlgorithm(this.g1g_1, _Curve___get_name__impl__klrusc(curve)), EcPublicKeyProcessor_getInstance(), this.h1g_1);
  };
  protoOf(WebCryptoEc).m1g = function (curve) {
    return new WebCryptoKeyDecoder(EcKeyAlgorithm(this.g1g_1, _Curve___get_name__impl__klrusc(curve)), EcPrivateKeyProcessor_getInstance(), this.i1g_1);
  };
  function EcPublicKeyProcessor() {
    EcPublicKeyProcessor_instance = this;
    WebCryptoKeyProcessor.call(this);
  }
  protoOf(EcPublicKeyProcessor).n1g = function (format) {
    var tmp;
    if (equals(format, JWK_getInstance())) {
      tmp = 'jwk';
    } else if (equals(format, RAW_getInstance())) {
      tmp = 'raw';
    } else if (equals(format, DER_getInstance()) || equals(format, PEM_getInstance())) {
      tmp = 'spki';
    } else {
      noWhenBranchMatchedException();
    }
    return tmp;
  };
  protoOf(EcPublicKeyProcessor).k1f = function (format) {
    return this.n1g(format instanceof Format_0 ? format : THROW_CCE());
  };
  protoOf(EcPublicKeyProcessor).o1g = function (algorithm, format, key) {
    var tmp;
    if (equals(format, JWK_getInstance())) {
      tmp = key;
    } else if (equals(format, RAW_getInstance())) {
      tmp = key;
    } else if (equals(format, DER_getInstance())) {
      tmp = key;
    } else if (equals(format, PEM_getInstance())) {
      tmp = unwrapPem(Companion_getInstance_10().m1b_1, key);
    } else {
      noWhenBranchMatchedException();
    }
    return tmp;
  };
  protoOf(EcPublicKeyProcessor).m1f = function (algorithm, format, key) {
    return this.o1g(algorithm, format instanceof Format_0 ? format : THROW_CCE(), key);
  };
  var EcPublicKeyProcessor_instance;
  function EcPublicKeyProcessor_getInstance() {
    if (EcPublicKeyProcessor_instance == null)
      new EcPublicKeyProcessor();
    return EcPublicKeyProcessor_instance;
  }
  function convertSec1ToPkcs8($this, input) {
    var ecPrivateKey = Default_getInstance().z1b(Companion_instance.k19(), input);
    $l$block: {
      // Inline function 'kotlin.checkNotNull' call
      var tmp = ecPrivateKey.i1e_1;
      var value = tmp == null ? null : new EcParameters(tmp);
      // Inline function 'kotlin.contracts.contract' call
      if (value == null) {
        // Inline function 'dev.whyoleg.cryptography.providers.webcrypto.algorithms.EcPrivateKeyProcessor.convertSec1ToPkcs8.<anonymous>' call
        var message = 'EC Parameters are not present in the key';
        throw IllegalStateException_init_$Create$(toString(message));
      } else {
        value.c1e_1;
        break $l$block;
      }
    }
    var privateKeyInfo = new PrivateKeyInfo(0, new EcKeyAlgorithmIdentifier(ecPrivateKey.i1e_1), input);
    return Default_getInstance().y1b(Companion_getInstance_11().k19(), privateKeyInfo);
  }
  function EcPrivateKeyProcessor() {
    EcPrivateKeyProcessor_instance = this;
    WebCryptoKeyProcessor.call(this);
  }
  protoOf(EcPrivateKeyProcessor).p1g = function (format) {
    var tmp;
    if (equals(format, JWK_getInstance_0())) {
      tmp = 'jwk';
    } else if (equals(format, RAW_getInstance_0()) || equals(format, Generic_getInstance()) || (equals(format, Generic_getInstance_0()) || (equals(format, SEC1_getInstance()) || equals(format, SEC1_getInstance_0())))) {
      tmp = 'pkcs8';
    } else {
      noWhenBranchMatchedException();
    }
    return tmp;
  };
  protoOf(EcPrivateKeyProcessor).k1f = function (format) {
    return this.p1g(format instanceof Format_1 ? format : THROW_CCE());
  };
  protoOf(EcPrivateKeyProcessor).q1g = function (algorithm, format, key) {
    var tmp;
    if (equals(format, JWK_getInstance_0())) {
      tmp = key;
    } else if (equals(format, RAW_getInstance_0())) {
      var namedCurve = get_ecKeyAlgorithmNamedCurve(algorithm);
      var tmp_0;
      if (namedCurve === _Curve___get_name__impl__klrusc(Companion_instance_1.p1a())) {
        tmp_0 = get_secp256r1(Companion_instance_0);
      } else if (namedCurve === _Curve___get_name__impl__klrusc(Companion_instance_1.q1a())) {
        tmp_0 = get_secp384r1(Companion_instance_0);
      } else if (namedCurve === _Curve___get_name__impl__klrusc(Companion_instance_1.r1a())) {
        tmp_0 = get_secp521r1(Companion_instance_0);
      } else {
        var message = 'Unknown curve: ' + namedCurve;
        throw IllegalStateException_init_$Create$(toString(message));
      }
      var curveId = tmp_0;
      tmp = wrapPrivateKey(0, new EcKeyAlgorithmIdentifier(_EcParameters___init__impl__i4ww9z(curveId)), Default_getInstance().y1b(Companion_instance.k19(), new EcPrivateKey(1, key)));
    } else if (equals(format, Generic_getInstance())) {
      tmp = key;
    } else if (equals(format, Generic_getInstance_0())) {
      tmp = unwrapPem(Companion_getInstance_10().n1b_1, key);
    } else if (equals(format, SEC1_getInstance())) {
      tmp = convertSec1ToPkcs8(this, key);
    } else if (equals(format, SEC1_getInstance_0())) {
      tmp = convertSec1ToPkcs8(this, unwrapPem(Companion_getInstance_10().q1b_1, key));
    } else {
      noWhenBranchMatchedException();
    }
    return tmp;
  };
  protoOf(EcPrivateKeyProcessor).m1f = function (algorithm, format, key) {
    return this.q1g(algorithm, format instanceof Format_1 ? format : THROW_CCE(), key);
  };
  var EcPrivateKeyProcessor_instance;
  function EcPrivateKeyProcessor_getInstance() {
    if (EcPrivateKeyProcessor_instance == null)
      new EcPrivateKeyProcessor();
    return EcPrivateKeyProcessor_instance;
  }
  function EcdhKeyPair(publicKey, privateKey) {
    this.r1g_1 = publicKey;
    this.s1g_1 = privateKey;
  }
  function EcdhPublicKey(publicKey) {
    EcPublicKey.call(this, publicKey);
  }
  protoOf(EcdhPublicKey).w1g = function (other, $completion) {
    // Inline function 'kotlin.check' call
    // Inline function 'kotlin.contracts.contract' call
    // Inline function 'kotlin.check' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(other instanceof EcdhPrivateKey)) {
      // Inline function 'kotlin.check.<anonymous>' call
      var message = 'Check failed.';
      throw IllegalStateException_init_$Create$(toString(message));
    }
    return WebCrypto_getInstance().x1g(EcdhKeyDeriveAlgorithm(this.a1g_1), other.d1g_1, _BinarySize___get_inBits__impl__p3mjjl(curveOrderSize(get_ecKeyAlgorithmNamedCurve(get_algorithm(this.a1g_1)))), $completion);
  };
  protoOf(EcdhPublicKey).x1a = function (other, $completion) {
    return this.w1g(isInterface(other, PrivateKey) ? other : THROW_CCE(), $completion);
  };
  protoOf(EcdhPublicKey).y1g = function (other) {
    nonBlocking();
  };
  protoOf(EcdhPublicKey).y1a = function (other) {
    return this.y1g(isInterface(other, PrivateKey) ? other : THROW_CCE());
  };
  function EcdhPrivateKey(privateKey) {
    EcPrivateKey_0.call(this, privateKey);
  }
  protoOf(EcdhPrivateKey).s1a = function () {
    return this;
  };
  protoOf(EcdhPrivateKey).c1h = function (other, $completion) {
    // Inline function 'kotlin.check' call
    // Inline function 'kotlin.contracts.contract' call
    // Inline function 'kotlin.check' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(other instanceof EcdhPublicKey)) {
      // Inline function 'kotlin.check.<anonymous>' call
      var message = 'Check failed.';
      throw IllegalStateException_init_$Create$(toString(message));
    }
    return WebCrypto_getInstance().x1g(EcdhKeyDeriveAlgorithm(other.a1g_1), this.d1g_1, _BinarySize___get_inBits__impl__p3mjjl(curveOrderSize(get_ecKeyAlgorithmNamedCurve(get_algorithm(this.d1g_1)))), $completion);
  };
  protoOf(EcdhPrivateKey).x1a = function (other, $completion) {
    return this.c1h(isInterface(other, PublicKey) ? other : THROW_CCE(), $completion);
  };
  protoOf(EcdhPrivateKey).d1h = function (other) {
    nonBlocking();
  };
  protoOf(EcdhPrivateKey).y1a = function (other) {
    return this.d1h(isInterface(other, PublicKey) ? other : THROW_CCE());
  };
  function WebCryptoEcdh$EcdhPublicKey$_init_$ref_x8kjxe() {
    var l = function (p0) {
      return new EcdhPublicKey(p0);
    };
    l.callableName = '<init>';
    return l;
  }
  function WebCryptoEcdh$EcdhPrivateKey$_init_$ref_1xgfie() {
    var l = function (p0) {
      return new EcdhPrivateKey(p0);
    };
    l.callableName = '<init>';
    return l;
  }
  function WebCryptoEcdh$EcdhKeyPair$_init_$ref_fqllqz() {
    var l = function (p0, p1) {
      return new EcdhKeyPair(p0, p1);
    };
    l.callableName = '<init>';
    return l;
  }
  function WebCryptoEcdh() {
    WebCryptoEcdh_instance = this;
    // Inline function 'kotlin.arrayOf' call
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    var tmp = [];
    var tmp_0 = new WebCryptoKeyWrapper(tmp, WebCryptoEcdh$EcdhPublicKey$_init_$ref_x8kjxe());
    // Inline function 'kotlin.arrayOf' call
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    var tmp_1 = ['deriveBits'];
    var tmp_2 = new WebCryptoKeyWrapper(tmp_1, WebCryptoEcdh$EcdhPrivateKey$_init_$ref_1xgfie());
    WebCryptoEc.call(this, 'ECDH', tmp_0, tmp_2, WebCryptoEcdh$EcdhKeyPair$_init_$ref_fqllqz());
  }
  var WebCryptoEcdh_instance;
  function WebCryptoEcdh_getInstance() {
    if (WebCryptoEcdh_instance == null)
      new WebCryptoEcdh();
    return WebCryptoEcdh_instance;
  }
  function EcdsaKeyPair(publicKey, privateKey) {
    this.e1h_1 = publicKey;
    this.f1h_1 = privateKey;
  }
  function EcdsaPublicKey(publicKey) {
    EcPublicKey.call(this, publicKey);
  }
  function EcdsaPrivateKey(privateKey) {
    EcPrivateKey_0.call(this, privateKey);
  }
  function WebCryptoEcdsa$EcdsaPublicKey$_init_$ref_5egfgi() {
    var l = function (p0) {
      return new EcdsaPublicKey(p0);
    };
    l.callableName = '<init>';
    return l;
  }
  function WebCryptoEcdsa$EcdsaPrivateKey$_init_$ref_8mqrbu() {
    var l = function (p0) {
      return new EcdsaPrivateKey(p0);
    };
    l.callableName = '<init>';
    return l;
  }
  function WebCryptoEcdsa$EcdsaKeyPair$_init_$ref_l5vyez() {
    var l = function (p0, p1) {
      return new EcdsaKeyPair(p0, p1);
    };
    l.callableName = '<init>';
    return l;
  }
  function WebCryptoEcdsa() {
    WebCryptoEcdsa_instance = this;
    // Inline function 'kotlin.arrayOf' call
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    var tmp = ['verify'];
    var tmp_0 = new WebCryptoKeyWrapper(tmp, WebCryptoEcdsa$EcdsaPublicKey$_init_$ref_5egfgi());
    // Inline function 'kotlin.arrayOf' call
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    var tmp_1 = ['sign'];
    var tmp_2 = new WebCryptoKeyWrapper(tmp_1, WebCryptoEcdsa$EcdsaPrivateKey$_init_$ref_8mqrbu());
    WebCryptoEc.call(this, 'ECDSA', tmp_0, tmp_2, WebCryptoEcdsa$EcdsaKeyPair$_init_$ref_l5vyez());
  }
  var WebCryptoEcdsa_instance;
  function WebCryptoEcdsa_getInstance() {
    if (WebCryptoEcdsa_instance == null)
      new WebCryptoEcdsa();
    return WebCryptoEcdsa_instance;
  }
  function curveOrderSize(namedCurve) {
    var tmp = Companion_instance_2;
    var tmp_0;
    if (namedCurve === _Curve___get_name__impl__klrusc(Companion_instance_1.p1a())) {
      tmp_0 = 32;
    } else if (namedCurve === _Curve___get_name__impl__klrusc(Companion_instance_1.q1a())) {
      tmp_0 = 48;
    } else if (namedCurve === _Curve___get_name__impl__klrusc(Companion_instance_1.r1a())) {
      tmp_0 = 66;
    } else {
      var message = 'Unknown curve: ' + namedCurve;
      throw IllegalStateException_init_$Create$(toString(message));
    }
    return tmp.c1a(tmp_0);
  }
  function $deriveSecretToByteArrayCOROUTINE$7(_this__u8e3s4, input, resultContinuation) {
    CoroutineImpl.call(this, resultContinuation);
    this.o1h_1 = _this__u8e3s4;
    this.p1h_1 = input;
  }
  protoOf($deriveSecretToByteArrayCOROUTINE$7).ba = function () {
    var suspendResult = this.l9_1;
    $sm: do
      try {
        var tmp = this.j9_1;
        switch (tmp) {
          case 0:
            this.k9_1 = 3;
            this.j9_1 = 1;
            var tmp_0 = WebCrypto_getInstance();
            var tmp_1 = Algorithm('HKDF');
            var this_0 = ['deriveBits'];
            suspendResult = tmp_0.r1h('raw', this.p1h_1, tmp_1, false, this_0, this);
            if (suspendResult === get_COROUTINE_SUSPENDED()) {
              return suspendResult;
            }

            continue $sm;
          case 1:
            this.q1h_1 = suspendResult;
            this.j9_1 = 2;
            suspendResult = WebCrypto_getInstance().x1g(this.o1h_1.s1h_1, this.q1h_1, _BinarySize___get_inBits__impl__p3mjjl(this.o1h_1.t1h_1), this);
            if (suspendResult === get_COROUTINE_SUSPENDED()) {
              return suspendResult;
            }

            continue $sm;
          case 2:
            return suspendResult;
          case 3:
            throw this.m9_1;
        }
      } catch ($p) {
        var e = $p;
        if (this.k9_1 === 3) {
          throw e;
        } else {
          this.j9_1 = this.k9_1;
          this.m9_1 = e;
        }
      }
     while (true);
  };
  function HkdfSecretDerivation(algorithm, outputSize) {
    this.s1h_1 = algorithm;
    this.t1h_1 = outputSize;
  }
  protoOf(HkdfSecretDerivation).v1a = function (input, $completion) {
    var tmp = new $deriveSecretToByteArrayCOROUTINE$7(this, input, $completion);
    tmp.l9_1 = Unit_instance;
    tmp.m9_1 = null;
    return tmp.ba();
  };
  protoOf(HkdfSecretDerivation).w1a = function (input) {
    nonBlocking();
  };
  function WebCryptoHkdf() {
  }
  protoOf(WebCryptoHkdf).u1h = function (digest, outputSize, salt, info) {
    var tmp = hashAlgorithmName(digest);
    var tmp_0;
    if (salt == null) {
      tmp_0 = null;
    } else {
      // Inline function 'kotlin.takeIf' call
      // Inline function 'kotlin.contracts.contract' call
      var tmp_1;
      // Inline function 'kotlin.collections.isNotEmpty' call
      // Inline function 'kotlin.collections.isNotEmpty' call
      // Inline function 'kotlin.collections.isEmpty' call
      if (!(salt.length === 0)) {
        tmp_1 = salt;
      } else {
        tmp_1 = null;
      }
      tmp_0 = tmp_1;
    }
    var tmp1_elvis_lhs = tmp_0;
    var tmp_2 = tmp1_elvis_lhs == null ? new Int8Array(digestSize(digest)) : tmp1_elvis_lhs;
    var tmp_3;
    if (info == null) {
      tmp_3 = null;
    } else {
      // Inline function 'kotlin.takeIf' call
      // Inline function 'kotlin.contracts.contract' call
      var tmp_4;
      // Inline function 'kotlin.collections.isNotEmpty' call
      // Inline function 'kotlin.collections.isNotEmpty' call
      // Inline function 'kotlin.collections.isEmpty' call
      if (!(info.length === 0)) {
        tmp_4 = info;
      } else {
        tmp_4 = null;
      }
      tmp_3 = tmp_4;
    }
    var tmp3_elvis_lhs = tmp_3;
    return new HkdfSecretDerivation(HkdfDeriveAlgorithm(tmp, tmp_2, tmp3_elvis_lhs == null ? get_EmptyByteArray() : tmp3_elvis_lhs), outputSize);
  };
  var WebCryptoHkdf_instance;
  function WebCryptoHkdf_getInstance() {
    return WebCryptoHkdf_instance;
  }
  function HmacKey(key) {
    WebCryptoEncodableKey.call(this, key, HmacKeyProcessor_getInstance());
    this.x1h_1 = key;
  }
  protoOf(HmacKey).y1h = function () {
    return new WebCryptoSignatureGenerator(Algorithm('HMAC'), this.x1h_1);
  };
  function WebCryptoHmac$HmacKey$_init_$ref_enwl2t() {
    var l = function (p0) {
      return new HmacKey(p0);
    };
    l.callableName = '<init>';
    return l;
  }
  function WebCryptoHmac() {
    WebCryptoHmac_instance = this;
    var tmp = this;
    // Inline function 'kotlin.arrayOf' call
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    var tmp_0 = ['sign', 'verify'];
    tmp.z1h_1 = new WebCryptoKeyWrapper(tmp_0, WebCryptoHmac$HmacKey$_init_$ref_enwl2t());
  }
  protoOf(WebCryptoHmac).a1i = function (digest) {
    return new WebCryptoKeyDecoder(HmacKeyAlgorithm(hashAlgorithmName(digest), null), HmacKeyProcessor_getInstance(), this.z1h_1);
  };
  var WebCryptoHmac_instance;
  function WebCryptoHmac_getInstance() {
    if (WebCryptoHmac_instance == null)
      new WebCryptoHmac();
    return WebCryptoHmac_instance;
  }
  function HmacKeyProcessor() {
    HmacKeyProcessor_instance = this;
    WebCryptoKeyProcessor.call(this);
  }
  protoOf(HmacKeyProcessor).b1i = function (format) {
    var tmp;
    switch (format.v2_1) {
      case 0:
        tmp = 'raw';
        break;
      case 1:
        tmp = 'jwk';
        break;
      default:
        noWhenBranchMatchedException();
        break;
    }
    return tmp;
  };
  protoOf(HmacKeyProcessor).k1f = function (format) {
    return this.b1i(format instanceof Format_2 ? format : THROW_CCE());
  };
  protoOf(HmacKeyProcessor).c1i = function (algorithm, format, key) {
    return key;
  };
  protoOf(HmacKeyProcessor).m1f = function (algorithm, format, key) {
    return this.c1i(algorithm, format instanceof Format_2 ? format : THROW_CCE(), key);
  };
  var HmacKeyProcessor_instance;
  function HmacKeyProcessor_getInstance() {
    if (HmacKeyProcessor_instance == null)
      new HmacKeyProcessor();
    return HmacKeyProcessor_instance;
  }
  function WebCryptoPbkdf2() {
  }
  var WebCryptoPbkdf2_instance;
  function WebCryptoPbkdf2_getInstance() {
    return WebCryptoPbkdf2_instance;
  }
  function RsaPublicKey(publicKey) {
    WebCryptoEncodableKey.call(this, publicKey, RsaPublicKeyProcessor_getInstance());
    this.f1i_1 = publicKey;
  }
  function RsaPrivateKey(privateKey) {
    WebCryptoEncodableKey.call(this, privateKey, RsaPrivateKeyProcessor_getInstance());
    this.i1i_1 = privateKey;
  }
  function WebCryptoRsa$keyPairWrapper$lambda($keyPairWrapper, this$0) {
    return function (it) {
      return $keyPairWrapper(this$0.k1i_1.f1g_1(it.publicKey), this$0.l1i_1.f1g_1(it.privateKey));
    };
  }
  function WebCryptoRsa(algorithmName, publicKeyWrapper, privateKeyWrapper, keyPairWrapper) {
    this.j1i_1 = algorithmName;
    this.k1i_1 = publicKeyWrapper;
    this.l1i_1 = privateKeyWrapper;
    var tmp = this;
    // Inline function 'kotlin.collections.plus' call
    var this_0 = this.k1i_1.e1g_1;
    var elements = this.l1i_1.e1g_1;
    // Inline function 'kotlin.js.asDynamic' call
    tmp.m1i_1 = this_0.concat(elements);
    var tmp_0 = this;
    tmp_0.n1i_1 = WebCryptoRsa$keyPairWrapper$lambda(keyPairWrapper, this);
  }
  function RsaPublicKeyProcessor() {
    RsaPublicKeyProcessor_instance = this;
    WebCryptoKeyProcessor.call(this);
  }
  protoOf(RsaPublicKeyProcessor).o1i = function (format) {
    var tmp;
    if (equals(format, JWK_getInstance_1())) {
      tmp = 'jwk';
    } else if (equals(format, Generic_getInstance_1()) || equals(format, Generic_getInstance_2()) || (equals(format, PKCS1_getInstance()) || equals(format, PKCS1_getInstance_0()))) {
      tmp = 'spki';
    } else {
      noWhenBranchMatchedException();
    }
    return tmp;
  };
  protoOf(RsaPublicKeyProcessor).k1f = function (format) {
    return this.o1i(format instanceof Format_3 ? format : THROW_CCE());
  };
  protoOf(RsaPublicKeyProcessor).p1i = function (algorithm, format, key) {
    var tmp;
    if (equals(format, JWK_getInstance_1())) {
      tmp = key;
    } else if (equals(format, Generic_getInstance_1())) {
      tmp = key;
    } else if (equals(format, Generic_getInstance_2())) {
      tmp = unwrapPem(Companion_getInstance_10().m1b_1, key);
    } else if (equals(format, PKCS1_getInstance())) {
      tmp = wrapPublicKey(RsaKeyAlgorithmIdentifier_instance, key);
    } else if (equals(format, PKCS1_getInstance_0())) {
      tmp = wrapPublicKey(RsaKeyAlgorithmIdentifier_instance, unwrapPem(Companion_getInstance_10().o1b_1, key));
    } else {
      noWhenBranchMatchedException();
    }
    return tmp;
  };
  protoOf(RsaPublicKeyProcessor).m1f = function (algorithm, format, key) {
    return this.p1i(algorithm, format instanceof Format_3 ? format : THROW_CCE(), key);
  };
  var RsaPublicKeyProcessor_instance;
  function RsaPublicKeyProcessor_getInstance() {
    if (RsaPublicKeyProcessor_instance == null)
      new RsaPublicKeyProcessor();
    return RsaPublicKeyProcessor_instance;
  }
  function RsaPrivateKeyProcessor() {
    RsaPrivateKeyProcessor_instance = this;
    WebCryptoKeyProcessor.call(this);
  }
  protoOf(RsaPrivateKeyProcessor).q1i = function (format) {
    var tmp;
    if (equals(format, JWK_getInstance_2())) {
      tmp = 'jwk';
    } else if (equals(format, Generic_getInstance_3()) || equals(format, Generic_getInstance_4()) || (equals(format, PKCS1_getInstance_1()) || equals(format, PKCS1_getInstance_2()))) {
      tmp = 'pkcs8';
    } else {
      noWhenBranchMatchedException();
    }
    return tmp;
  };
  protoOf(RsaPrivateKeyProcessor).k1f = function (format) {
    return this.q1i(format instanceof Format_4 ? format : THROW_CCE());
  };
  protoOf(RsaPrivateKeyProcessor).r1i = function (algorithm, format, key) {
    var tmp;
    if (equals(format, JWK_getInstance_2())) {
      tmp = key;
    } else if (equals(format, Generic_getInstance_3())) {
      tmp = key;
    } else if (equals(format, Generic_getInstance_4())) {
      tmp = unwrapPem(Companion_getInstance_10().n1b_1, key);
    } else if (equals(format, PKCS1_getInstance_1())) {
      tmp = wrapPrivateKey(0, RsaKeyAlgorithmIdentifier_instance, key);
    } else if (equals(format, PKCS1_getInstance_2())) {
      tmp = wrapPrivateKey(0, RsaKeyAlgorithmIdentifier_instance, unwrapPem(Companion_getInstance_10().p1b_1, key));
    } else {
      noWhenBranchMatchedException();
    }
    return tmp;
  };
  protoOf(RsaPrivateKeyProcessor).m1f = function (algorithm, format, key) {
    return this.r1i(algorithm, format instanceof Format_4 ? format : THROW_CCE(), key);
  };
  var RsaPrivateKeyProcessor_instance;
  function RsaPrivateKeyProcessor_getInstance() {
    if (RsaPrivateKeyProcessor_instance == null)
      new RsaPrivateKeyProcessor();
    return RsaPrivateKeyProcessor_instance;
  }
  function RsaOaepKeyPair(publicKey, privateKey) {
    this.s1i_1 = publicKey;
    this.t1i_1 = privateKey;
  }
  function RsaOaepPublicKey(publicKey) {
    RsaPublicKey.call(this, publicKey);
  }
  function RsaOaepPrivateKey(privateKey) {
    RsaPrivateKey.call(this, privateKey);
  }
  function WebCryptoRsaOaep$RsaOaepPublicKey$_init_$ref_2xzkj4() {
    var l = function (p0) {
      return new RsaOaepPublicKey(p0);
    };
    l.callableName = '<init>';
    return l;
  }
  function WebCryptoRsaOaep$RsaOaepPrivateKey$_init_$ref_ds5c3s() {
    var l = function (p0) {
      return new RsaOaepPrivateKey(p0);
    };
    l.callableName = '<init>';
    return l;
  }
  function WebCryptoRsaOaep$RsaOaepKeyPair$_init_$ref_whtofn() {
    var l = function (p0, p1) {
      return new RsaOaepKeyPair(p0, p1);
    };
    l.callableName = '<init>';
    return l;
  }
  function WebCryptoRsaOaep() {
    WebCryptoRsaOaep_instance = this;
    // Inline function 'kotlin.arrayOf' call
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    var tmp = ['encrypt'];
    var tmp_0 = new WebCryptoKeyWrapper(tmp, WebCryptoRsaOaep$RsaOaepPublicKey$_init_$ref_2xzkj4());
    // Inline function 'kotlin.arrayOf' call
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    var tmp_1 = ['decrypt'];
    var tmp_2 = new WebCryptoKeyWrapper(tmp_1, WebCryptoRsaOaep$RsaOaepPrivateKey$_init_$ref_ds5c3s());
    WebCryptoRsa.call(this, 'RSA-OAEP', tmp_0, tmp_2, WebCryptoRsaOaep$RsaOaepKeyPair$_init_$ref_whtofn());
  }
  var WebCryptoRsaOaep_instance;
  function WebCryptoRsaOaep_getInstance() {
    if (WebCryptoRsaOaep_instance == null)
      new WebCryptoRsaOaep();
    return WebCryptoRsaOaep_instance;
  }
  function RsaPkcs1KeyPair(publicKey, privateKey) {
    this.u1i_1 = publicKey;
    this.v1i_1 = privateKey;
  }
  function RsaPkcs1PublicKey(publicKey) {
    RsaPublicKey.call(this, publicKey);
  }
  function RsaPkcs1PrivateKey(privateKey) {
    RsaPrivateKey.call(this, privateKey);
  }
  function WebCryptoRsaPkcs1$RsaPkcs1PublicKey$_init_$ref_kiwvhe() {
    var l = function (p0) {
      return new RsaPkcs1PublicKey(p0);
    };
    l.callableName = '<init>';
    return l;
  }
  function WebCryptoRsaPkcs1$RsaPkcs1PrivateKey$_init_$ref_y0qpl2() {
    var l = function (p0) {
      return new RsaPkcs1PrivateKey(p0);
    };
    l.callableName = '<init>';
    return l;
  }
  function WebCryptoRsaPkcs1$RsaPkcs1KeyPair$_init_$ref_s062w5() {
    var l = function (p0, p1) {
      return new RsaPkcs1KeyPair(p0, p1);
    };
    l.callableName = '<init>';
    return l;
  }
  function WebCryptoRsaPkcs1() {
    WebCryptoRsaPkcs1_instance = this;
    // Inline function 'kotlin.arrayOf' call
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    var tmp = ['verify'];
    var tmp_0 = new WebCryptoKeyWrapper(tmp, WebCryptoRsaPkcs1$RsaPkcs1PublicKey$_init_$ref_kiwvhe());
    // Inline function 'kotlin.arrayOf' call
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    var tmp_1 = ['sign'];
    var tmp_2 = new WebCryptoKeyWrapper(tmp_1, WebCryptoRsaPkcs1$RsaPkcs1PrivateKey$_init_$ref_y0qpl2());
    WebCryptoRsa.call(this, 'RSASSA-PKCS1-v1_5', tmp_0, tmp_2, WebCryptoRsaPkcs1$RsaPkcs1KeyPair$_init_$ref_s062w5());
  }
  var WebCryptoRsaPkcs1_instance;
  function WebCryptoRsaPkcs1_getInstance() {
    if (WebCryptoRsaPkcs1_instance == null)
      new WebCryptoRsaPkcs1();
    return WebCryptoRsaPkcs1_instance;
  }
  function RsaPssKeyPair(publicKey, privateKey) {
    this.w1i_1 = publicKey;
    this.x1i_1 = privateKey;
  }
  function RsaPssPublicKey(publicKey) {
    RsaPublicKey.call(this, publicKey);
  }
  function RsaPssPrivateKey(privateKey) {
    RsaPrivateKey.call(this, privateKey);
  }
  function WebCryptoRsaPss$RsaPssPublicKey$_init_$ref_o813fm() {
    var l = function (p0) {
      return new RsaPssPublicKey(p0);
    };
    l.callableName = '<init>';
    return l;
  }
  function WebCryptoRsaPss$RsaPssPrivateKey$_init_$ref_6l5g3q() {
    var l = function (p0) {
      return new RsaPssPrivateKey(p0);
    };
    l.callableName = '<init>';
    return l;
  }
  function WebCryptoRsaPss$RsaPssKeyPair$_init_$ref_3h8asb() {
    var l = function (p0, p1) {
      return new RsaPssKeyPair(p0, p1);
    };
    l.callableName = '<init>';
    return l;
  }
  function WebCryptoRsaPss() {
    WebCryptoRsaPss_instance = this;
    // Inline function 'kotlin.arrayOf' call
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    var tmp = ['verify'];
    var tmp_0 = new WebCryptoKeyWrapper(tmp, WebCryptoRsaPss$RsaPssPublicKey$_init_$ref_o813fm());
    // Inline function 'kotlin.arrayOf' call
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    var tmp_1 = ['sign'];
    var tmp_2 = new WebCryptoKeyWrapper(tmp_1, WebCryptoRsaPss$RsaPssPrivateKey$_init_$ref_6l5g3q());
    WebCryptoRsa.call(this, 'RSA-PSS', tmp_0, tmp_2, WebCryptoRsaPss$RsaPssKeyPair$_init_$ref_3h8asb());
  }
  var WebCryptoRsaPss_instance;
  function WebCryptoRsaPss_getInstance() {
    if (WebCryptoRsaPss_instance == null)
      new WebCryptoRsaPss();
    return WebCryptoRsaPss_instance;
  }
  function unwrapPem(label, key) {
    return ensurePemLabel(Pem_instance.i1b(key), label).l1b_1;
  }
  function wrapPrivateKey(version, identifier, key) {
    return Default_getInstance().y1b(Companion_getInstance_11().k19(), new PrivateKeyInfo(version, identifier, key));
  }
  function wrapPublicKey(identifier, key) {
    return Default_getInstance().y1b(Companion_getInstance_12().k19(), new SubjectPublicKeyInfo(identifier, new BitArray(0, key)));
  }
  function hashAlgorithmName(_this__u8e3s4) {
    var tmp;
    if (equals(_this__u8e3s4, SHA1_getInstance())) {
      tmp = 'SHA-1';
    } else if (equals(_this__u8e3s4, SHA256_getInstance())) {
      tmp = 'SHA-256';
    } else if (equals(_this__u8e3s4, SHA384_getInstance())) {
      tmp = 'SHA-384';
    } else if (equals(_this__u8e3s4, SHA512_getInstance())) {
      tmp = 'SHA-512';
    } else {
      throw IllegalStateException_init_$Create$('Unsupported hash algorithm: ' + _this__u8e3s4.d1a_1);
    }
    return tmp;
  }
  function digestSize(_this__u8e3s4) {
    var tmp;
    if (equals(_this__u8e3s4, SHA1_getInstance())) {
      tmp = 20;
    } else if (equals(_this__u8e3s4, SHA256_getInstance())) {
      tmp = 32;
    } else if (equals(_this__u8e3s4, SHA384_getInstance())) {
      tmp = 48;
    } else if (equals(_this__u8e3s4, SHA512_getInstance())) {
      tmp = 64;
    } else {
      throw IllegalStateException_init_$Create$('Unsupported hash algorithm: ' + _this__u8e3s4.d1a_1);
    }
    return tmp;
  }
  function nonBlocking() {
    throw IllegalStateException_init_$Create$('Only non-blocking(suspend) calls are supported in WebCrypto');
  }
  function WebCryptoEncodableKey(key, keyProcessor) {
    this.y1i_1 = key;
    this.z1i_1 = keyProcessor;
  }
  function $decodeFromByteArrayCOROUTINE$11(_this__u8e3s4, format, bytes, resultContinuation) {
    CoroutineImpl.call(this, resultContinuation);
    this.i1j_1 = _this__u8e3s4;
    this.j1j_1 = format;
    this.k1j_1 = bytes;
  }
  protoOf($decodeFromByteArrayCOROUTINE$11).ba = function () {
    var suspendResult = this.l9_1;
    $sm: do
      try {
        var tmp = this.j9_1;
        switch (tmp) {
          case 0:
            this.k9_1 = 2;
            this.l1j_1 = this.i1j_1.o1j_1.f1g_1;
            this.j9_1 = 1;
            suspendResult = WebCrypto_getInstance().r1h(this.i1j_1.n1j_1.k1f(this.j1j_1), this.i1j_1.n1j_1.m1f(this.i1j_1.m1j_1, this.j1j_1, this.k1j_1), this.i1j_1.m1j_1, true, this.i1j_1.o1j_1.e1g_1, this);
            if (suspendResult === get_COROUTINE_SUSPENDED()) {
              return suspendResult;
            }

            continue $sm;
          case 1:
            var ARGUMENT = suspendResult;
            return this.l1j_1(ARGUMENT);
          case 2:
            throw this.m9_1;
        }
      } catch ($p) {
        var e = $p;
        if (this.k9_1 === 2) {
          throw e;
        } else {
          this.j9_1 = this.k9_1;
          this.m9_1 = e;
        }
      }
     while (true);
  };
  function WebCryptoKeyDecoder(algorithm, keyProcessor, keyWrapper) {
    this.m1j_1 = algorithm;
    this.n1j_1 = keyProcessor;
    this.o1j_1 = keyWrapper;
  }
  protoOf(WebCryptoKeyDecoder).t1a = function (format, bytes, $completion) {
    var tmp = new $decodeFromByteArrayCOROUTINE$11(this, format, bytes, $completion);
    tmp.l9_1 = Unit_instance;
    tmp.m9_1 = null;
    return tmp.ba();
  };
  protoOf(WebCryptoKeyDecoder).u1a = function (format, bytes) {
    nonBlocking();
  };
  function WebCryptoKeyProcessor() {
  }
  function WebCryptoKeyWrapper(usages, wrap) {
    this.e1g_1 = usages;
    this.f1g_1 = wrap;
  }
  function WebCryptoSignatureGenerator(algorithm, key) {
    this.p1j_1 = algorithm;
    this.q1j_1 = key;
  }
  protoOf(WebCryptoSignatureGenerator).a1b = function (data, $completion) {
    return WebCrypto_getInstance().r1j(this.p1j_1, this.q1j_1, data, $completion);
  };
  protoOf(WebCryptoSignatureGenerator).z1a = function () {
    nonBlocking();
  };
  protoOf(WebCryptoSignatureGenerator).b1b = function (data) {
    nonBlocking();
  };
  function get_initHook() {
    return initHook;
  }
  var initHook;
  function initHook$init$() {
    Registry_getInstance().g1a(get_defaultProvider());
    return Unit_instance;
  }
  function AesGcmCipherAlgorithm(additionalData, iv, tagLength) {
    // Inline function 'kotlin.js.unsafeCast' call
    return additionalData == null ? {name: 'AES-GCM', iv: iv, tagLength: tagLength} : {name: 'AES-GCM', iv: iv, tagLength: tagLength, additionalData: additionalData};
  }
  function EcdhKeyDeriveAlgorithm(publicKey) {
    // Inline function 'kotlin.js.unsafeCast' call
    return {name: 'ECDH', public: publicKey};
  }
  function get_ecKeyAlgorithmNamedCurve(_this__u8e3s4) {
    return ecKeyAlgorithmNamedCurve(_this__u8e3s4);
  }
  function HkdfDeriveAlgorithm(hash, salt, info) {
    // Inline function 'kotlin.js.unsafeCast' call
    return {name: 'HKDF', hash: hash, salt: salt, info: info};
  }
  function Algorithm(name) {
    // Inline function 'kotlin.js.unsafeCast' call
    return {name: name};
  }
  function HmacKeyAlgorithm(hash, length) {
    // Inline function 'kotlin.js.unsafeCast' call
    return length == null ? {name: 'HMAC', hash: hash} : {name: 'HMAC', hash: hash, length: length};
  }
  function ecKeyAlgorithmNamedCurve(algorithm) {
    // Inline function 'kotlin.js.unsafeCast' call
    return algorithm.namedCurve;
  }
  function EcKeyAlgorithm(name, namedCurve) {
    // Inline function 'kotlin.js.unsafeCast' call
    return {name: name, namedCurve: namedCurve};
  }
  function get_algorithm(_this__u8e3s4) {
    return keyAlgorithm(_this__u8e3s4);
  }
  function keyAlgorithm(key) {
    // Inline function 'kotlin.js.unsafeCast' call
    return key.algorithm;
  }
  function getSubtleCrypto() {
    // Inline function 'kotlin.js.unsafeCast' call
    return function () {
      var isNodeJs = typeof process !== 'undefined' && process.versions != null && process.versions.node != null;
      if (isNodeJs) {
        return eval('require')('node:crypto').webcrypto.subtle;
      } else {
        return (window ? window.crypto ? window.crypto : window.msCrypto : self.crypto).subtle;
      }
    }();
  }
  function $encryptCOROUTINE$16(_this__u8e3s4, algorithm, key, data, resultContinuation) {
    CoroutineImpl.call(this, resultContinuation);
    this.a1k_1 = _this__u8e3s4;
    this.b1k_1 = algorithm;
    this.c1k_1 = key;
    this.d1k_1 = data;
  }
  protoOf($encryptCOROUTINE$16).ba = function () {
    var suspendResult = this.l9_1;
    $sm: do
      try {
        var tmp = this.j9_1;
        switch (tmp) {
          case 0:
            this.k9_1 = 2;
            this.j9_1 = 1;
            suspendResult = await_0(this.a1k_1.t1f_1.encrypt(this.b1k_1, this.c1k_1, this.d1k_1), this);
            if (suspendResult === get_COROUTINE_SUSPENDED()) {
              return suspendResult;
            }

            continue $sm;
          case 1:
            var ARGUMENT = suspendResult;
            return toByteArray(ARGUMENT);
          case 2:
            throw this.m9_1;
        }
      } catch ($p) {
        var e = $p;
        if (this.k9_1 === 2) {
          throw e;
        } else {
          this.j9_1 = this.k9_1;
          this.m9_1 = e;
        }
      }
     while (true);
  };
  function $decryptCOROUTINE$17(_this__u8e3s4, algorithm, key, data, resultContinuation) {
    CoroutineImpl.call(this, resultContinuation);
    this.m1k_1 = _this__u8e3s4;
    this.n1k_1 = algorithm;
    this.o1k_1 = key;
    this.p1k_1 = data;
  }
  protoOf($decryptCOROUTINE$17).ba = function () {
    var suspendResult = this.l9_1;
    $sm: do
      try {
        var tmp = this.j9_1;
        switch (tmp) {
          case 0:
            this.k9_1 = 2;
            this.j9_1 = 1;
            suspendResult = await_0(this.m1k_1.t1f_1.decrypt(this.n1k_1, this.o1k_1, this.p1k_1), this);
            if (suspendResult === get_COROUTINE_SUSPENDED()) {
              return suspendResult;
            }

            continue $sm;
          case 1:
            var ARGUMENT = suspendResult;
            return toByteArray(ARGUMENT);
          case 2:
            throw this.m9_1;
        }
      } catch ($p) {
        var e = $p;
        if (this.k9_1 === 2) {
          throw e;
        } else {
          this.j9_1 = this.k9_1;
          this.m9_1 = e;
        }
      }
     while (true);
  };
  function $signCOROUTINE$18(_this__u8e3s4, algorithm, key, data, resultContinuation) {
    CoroutineImpl.call(this, resultContinuation);
    this.y1k_1 = _this__u8e3s4;
    this.z1k_1 = algorithm;
    this.a1l_1 = key;
    this.b1l_1 = data;
  }
  protoOf($signCOROUTINE$18).ba = function () {
    var suspendResult = this.l9_1;
    $sm: do
      try {
        var tmp = this.j9_1;
        switch (tmp) {
          case 0:
            this.k9_1 = 2;
            this.j9_1 = 1;
            suspendResult = await_0(this.y1k_1.t1f_1.sign(this.z1k_1, this.a1l_1, this.b1l_1), this);
            if (suspendResult === get_COROUTINE_SUSPENDED()) {
              return suspendResult;
            }

            continue $sm;
          case 1:
            var ARGUMENT = suspendResult;
            return toByteArray(ARGUMENT);
          case 2:
            throw this.m9_1;
        }
      } catch ($p) {
        var e = $p;
        if (this.k9_1 === 2) {
          throw e;
        } else {
          this.j9_1 = this.k9_1;
          this.m9_1 = e;
        }
      }
     while (true);
  };
  function $deriveBitsCOROUTINE$19(_this__u8e3s4, algorithm, baseKey, length, resultContinuation) {
    CoroutineImpl.call(this, resultContinuation);
    this.k1l_1 = _this__u8e3s4;
    this.l1l_1 = algorithm;
    this.m1l_1 = baseKey;
    this.n1l_1 = length;
  }
  protoOf($deriveBitsCOROUTINE$19).ba = function () {
    var suspendResult = this.l9_1;
    $sm: do
      try {
        var tmp = this.j9_1;
        switch (tmp) {
          case 0:
            this.k9_1 = 2;
            this.j9_1 = 1;
            suspendResult = await_0(this.k1l_1.t1f_1.deriveBits(this.l1l_1, this.m1l_1, this.n1l_1), this);
            if (suspendResult === get_COROUTINE_SUSPENDED()) {
              return suspendResult;
            }

            continue $sm;
          case 1:
            var ARGUMENT = suspendResult;
            return toByteArray(ARGUMENT);
          case 2:
            throw this.m9_1;
        }
      } catch ($p) {
        var e = $p;
        if (this.k9_1 === 2) {
          throw e;
        } else {
          this.j9_1 = this.k9_1;
          this.m9_1 = e;
        }
      }
     while (true);
  };
  function WebCrypto() {
    WebCrypto_instance = this;
    this.t1f_1 = getSubtleCrypto();
  }
  protoOf(WebCrypto).u1f = function (algorithm, key, data, $completion) {
    var tmp = new $encryptCOROUTINE$16(this, algorithm, key, data, $completion);
    tmp.l9_1 = Unit_instance;
    tmp.m9_1 = null;
    return tmp.ba();
  };
  protoOf(WebCrypto).v1f = function (algorithm, key, data, $completion) {
    var tmp = new $decryptCOROUTINE$17(this, algorithm, key, data, $completion);
    tmp.l9_1 = Unit_instance;
    tmp.m9_1 = null;
    return tmp.ba();
  };
  protoOf(WebCrypto).r1j = function (algorithm, key, data, $completion) {
    var tmp = new $signCOROUTINE$18(this, algorithm, key, data, $completion);
    tmp.l9_1 = Unit_instance;
    tmp.m9_1 = null;
    return tmp.ba();
  };
  protoOf(WebCrypto).x1g = function (algorithm, baseKey, length, $completion) {
    var tmp = new $deriveBitsCOROUTINE$19(this, algorithm, baseKey, length, $completion);
    tmp.l9_1 = Unit_instance;
    tmp.m9_1 = null;
    return tmp.ba();
  };
  protoOf(WebCrypto).r1h = function (format, keyData, algorithm, extractable, keyUsages, $completion) {
    var key = format === 'jwk' ? JSON.parse(decodeToString(keyData)) : keyData;
    return await_0(this.t1f_1.importKey(format, key, algorithm, extractable, keyUsages), $completion);
  };
  var WebCrypto_instance;
  function WebCrypto_getInstance() {
    if (WebCrypto_instance == null)
      new WebCrypto();
    return WebCrypto_instance;
  }
  function toByteArray(_this__u8e3s4) {
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    return new Int8Array(_this__u8e3s4);
  }
  function await_0(_this__u8e3s4, $completion) {
    var tmp = new $awaitCOROUTINE$21(_this__u8e3s4, $completion);
    tmp.l9_1 = Unit_instance;
    tmp.m9_1 = null;
    return tmp.ba();
  }
  function await$lambda($continuation) {
    return function (it) {
      // Inline function 'kotlin.coroutines.resume' call
      var this_0 = $continuation;
      // Inline function 'kotlin.Companion.success' call
      var tmp$ret$0 = _Result___init__impl__xyqfz8(it);
      this_0.ea(tmp$ret$0);
      return null;
    };
  }
  function await$lambda_0($continuation) {
    return function (it) {
      // Inline function 'kotlin.coroutines.resumeWithException' call
      var this_0 = $continuation;
      // Inline function 'kotlin.Companion.failure' call
      var tmp$ret$0 = _Result___init__impl__xyqfz8(createFailure(it));
      this_0.ea(tmp$ret$0);
      return null;
    };
  }
  function $awaitCOROUTINE$21(_this__u8e3s4, resultContinuation) {
    CoroutineImpl.call(this, resultContinuation);
    this.w1l_1 = _this__u8e3s4;
  }
  protoOf($awaitCOROUTINE$21).ba = function () {
    var suspendResult = this.l9_1;
    $sm: do
      try {
        var tmp = this.j9_1;
        switch (tmp) {
          case 0:
            this.k9_1 = 2;
            this.j9_1 = 1;
            var safe = SafeContinuation_init_$Create$(intercepted(this));
            var tmp_0 = await$lambda(safe);
            this.w1l_1.then(tmp_0, await$lambda_0(safe));
            suspendResult = returnIfSuspended(safe.na(), this);
            if (suspendResult === get_COROUTINE_SUSPENDED()) {
              return suspendResult;
            }

            continue $sm;
          case 1:
            return suspendResult;
          case 2:
            throw this.m9_1;
        }
      } catch ($p) {
        var e = $p;
        if (this.k9_1 === 2) {
          throw e;
        } else {
          this.j9_1 = this.k9_1;
          this.m9_1 = e;
        }
      }
     while (true);
  };
  //region block: init
  WebCryptoHkdf_instance = new WebCryptoHkdf();
  WebCryptoPbkdf2_instance = new WebCryptoPbkdf2();
  initHook = initHook$init$();
  //endregion
  //region block: exports
  function $jsExportAll$(_) {
    var $dev = _.dev || (_.dev = {});
    var $dev$whyoleg = $dev.whyoleg || ($dev.whyoleg = {});
    var $dev$whyoleg$cryptography = $dev$whyoleg.cryptography || ($dev$whyoleg.cryptography = {});
    var $dev$whyoleg$cryptography$providers = $dev$whyoleg$cryptography.providers || ($dev$whyoleg$cryptography.providers = {});
    var $dev$whyoleg$cryptography$providers$webcrypto = $dev$whyoleg$cryptography$providers.webcrypto || ($dev$whyoleg$cryptography$providers.webcrypto = {});
    defineProp($dev$whyoleg$cryptography$providers$webcrypto, 'initHook', get_initHook);
  }
  $jsExportAll$(_);
  _.$jsExportAll$ = $jsExportAll$;
  //endregion
  return _;
}));
