// @sphereon/kmp-mdoc-core/cryptography-kotlin-cryptography-serialization-asn1-modules.js
(function (factory) {
  if (typeof define === 'function' && define.amd)
    define(['exports', './kotlin-kotlin-stdlib.js', './cryptography-kotlin-cryptography-serialization-asn1.js', './kotlinx-serialization-kotlinx-serialization-core.js'], factory);
  else if (typeof exports === 'object')
    factory(module.exports, require('./kotlin-kotlin-stdlib.js'), require('./cryptography-kotlin-cryptography-serialization-asn1.js'), require('./kotlinx-serialization-kotlinx-serialization-core.js'));
  else {
    if (typeof globalThis['kotlin-kotlin-stdlib'] === 'undefined') {
      throw new Error("Error loading module 'cryptography-kotlin-cryptography-serialization-asn1-modules'. Its dependency 'kotlin-kotlin-stdlib' was not found. Please, check whether 'kotlin-kotlin-stdlib' is loaded prior to 'cryptography-kotlin-cryptography-serialization-asn1-modules'.");
    }
    if (typeof globalThis['cryptography-kotlin-cryptography-serialization-asn1'] === 'undefined') {
      throw new Error("Error loading module 'cryptography-kotlin-cryptography-serialization-asn1-modules'. Its dependency 'cryptography-kotlin-cryptography-serialization-asn1' was not found. Please, check whether 'cryptography-kotlin-cryptography-serialization-asn1' is loaded prior to 'cryptography-kotlin-cryptography-serialization-asn1-modules'.");
    }
    if (typeof globalThis['kotlinx-serialization-kotlinx-serialization-core'] === 'undefined') {
      throw new Error("Error loading module 'cryptography-kotlin-cryptography-serialization-asn1-modules'. Its dependency 'kotlinx-serialization-kotlinx-serialization-core' was not found. Please, check whether 'kotlinx-serialization-kotlinx-serialization-core' is loaded prior to 'cryptography-kotlin-cryptography-serialization-asn1-modules'.");
    }
    globalThis['cryptography-kotlin-cryptography-serialization-asn1-modules'] = factory(typeof globalThis['cryptography-kotlin-cryptography-serialization-asn1-modules'] === 'undefined' ? {} : globalThis['cryptography-kotlin-cryptography-serialization-asn1-modules'], globalThis['kotlin-kotlin-stdlib'], globalThis['cryptography-kotlin-cryptography-serialization-asn1'], globalThis['kotlinx-serialization-kotlinx-serialization-core']);
  }
}(function (_, kotlin_kotlin, kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1, kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core) {
  'use strict';
  //region block: imports
  var protoOf = kotlin_kotlin.$_$.xa;
  var initMetadataForInterface = kotlin_kotlin.$_$.z9;
  var Companion_instance = kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1.$_$.k;
  var CONTEXTUAL_getInstance = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.k;
  var buildSerialDescriptor = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.n1;
  var Unit_instance = kotlin_kotlin.$_$.q4;
  var OPEN_getInstance = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.f;
  var ObjectIdentifier = kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1.$_$.c;
  var THROW_CCE = kotlin_kotlin.$_$.ge;
  var isInterface = kotlin_kotlin.$_$.ka;
  var toString = kotlin_kotlin.$_$.bb;
  var IllegalStateException_init_$Create$ = kotlin_kotlin.$_$.q1;
  var KSerializer = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.r2;
  var initMetadataForClass = kotlin_kotlin.$_$.w9;
  var VOID = kotlin_kotlin.$_$.f;
  var _ObjectIdentifier___init__impl__7hutfr = kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1.$_$.e;
  var initMetadataForCompanion = kotlin_kotlin.$_$.x9;
  var InlineClassDescriptor = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.z1;
  var $serializer_getInstance = kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1.$_$.j;
  var typeParametersSerializers = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.x1;
  var GeneratedSerializer = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.y1;
  var initMetadataForObject = kotlin_kotlin.$_$.ba;
  var ObjectIdentifier__toString_impl_dtwsrd = kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1.$_$.g;
  var ObjectIdentifier__hashCode_impl_7x941i = kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1.$_$.f;
  var PluginGeneratedSerialDescriptor = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.e2;
  var TagType_EXPLICIT_getInstance = kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1.$_$.d;
  var ContextSpecificTag = kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1.$_$.b;
  var ByteArraySerializer_getInstance = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.r;
  var $serializer_getInstance_0 = kotlin_dev_whyoleg_cryptography_cryptography_serialization_asn1.$_$.h;
  var UnknownFieldException_init_$Create$ = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.e;
  var IntSerializer_getInstance = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.t;
  var get_nullable = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.z;
  var throwMissingFieldException = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.j2;
  var objectCreate = kotlin_kotlin.$_$.wa;
  var NothingSerializer = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.y;
  var getKClass = kotlin_kotlin.$_$.e;
  var ContextualSerializer = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.p2;
  //endregion
  //region block: pre-declaration
  initMetadataForInterface(AlgorithmIdentifier, 'AlgorithmIdentifier');
  initMetadataForClass(AlgorithmIdentifierSerializer, 'AlgorithmIdentifierSerializer', VOID, VOID, [KSerializer]);
  initMetadataForInterface(KeyAlgorithmIdentifier, 'KeyAlgorithmIdentifier', VOID, VOID, [AlgorithmIdentifier], VOID, VOID, {0: KeyAlgorithmIdentifierSerializer_getInstance});
  initMetadataForClass(EcKeyAlgorithmIdentifier, 'EcKeyAlgorithmIdentifier', VOID, VOID, [KeyAlgorithmIdentifier]);
  initMetadataForCompanion(Companion);
  initMetadataForObject($serializer, '$serializer', VOID, VOID, [GeneratedSerializer]);
  initMetadataForClass(EcParameters, 'EcParameters', VOID, VOID, VOID, VOID, VOID, {0: $serializer_getInstance_1});
  initMetadataForCompanion(Companion_0);
  initMetadataForObject($serializer_0, '$serializer', VOID, VOID, [GeneratedSerializer]);
  initMetadataForClass(EcPrivateKey, 'EcPrivateKey', VOID, VOID, VOID, VOID, VOID, {0: $serializer_getInstance_2});
  initMetadataForClass(UnknownKeyAlgorithmIdentifier, 'UnknownKeyAlgorithmIdentifier', VOID, VOID, [KeyAlgorithmIdentifier]);
  initMetadataForObject(KeyAlgorithmIdentifierSerializer, 'KeyAlgorithmIdentifierSerializer', VOID, AlgorithmIdentifierSerializer);
  initMetadataForCompanion(Companion_1);
  initMetadataForObject($serializer_1, '$serializer', VOID, VOID, [GeneratedSerializer]);
  initMetadataForClass(PrivateKeyInfo, 'PrivateKeyInfo', VOID, VOID, VOID, VOID, VOID, {0: $serializer_getInstance_3});
  initMetadataForCompanion(Companion_2);
  initMetadataForObject($serializer_2, '$serializer', VOID, VOID, [GeneratedSerializer]);
  initMetadataForClass(SubjectPublicKeyInfo, 'SubjectPublicKeyInfo', VOID, VOID, VOID, VOID, VOID, {0: $serializer_getInstance_4});
  initMetadataForObject(RsaKeyAlgorithmIdentifier, 'RsaKeyAlgorithmIdentifier', VOID, VOID, [KeyAlgorithmIdentifier]);
  //endregion
  function AlgorithmIdentifier() {
  }
  function AlgorithmIdentifierSerializer$descriptor$lambda($this$buildSerialDescriptor) {
    $this$buildSerialDescriptor.eo('algorithm', Companion_instance.k19().ln());
    $this$buildSerialDescriptor.eo('parameters', buildSerialDescriptor('Any', CONTEXTUAL_getInstance(), []));
    return Unit_instance;
  }
  function AlgorithmIdentifierSerializer() {
    var tmp = this;
    var tmp_0 = OPEN_getInstance();
    tmp.t1d_1 = buildSerialDescriptor('AlgorithmIdentifier', tmp_0, [], AlgorithmIdentifierSerializer$descriptor$lambda);
  }
  protoOf(AlgorithmIdentifierSerializer).w1d = function (_this__u8e3s4, serializer, value) {
    _this__u8e3s4.ns(this.t1d_1, 1, serializer, value);
  };
  protoOf(AlgorithmIdentifierSerializer).x1d = function (_this__u8e3s4, serializer) {
    return _this__u8e3s4.kr(this.t1d_1, 1, serializer);
  };
  protoOf(AlgorithmIdentifierSerializer).ln = function () {
    return this.t1d_1;
  };
  protoOf(AlgorithmIdentifierSerializer).y1d = function (encoder, value) {
    var descriptor = this.t1d_1;
    var composite = encoder.vq(descriptor);
    // Inline function 'dev.whyoleg.cryptography.serialization.asn1.modules.AlgorithmIdentifierSerializer.serialize.<anonymous>' call
    composite.ms(this.t1d_1, 0, Companion_instance.k19(), new ObjectIdentifier(value.s1d()));
    this.u1d(composite, value);
    composite.wq(descriptor);
    return Unit_instance;
  };
  protoOf(AlgorithmIdentifierSerializer).yn = function (encoder, value) {
    return this.y1d(encoder, (!(value == null) ? isInterface(value, AlgorithmIdentifier) : false) ? value : THROW_CCE());
  };
  protoOf(AlgorithmIdentifierSerializer).zn = function (decoder) {
    // Inline function 'kotlinx.serialization.encoding.decodeStructure' call
    var descriptor = this.t1d_1;
    var composite = decoder.vq(descriptor);
    // Inline function 'dev.whyoleg.cryptography.serialization.asn1.modules.AlgorithmIdentifierSerializer.deserialize.<anonymous>' call
    // Inline function 'kotlin.check' call
    // Inline function 'kotlin.contracts.contract' call
    // Inline function 'kotlin.check' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(composite.mr(this.t1d_1) === 0)) {
      // Inline function 'kotlin.check.<anonymous>' call
      var message = 'Check failed.';
      throw IllegalStateException_init_$Create$(toString(message));
    }
    var algorithm = composite.ir(this.t1d_1, 0, Companion_instance.k19()).n1c_1;
    // Inline function 'kotlin.check' call
    // Inline function 'kotlin.contracts.contract' call
    // Inline function 'kotlin.check' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(composite.mr(this.t1d_1) === 1)) {
      // Inline function 'kotlin.check.<anonymous>' call
      var message_0 = 'Check failed.';
      throw IllegalStateException_init_$Create$(toString(message_0));
    }
    var parameters = this.v1d(composite, algorithm);
    // Inline function 'kotlin.check' call
    // Inline function 'kotlin.contracts.contract' call
    // Inline function 'kotlin.check' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(composite.mr(this.t1d_1) === -1)) {
      // Inline function 'kotlin.check.<anonymous>' call
      var message_1 = 'Check failed.';
      throw IllegalStateException_init_$Create$(toString(message_1));
    }
    var result = parameters;
    composite.wq(descriptor);
    return result;
  };
  function get_secp256r1(_this__u8e3s4) {
    return _ObjectIdentifier___init__impl__7hutfr('1.2.840.10045.3.1.7');
  }
  function get_secp384r1(_this__u8e3s4) {
    return _ObjectIdentifier___init__impl__7hutfr('1.3.132.0.34');
  }
  function get_secp521r1(_this__u8e3s4) {
    return _ObjectIdentifier___init__impl__7hutfr('1.3.132.0.35');
  }
  function EcKeyAlgorithmIdentifier(parameters) {
    this.z1d_1 = parameters;
  }
  protoOf(EcKeyAlgorithmIdentifier).s1d = function () {
    return get_EC(Companion_instance);
  };
  function _EcParameters___init__impl__i4ww9z(namedCurve) {
    return namedCurve;
  }
  function _EcParameters___get_namedCurve__impl__oibidm($this) {
    return $this;
  }
  function Companion() {
  }
  protoOf(Companion).k19 = function () {
    return $serializer_getInstance_1();
  };
  var Companion_instance_0;
  function Companion_getInstance() {
    return Companion_instance_0;
  }
  function $serializer() {
    $serializer_instance = this;
    var tmp0_serialDesc = new InlineClassDescriptor('dev.whyoleg.cryptography.serialization.asn1.modules.EcParameters', this);
    tmp0_serialDesc.iy('namedCurve', false);
    this.a1e_1 = tmp0_serialDesc;
  }
  protoOf($serializer).b1e = function (encoder, value) {
    var tmp0_inlineEncoder = encoder.bs(this.a1e_1);
    if (tmp0_inlineEncoder == null)
      null;
    else {
      tmp0_inlineEncoder.xn($serializer_getInstance(), new ObjectIdentifier(_EcParameters___get_namedCurve__impl__oibidm(value)));
    }
  };
  protoOf($serializer).yn = function (encoder, value) {
    return this.b1e(encoder, value instanceof EcParameters ? value.c1e_1 : THROW_CCE());
  };
  protoOf($serializer).d1e = function (decoder) {
    return _EcParameters___init__impl__i4ww9z(decoder.tq(this.a1e_1).ao($serializer_getInstance()).n1c_1);
  };
  protoOf($serializer).zn = function (decoder) {
    return new EcParameters(this.d1e(decoder));
  };
  protoOf($serializer).ln = function () {
    return this.a1e_1;
  };
  protoOf($serializer).yy = function () {
    // Inline function 'kotlin.arrayOf' call
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    return [$serializer_getInstance()];
  };
  var $serializer_instance;
  function $serializer_getInstance_1() {
    if ($serializer_instance == null)
      new $serializer();
    return $serializer_instance;
  }
  function EcParameters__toString_impl_mgxjuh($this) {
    return 'EcParameters(namedCurve=' + ObjectIdentifier__toString_impl_dtwsrd($this) + ')';
  }
  function EcParameters__hashCode_impl_prn1m($this) {
    return ObjectIdentifier__hashCode_impl_7x941i($this);
  }
  function EcParameters__equals_impl_5tm4e2($this, other) {
    if (!(other instanceof EcParameters))
      return false;
    if (!($this === (other instanceof EcParameters ? other.c1e_1 : THROW_CCE())))
      return false;
    return true;
  }
  function EcParameters(namedCurve) {
    this.c1e_1 = namedCurve;
  }
  protoOf(EcParameters).toString = function () {
    return EcParameters__toString_impl_mgxjuh(this.c1e_1);
  };
  protoOf(EcParameters).hashCode = function () {
    return EcParameters__hashCode_impl_prn1m(this.c1e_1);
  };
  protoOf(EcParameters).equals = function (other) {
    return EcParameters__equals_impl_5tm4e2(this.c1e_1, other);
  };
  function Companion_0() {
  }
  protoOf(Companion_0).k19 = function () {
    return $serializer_getInstance_2();
  };
  var Companion_instance_1;
  function Companion_getInstance_0() {
    return Companion_instance_1;
  }
  function $serializer_0() {
    $serializer_instance_0 = this;
    var tmp0_serialDesc = new PluginGeneratedSerialDescriptor('dev.whyoleg.cryptography.serialization.asn1.modules.EcPrivateKey', this, 4);
    tmp0_serialDesc.iy('version', false);
    tmp0_serialDesc.iy('privateKey', false);
    tmp0_serialDesc.iy('parameters', true);
    tmp0_serialDesc.jy(new ContextSpecificTag(0, TagType_EXPLICIT_getInstance()));
    tmp0_serialDesc.iy('publicKey', true);
    tmp0_serialDesc.jy(new ContextSpecificTag(1, TagType_EXPLICIT_getInstance()));
    this.e1e_1 = tmp0_serialDesc;
  }
  protoOf($serializer_0).f1e = function (encoder, value) {
    var tmp0_desc = this.e1e_1;
    var tmp1_output = encoder.vq(tmp0_desc);
    tmp1_output.fs(tmp0_desc, 0, value.g1e_1);
    tmp1_output.ms(tmp0_desc, 1, ByteArraySerializer_getInstance(), value.h1e_1);
    var tmp;
    if (tmp1_output.rs(tmp0_desc, 2)) {
      tmp = true;
    } else {
      var tmp_0 = value.i1e_1;
      tmp = !((tmp_0 == null ? null : new EcParameters(tmp_0)) == null);
    }
    if (tmp) {
      var tmp_1 = $serializer_getInstance_1();
      var tmp_2 = value.i1e_1;
      tmp1_output.ns(tmp0_desc, 2, tmp_1, tmp_2 == null ? null : new EcParameters(tmp_2));
    }
    if (tmp1_output.rs(tmp0_desc, 3) ? true : !(value.j1e_1 == null)) {
      tmp1_output.ns(tmp0_desc, 3, $serializer_getInstance_0(), value.j1e_1);
    }
    tmp1_output.wq(tmp0_desc);
  };
  protoOf($serializer_0).yn = function (encoder, value) {
    return this.f1e(encoder, value instanceof EcPrivateKey ? value : THROW_CCE());
  };
  protoOf($serializer_0).zn = function (decoder) {
    var tmp0_desc = this.e1e_1;
    var tmp1_flag = true;
    var tmp2_index = 0;
    var tmp3_bitMask0 = 0;
    var tmp4_local0 = 0;
    var tmp5_local1 = null;
    var tmp6_local2 = null;
    var tmp7_local3 = null;
    var tmp8_input = decoder.vq(tmp0_desc);
    if (tmp8_input.lr()) {
      tmp4_local0 = tmp8_input.ar(tmp0_desc, 0);
      tmp3_bitMask0 = tmp3_bitMask0 | 1;
      tmp5_local1 = tmp8_input.hr(tmp0_desc, 1, ByteArraySerializer_getInstance(), tmp5_local1);
      tmp3_bitMask0 = tmp3_bitMask0 | 2;
      var tmp = $serializer_getInstance_1();
      var tmp_0 = tmp6_local2;
      var tmp_1 = tmp8_input.jr(tmp0_desc, 2, tmp, tmp_0 == null ? null : new EcParameters(tmp_0));
      tmp6_local2 = tmp_1 == null ? null : tmp_1.c1e_1;
      tmp3_bitMask0 = tmp3_bitMask0 | 4;
      tmp7_local3 = tmp8_input.jr(tmp0_desc, 3, $serializer_getInstance_0(), tmp7_local3);
      tmp3_bitMask0 = tmp3_bitMask0 | 8;
    } else
      while (tmp1_flag) {
        tmp2_index = tmp8_input.mr(tmp0_desc);
        switch (tmp2_index) {
          case -1:
            tmp1_flag = false;
            break;
          case 0:
            tmp4_local0 = tmp8_input.ar(tmp0_desc, 0);
            tmp3_bitMask0 = tmp3_bitMask0 | 1;
            break;
          case 1:
            tmp5_local1 = tmp8_input.hr(tmp0_desc, 1, ByteArraySerializer_getInstance(), tmp5_local1);
            tmp3_bitMask0 = tmp3_bitMask0 | 2;
            break;
          case 2:
            var tmp_2 = $serializer_getInstance_1();
            var tmp_3 = tmp6_local2;
            var tmp_4 = tmp8_input.jr(tmp0_desc, 2, tmp_2, tmp_3 == null ? null : new EcParameters(tmp_3));
            tmp6_local2 = tmp_4 == null ? null : tmp_4.c1e_1;
            tmp3_bitMask0 = tmp3_bitMask0 | 4;
            break;
          case 3:
            tmp7_local3 = tmp8_input.jr(tmp0_desc, 3, $serializer_getInstance_0(), tmp7_local3);
            tmp3_bitMask0 = tmp3_bitMask0 | 8;
            break;
          default:
            throw UnknownFieldException_init_$Create$(tmp2_index);
        }
      }
    tmp8_input.wq(tmp0_desc);
    return EcPrivateKey_init_$Create$(tmp3_bitMask0, tmp4_local0, tmp5_local1, tmp6_local2, tmp7_local3, null);
  };
  protoOf($serializer_0).ln = function () {
    return this.e1e_1;
  };
  protoOf($serializer_0).yy = function () {
    // Inline function 'kotlin.arrayOf' call
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    return [IntSerializer_getInstance(), ByteArraySerializer_getInstance(), get_nullable($serializer_getInstance_1()), get_nullable($serializer_getInstance_0())];
  };
  var $serializer_instance_0;
  function $serializer_getInstance_2() {
    if ($serializer_instance_0 == null)
      new $serializer_0();
    return $serializer_instance_0;
  }
  function EcPrivateKey_init_$Init$(seen0, version, privateKey, parameters, publicKey, serializationConstructorMarker, $this) {
    if (!(3 === (3 & seen0))) {
      throwMissingFieldException(seen0, 3, $serializer_getInstance_2().e1e_1);
    }
    $this.g1e_1 = version;
    $this.h1e_1 = privateKey;
    if (0 === (seen0 & 4))
      $this.i1e_1 = null;
    else
      $this.i1e_1 = parameters;
    if (0 === (seen0 & 8))
      $this.j1e_1 = null;
    else
      $this.j1e_1 = publicKey;
    return $this;
  }
  function EcPrivateKey_init_$Create$(seen0, version, privateKey, parameters, publicKey, serializationConstructorMarker) {
    return EcPrivateKey_init_$Init$(seen0, version, privateKey, parameters, publicKey, serializationConstructorMarker, objectCreate(protoOf(EcPrivateKey)));
  }
  function EcPrivateKey(version, privateKey, parameters, publicKey) {
    parameters = parameters === VOID ? null : parameters;
    publicKey = publicKey === VOID ? null : publicKey;
    this.g1e_1 = version;
    this.h1e_1 = privateKey;
    this.i1e_1 = parameters;
    this.j1e_1 = publicKey;
  }
  function get_EC(_this__u8e3s4) {
    return _ObjectIdentifier___init__impl__7hutfr('1.2.840.10045.2.1');
  }
  function KeyAlgorithmIdentifier() {
  }
  function UnknownKeyAlgorithmIdentifier(algorithm) {
    this.k1e_1 = algorithm;
  }
  protoOf(UnknownKeyAlgorithmIdentifier).s1d = function () {
    return this.k1e_1;
  };
  protoOf(UnknownKeyAlgorithmIdentifier).l1e = function () {
    return null;
  };
  function KeyAlgorithmIdentifierSerializer() {
    KeyAlgorithmIdentifierSerializer_instance = this;
    AlgorithmIdentifierSerializer.call(this);
  }
  protoOf(KeyAlgorithmIdentifierSerializer).n1e = function (_this__u8e3s4, value) {
    var tmp;
    if (value instanceof RsaKeyAlgorithmIdentifier) {
      this.w1d(_this__u8e3s4, NothingSerializer(), RsaKeyAlgorithmIdentifier_instance.l1e());
      tmp = Unit_instance;
    } else {
      if (value instanceof EcKeyAlgorithmIdentifier) {
        var tmp_0 = Companion_instance_0.k19();
        var tmp_1 = value.z1d_1;
        this.w1d(_this__u8e3s4, tmp_0, tmp_1 == null ? null : new EcParameters(tmp_1));
        tmp = Unit_instance;
      } else {
        if (value instanceof UnknownKeyAlgorithmIdentifier) {
          this.w1d(_this__u8e3s4, NothingSerializer(), value.l1e());
          tmp = Unit_instance;
        } else {
          this.w1d(_this__u8e3s4, NothingSerializer(), null);
          tmp = Unit_instance;
        }
      }
    }
    return tmp;
  };
  protoOf(KeyAlgorithmIdentifierSerializer).u1d = function (_this__u8e3s4, value) {
    return this.n1e(_this__u8e3s4, isInterface(value, KeyAlgorithmIdentifier) ? value : THROW_CCE());
  };
  protoOf(KeyAlgorithmIdentifierSerializer).v1d = function (_this__u8e3s4, algorithm) {
    var tmp;
    if (algorithm === get_RSA(Companion_instance)) {
      this.x1d(_this__u8e3s4, NothingSerializer());
      tmp = RsaKeyAlgorithmIdentifier_instance;
    } else if (algorithm === get_EC(Companion_instance)) {
      var tmp_0 = this.x1d(_this__u8e3s4, Companion_instance_0.k19());
      tmp = new EcKeyAlgorithmIdentifier(tmp_0 == null ? null : tmp_0.c1e_1);
    } else {
      tmp = new UnknownKeyAlgorithmIdentifier(algorithm);
    }
    return tmp;
  };
  var KeyAlgorithmIdentifierSerializer_instance;
  function KeyAlgorithmIdentifierSerializer_getInstance() {
    if (KeyAlgorithmIdentifierSerializer_instance == null)
      new KeyAlgorithmIdentifierSerializer();
    return KeyAlgorithmIdentifierSerializer_instance;
  }
  function Companion_1() {
    Companion_instance_2 = this;
    var tmp = this;
    // Inline function 'kotlin.arrayOf' call
    var tmp_0 = getKClass(KeyAlgorithmIdentifier);
    var tmp_1 = KeyAlgorithmIdentifierSerializer_getInstance();
    // Inline function 'kotlin.arrayOf' call
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    var tmp$ret$2 = [];
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    tmp.o1e_1 = [null, new ContextualSerializer(tmp_0, tmp_1, tmp$ret$2), null];
  }
  protoOf(Companion_1).k19 = function () {
    return $serializer_getInstance_3();
  };
  var Companion_instance_2;
  function Companion_getInstance_1() {
    if (Companion_instance_2 == null)
      new Companion_1();
    return Companion_instance_2;
  }
  function $serializer_1() {
    $serializer_instance_1 = this;
    var tmp0_serialDesc = new PluginGeneratedSerialDescriptor('dev.whyoleg.cryptography.serialization.asn1.modules.PrivateKeyInfo', this, 3);
    tmp0_serialDesc.iy('version', false);
    tmp0_serialDesc.iy('privateKeyAlgorithm', false);
    tmp0_serialDesc.iy('privateKey', false);
    this.p1e_1 = tmp0_serialDesc;
  }
  protoOf($serializer_1).q1e = function (encoder, value) {
    var tmp0_desc = this.p1e_1;
    var tmp1_output = encoder.vq(tmp0_desc);
    var tmp2_cached = Companion_getInstance_1().o1e_1;
    tmp1_output.fs(tmp0_desc, 0, value.r1e_1);
    tmp1_output.ms(tmp0_desc, 1, tmp2_cached[1], value.s1e_1);
    tmp1_output.ms(tmp0_desc, 2, ByteArraySerializer_getInstance(), value.t1e_1);
    tmp1_output.wq(tmp0_desc);
  };
  protoOf($serializer_1).yn = function (encoder, value) {
    return this.q1e(encoder, value instanceof PrivateKeyInfo ? value : THROW_CCE());
  };
  protoOf($serializer_1).zn = function (decoder) {
    var tmp0_desc = this.p1e_1;
    var tmp1_flag = true;
    var tmp2_index = 0;
    var tmp3_bitMask0 = 0;
    var tmp4_local0 = 0;
    var tmp5_local1 = null;
    var tmp6_local2 = null;
    var tmp7_input = decoder.vq(tmp0_desc);
    var tmp8_cached = Companion_getInstance_1().o1e_1;
    if (tmp7_input.lr()) {
      tmp4_local0 = tmp7_input.ar(tmp0_desc, 0);
      tmp3_bitMask0 = tmp3_bitMask0 | 1;
      tmp5_local1 = tmp7_input.hr(tmp0_desc, 1, tmp8_cached[1], tmp5_local1);
      tmp3_bitMask0 = tmp3_bitMask0 | 2;
      tmp6_local2 = tmp7_input.hr(tmp0_desc, 2, ByteArraySerializer_getInstance(), tmp6_local2);
      tmp3_bitMask0 = tmp3_bitMask0 | 4;
    } else
      while (tmp1_flag) {
        tmp2_index = tmp7_input.mr(tmp0_desc);
        switch (tmp2_index) {
          case -1:
            tmp1_flag = false;
            break;
          case 0:
            tmp4_local0 = tmp7_input.ar(tmp0_desc, 0);
            tmp3_bitMask0 = tmp3_bitMask0 | 1;
            break;
          case 1:
            tmp5_local1 = tmp7_input.hr(tmp0_desc, 1, tmp8_cached[1], tmp5_local1);
            tmp3_bitMask0 = tmp3_bitMask0 | 2;
            break;
          case 2:
            tmp6_local2 = tmp7_input.hr(tmp0_desc, 2, ByteArraySerializer_getInstance(), tmp6_local2);
            tmp3_bitMask0 = tmp3_bitMask0 | 4;
            break;
          default:
            throw UnknownFieldException_init_$Create$(tmp2_index);
        }
      }
    tmp7_input.wq(tmp0_desc);
    return PrivateKeyInfo_init_$Create$(tmp3_bitMask0, tmp4_local0, tmp5_local1, tmp6_local2, null);
  };
  protoOf($serializer_1).ln = function () {
    return this.p1e_1;
  };
  protoOf($serializer_1).yy = function () {
    var tmp0_cached = Companion_getInstance_1().o1e_1;
    // Inline function 'kotlin.arrayOf' call
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    return [IntSerializer_getInstance(), tmp0_cached[1], ByteArraySerializer_getInstance()];
  };
  var $serializer_instance_1;
  function $serializer_getInstance_3() {
    if ($serializer_instance_1 == null)
      new $serializer_1();
    return $serializer_instance_1;
  }
  function PrivateKeyInfo_init_$Init$(seen0, version, privateKeyAlgorithm, privateKey, serializationConstructorMarker, $this) {
    if (!(7 === (7 & seen0))) {
      throwMissingFieldException(seen0, 7, $serializer_getInstance_3().p1e_1);
    }
    $this.r1e_1 = version;
    $this.s1e_1 = privateKeyAlgorithm;
    $this.t1e_1 = privateKey;
    return $this;
  }
  function PrivateKeyInfo_init_$Create$(seen0, version, privateKeyAlgorithm, privateKey, serializationConstructorMarker) {
    return PrivateKeyInfo_init_$Init$(seen0, version, privateKeyAlgorithm, privateKey, serializationConstructorMarker, objectCreate(protoOf(PrivateKeyInfo)));
  }
  function PrivateKeyInfo(version, privateKeyAlgorithm, privateKey) {
    Companion_getInstance_1();
    this.r1e_1 = version;
    this.s1e_1 = privateKeyAlgorithm;
    this.t1e_1 = privateKey;
  }
  function Companion_2() {
    Companion_instance_3 = this;
    var tmp = this;
    // Inline function 'kotlin.arrayOf' call
    var tmp_0 = getKClass(KeyAlgorithmIdentifier);
    var tmp_1 = KeyAlgorithmIdentifierSerializer_getInstance();
    // Inline function 'kotlin.arrayOf' call
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    var tmp$ret$2 = [];
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    tmp.u1e_1 = [new ContextualSerializer(tmp_0, tmp_1, tmp$ret$2), null];
  }
  protoOf(Companion_2).k19 = function () {
    return $serializer_getInstance_4();
  };
  var Companion_instance_3;
  function Companion_getInstance_2() {
    if (Companion_instance_3 == null)
      new Companion_2();
    return Companion_instance_3;
  }
  function $serializer_2() {
    $serializer_instance_2 = this;
    var tmp0_serialDesc = new PluginGeneratedSerialDescriptor('dev.whyoleg.cryptography.serialization.asn1.modules.SubjectPublicKeyInfo', this, 2);
    tmp0_serialDesc.iy('algorithm', false);
    tmp0_serialDesc.iy('subjectPublicKey', false);
    this.v1e_1 = tmp0_serialDesc;
  }
  protoOf($serializer_2).w1e = function (encoder, value) {
    var tmp0_desc = this.v1e_1;
    var tmp1_output = encoder.vq(tmp0_desc);
    var tmp2_cached = Companion_getInstance_2().u1e_1;
    tmp1_output.ms(tmp0_desc, 0, tmp2_cached[0], value.x1e_1);
    tmp1_output.ms(tmp0_desc, 1, $serializer_getInstance_0(), value.y1e_1);
    tmp1_output.wq(tmp0_desc);
  };
  protoOf($serializer_2).yn = function (encoder, value) {
    return this.w1e(encoder, value instanceof SubjectPublicKeyInfo ? value : THROW_CCE());
  };
  protoOf($serializer_2).zn = function (decoder) {
    var tmp0_desc = this.v1e_1;
    var tmp1_flag = true;
    var tmp2_index = 0;
    var tmp3_bitMask0 = 0;
    var tmp4_local0 = null;
    var tmp5_local1 = null;
    var tmp6_input = decoder.vq(tmp0_desc);
    var tmp7_cached = Companion_getInstance_2().u1e_1;
    if (tmp6_input.lr()) {
      tmp4_local0 = tmp6_input.hr(tmp0_desc, 0, tmp7_cached[0], tmp4_local0);
      tmp3_bitMask0 = tmp3_bitMask0 | 1;
      tmp5_local1 = tmp6_input.hr(tmp0_desc, 1, $serializer_getInstance_0(), tmp5_local1);
      tmp3_bitMask0 = tmp3_bitMask0 | 2;
    } else
      while (tmp1_flag) {
        tmp2_index = tmp6_input.mr(tmp0_desc);
        switch (tmp2_index) {
          case -1:
            tmp1_flag = false;
            break;
          case 0:
            tmp4_local0 = tmp6_input.hr(tmp0_desc, 0, tmp7_cached[0], tmp4_local0);
            tmp3_bitMask0 = tmp3_bitMask0 | 1;
            break;
          case 1:
            tmp5_local1 = tmp6_input.hr(tmp0_desc, 1, $serializer_getInstance_0(), tmp5_local1);
            tmp3_bitMask0 = tmp3_bitMask0 | 2;
            break;
          default:
            throw UnknownFieldException_init_$Create$(tmp2_index);
        }
      }
    tmp6_input.wq(tmp0_desc);
    return SubjectPublicKeyInfo_init_$Create$(tmp3_bitMask0, tmp4_local0, tmp5_local1, null);
  };
  protoOf($serializer_2).ln = function () {
    return this.v1e_1;
  };
  protoOf($serializer_2).yy = function () {
    // Inline function 'kotlin.arrayOf' call
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    return [Companion_getInstance_2().u1e_1[0], $serializer_getInstance_0()];
  };
  var $serializer_instance_2;
  function $serializer_getInstance_4() {
    if ($serializer_instance_2 == null)
      new $serializer_2();
    return $serializer_instance_2;
  }
  function SubjectPublicKeyInfo_init_$Init$(seen0, algorithm, subjectPublicKey, serializationConstructorMarker, $this) {
    if (!(3 === (3 & seen0))) {
      throwMissingFieldException(seen0, 3, $serializer_getInstance_4().v1e_1);
    }
    $this.x1e_1 = algorithm;
    $this.y1e_1 = subjectPublicKey;
    return $this;
  }
  function SubjectPublicKeyInfo_init_$Create$(seen0, algorithm, subjectPublicKey, serializationConstructorMarker) {
    return SubjectPublicKeyInfo_init_$Init$(seen0, algorithm, subjectPublicKey, serializationConstructorMarker, objectCreate(protoOf(SubjectPublicKeyInfo)));
  }
  function SubjectPublicKeyInfo(algorithm, subjectPublicKey) {
    Companion_getInstance_2();
    this.x1e_1 = algorithm;
    this.y1e_1 = subjectPublicKey;
  }
  function RsaKeyAlgorithmIdentifier() {
  }
  protoOf(RsaKeyAlgorithmIdentifier).s1d = function () {
    return get_RSA(Companion_instance);
  };
  protoOf(RsaKeyAlgorithmIdentifier).l1e = function () {
    return null;
  };
  var RsaKeyAlgorithmIdentifier_instance;
  function RsaKeyAlgorithmIdentifier_getInstance() {
    return RsaKeyAlgorithmIdentifier_instance;
  }
  function get_RSA(_this__u8e3s4) {
    return _ObjectIdentifier___init__impl__7hutfr('1.2.840.113549.1.1.1');
  }
  //region block: post-declaration
  protoOf($serializer).zy = typeParametersSerializers;
  protoOf($serializer_0).zy = typeParametersSerializers;
  protoOf($serializer_1).zy = typeParametersSerializers;
  protoOf($serializer_2).zy = typeParametersSerializers;
  //endregion
  //region block: init
  Companion_instance_0 = new Companion();
  Companion_instance_1 = new Companion_0();
  RsaKeyAlgorithmIdentifier_instance = new RsaKeyAlgorithmIdentifier();
  //endregion
  //region block: exports
  _.$_$ = _.$_$ || {};
  _.$_$.a = EcKeyAlgorithmIdentifier;
  _.$_$.b = EcParameters;
  _.$_$.c = EcPrivateKey;
  _.$_$.d = PrivateKeyInfo;
  _.$_$.e = SubjectPublicKeyInfo;
  _.$_$.f = get_secp256r1;
  _.$_$.g = get_secp384r1;
  _.$_$.h = get_secp521r1;
  _.$_$.i = _EcParameters___init__impl__i4ww9z;
  _.$_$.j = Companion_instance_1;
  _.$_$.k = Companion_getInstance_1;
  _.$_$.l = RsaKeyAlgorithmIdentifier_instance;
  _.$_$.m = Companion_getInstance_2;
  //endregion
  return _;
}));
