// @sphereon/kmp-mdoc-core/cryptography-kotlin-cryptography-serialization-asn1.js
(function (factory) {
  if (typeof define === 'function' && define.amd)
    define(['exports', './kotlin-kotlin-stdlib.js', './kotlinx-serialization-kotlinx-serialization-core.js', './cryptography-kotlin-cryptography-bigint.js'], factory);
  else if (typeof exports === 'object')
    factory(module.exports, require('./kotlin-kotlin-stdlib.js'), require('./kotlinx-serialization-kotlinx-serialization-core.js'), require('./cryptography-kotlin-cryptography-bigint.js'));
  else {
    if (typeof globalThis['kotlin-kotlin-stdlib'] === 'undefined') {
      throw new Error("Error loading module 'cryptography-kotlin-cryptography-serialization-asn1'. Its dependency 'kotlin-kotlin-stdlib' was not found. Please, check whether 'kotlin-kotlin-stdlib' is loaded prior to 'cryptography-kotlin-cryptography-serialization-asn1'.");
    }
    if (typeof globalThis['kotlinx-serialization-kotlinx-serialization-core'] === 'undefined') {
      throw new Error("Error loading module 'cryptography-kotlin-cryptography-serialization-asn1'. Its dependency 'kotlinx-serialization-kotlinx-serialization-core' was not found. Please, check whether 'kotlinx-serialization-kotlinx-serialization-core' is loaded prior to 'cryptography-kotlin-cryptography-serialization-asn1'.");
    }
    if (typeof globalThis['cryptography-kotlin-cryptography-bigint'] === 'undefined') {
      throw new Error("Error loading module 'cryptography-kotlin-cryptography-serialization-asn1'. Its dependency 'cryptography-kotlin-cryptography-bigint' was not found. Please, check whether 'cryptography-kotlin-cryptography-bigint' is loaded prior to 'cryptography-kotlin-cryptography-serialization-asn1'.");
    }
    globalThis['cryptography-kotlin-cryptography-serialization-asn1'] = factory(typeof globalThis['cryptography-kotlin-cryptography-serialization-asn1'] === 'undefined' ? {} : globalThis['cryptography-kotlin-cryptography-serialization-asn1'], globalThis['kotlin-kotlin-stdlib'], globalThis['kotlinx-serialization-kotlinx-serialization-core'], globalThis['cryptography-kotlin-cryptography-bigint']);
  }
}(function (_, kotlin_kotlin, kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core, kotlin_dev_whyoleg_cryptography_cryptography_bigint) {
  'use strict';
  //region block: imports
  var imul = Math.imul;
  var clz32 = Math.clz32;
  var Unit_instance = kotlin_kotlin.$_$.q4;
  var Enum = kotlin_kotlin.$_$.ud;
  var protoOf = kotlin_kotlin.$_$.xa;
  var initMetadataForClass = kotlin_kotlin.$_$.w9;
  var VOID = kotlin_kotlin.$_$.f;
  var THROW_CCE = kotlin_kotlin.$_$.ge;
  var getStringHashCode = kotlin_kotlin.$_$.u9;
  var initMetadataForCompanion = kotlin_kotlin.$_$.x9;
  var PluginGeneratedSerialDescriptor = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.e2;
  var ByteArraySerializer_getInstance = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.r;
  var UnknownFieldException_init_$Create$ = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.e;
  var IntSerializer_getInstance = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.t;
  var typeParametersSerializers = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.x1;
  var GeneratedSerializer = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.y1;
  var initMetadataForObject = kotlin_kotlin.$_$.ba;
  var throwMissingFieldException = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.j2;
  var last = kotlin_kotlin.$_$.c7;
  var countTrailingZeroBits = kotlin_kotlin.$_$.we;
  var toString = kotlin_kotlin.$_$.bb;
  var IllegalStateException_init_$Create$ = kotlin_kotlin.$_$.q1;
  var objectCreate = kotlin_kotlin.$_$.wa;
  var EmptySerializersModule = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.k2;
  var InlineClassDescriptor = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.z1;
  var StringSerializer_getInstance = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.u;
  var copyOfRange = kotlin_kotlin.$_$.u5;
  var takeHighestOneBit = kotlin_kotlin.$_$.if;
  var copyOf = kotlin_kotlin.$_$.z5;
  var toByte = kotlin_kotlin.$_$.ya;
  var arrayCopy = kotlin_kotlin.$_$.f5;
  var AbstractDecoder = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.p1;
  var Companion_getInstance = kotlin_dev_whyoleg_cryptography_cryptography_bigint.$_$.h;
  var equals = kotlin_kotlin.$_$.p9;
  var ByteArraySerializer = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.v;
  var PolymorphicKind = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.f1;
  var CLASS_getInstance = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.m;
  var SerializationException_init_$Create$ = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.d;
  var AbstractEncoder = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.q1;
  var toBigInt = kotlin_dev_whyoleg_cryptography_cryptography_bigint.$_$.e;
  var toBigInt_0 = kotlin_dev_whyoleg_cryptography_cryptography_bigint.$_$.g;
  var toBigInt_1 = kotlin_dev_whyoleg_cryptography_cryptography_bigint.$_$.d;
  var toBigInt_2 = kotlin_dev_whyoleg_cryptography_cryptography_bigint.$_$.f;
  var BigInt = kotlin_dev_whyoleg_cryptography_cryptography_bigint.$_$.a;
  var isByteArray = kotlin_kotlin.$_$.ea;
  var decodeToBigInt = kotlin_dev_whyoleg_cryptography_cryptography_bigint.$_$.b;
  var noWhenBranchMatchedException = kotlin_kotlin.$_$.ff;
  var StringBuilder_init_$Create$ = kotlin_kotlin.$_$.c1;
  var _Char___init__impl__6a9atx = kotlin_kotlin.$_$.c2;
  var encodeToByteArray = kotlin_dev_whyoleg_cryptography_cryptography_bigint.$_$.c;
  var split = kotlin_kotlin.$_$.mc;
  var toInt = kotlin_kotlin.$_$.ad;
  var _UByte___init__impl__g9hnc4 = kotlin_kotlin.$_$.k2;
  var toString_0 = kotlin_kotlin.$_$.ed;
  var UByte = kotlin_kotlin.$_$.ke;
  //endregion
  //region block: pre-declaration
  initMetadataForClass(TagType, 'TagType', VOID, Enum);
  initMetadataForClass(ContextSpecificTag, 'ContextSpecificTag');
  initMetadataForCompanion(Companion);
  initMetadataForObject($serializer, '$serializer', VOID, VOID, [GeneratedSerializer]);
  initMetadataForClass(BitArray, 'BitArray', VOID, VOID, VOID, VOID, VOID, {0: $serializer_getInstance});
  initMetadataForClass(Der, 'Der');
  initMetadataForObject(Default, 'Default', VOID, Der);
  initMetadataForCompanion(Companion_0);
  initMetadataForObject($serializer_0, '$serializer', VOID, VOID, [GeneratedSerializer]);
  initMetadataForClass(ObjectIdentifier, 'ObjectIdentifier', VOID, VOID, VOID, VOID, VOID, {0: $serializer_getInstance_0});
  initMetadataForClass(ByteArrayInput, 'ByteArrayInput');
  initMetadataForClass(ByteArrayOutput, 'ByteArrayOutput', ByteArrayOutput);
  initMetadataForClass(DerDecoder, 'DerDecoder', VOID, AbstractDecoder);
  initMetadataForClass(DerEncoder, 'DerEncoder', VOID, AbstractEncoder);
  initMetadataForClass(DerInput, 'DerInput');
  initMetadataForClass(DerOutput, 'DerOutput');
  //endregion
  var TagType_IMPLICIT_instance;
  var TagType_EXPLICIT_instance;
  var TagType_entriesInitialized;
  function TagType_initEntries() {
    if (TagType_entriesInitialized)
      return Unit_instance;
    TagType_entriesInitialized = true;
    TagType_IMPLICIT_instance = new TagType('IMPLICIT', 0);
    TagType_EXPLICIT_instance = new TagType('EXPLICIT', 1);
  }
  function TagType(name, ordinal) {
    Enum.call(this, name, ordinal);
  }
  function TagType_EXPLICIT_getInstance() {
    TagType_initEntries();
    return TagType_EXPLICIT_instance;
  }
  function ContextSpecificTag(classIndex, type) {
    this.r1b_1 = classIndex;
    this.s1b_1 = type;
  }
  protoOf(ContextSpecificTag).equals = function (other) {
    if (!(other instanceof ContextSpecificTag))
      return false;
    var tmp0_other_with_cast = other instanceof ContextSpecificTag ? other : THROW_CCE();
    if (!(this.r1b_1 === tmp0_other_with_cast.r1b_1))
      return false;
    if (!this.s1b_1.equals(tmp0_other_with_cast.s1b_1))
      return false;
    return true;
  };
  protoOf(ContextSpecificTag).hashCode = function () {
    var result = imul(getStringHashCode('classIndex'), 127) ^ this.r1b_1;
    result = result + (imul(getStringHashCode('type'), 127) ^ this.s1b_1.hashCode()) | 0;
    return result;
  };
  protoOf(ContextSpecificTag).toString = function () {
    return '@dev.whyoleg.cryptography.serialization.asn1.ContextSpecificTag(' + 'classIndex=' + this.r1b_1 + ', ' + 'type=' + this.s1b_1.toString() + ')';
  };
  function Companion() {
  }
  protoOf(Companion).k19 = function () {
    return $serializer_getInstance();
  };
  var Companion_instance;
  function Companion_getInstance_0() {
    return Companion_instance;
  }
  function $serializer() {
    $serializer_instance = this;
    var tmp0_serialDesc = new PluginGeneratedSerialDescriptor('dev.whyoleg.cryptography.serialization.asn1.BitArray', this, 2);
    tmp0_serialDesc.iy('unusedBits', false);
    tmp0_serialDesc.iy('byteArray', false);
    this.t1b_1 = tmp0_serialDesc;
  }
  protoOf($serializer).u1b = function (encoder, value) {
    var tmp0_desc = this.t1b_1;
    var tmp1_output = encoder.vq(tmp0_desc);
    tmp1_output.fs(tmp0_desc, 0, value.v1b_1);
    tmp1_output.ms(tmp0_desc, 1, ByteArraySerializer_getInstance(), value.w1b_1);
    tmp1_output.wq(tmp0_desc);
  };
  protoOf($serializer).yn = function (encoder, value) {
    return this.u1b(encoder, value instanceof BitArray ? value : THROW_CCE());
  };
  protoOf($serializer).zn = function (decoder) {
    var tmp0_desc = this.t1b_1;
    var tmp1_flag = true;
    var tmp2_index = 0;
    var tmp3_bitMask0 = 0;
    var tmp4_local0 = 0;
    var tmp5_local1 = null;
    var tmp6_input = decoder.vq(tmp0_desc);
    if (tmp6_input.lr()) {
      tmp4_local0 = tmp6_input.ar(tmp0_desc, 0);
      tmp3_bitMask0 = tmp3_bitMask0 | 1;
      tmp5_local1 = tmp6_input.hr(tmp0_desc, 1, ByteArraySerializer_getInstance(), tmp5_local1);
      tmp3_bitMask0 = tmp3_bitMask0 | 2;
    } else
      while (tmp1_flag) {
        tmp2_index = tmp6_input.mr(tmp0_desc);
        switch (tmp2_index) {
          case -1:
            tmp1_flag = false;
            break;
          case 0:
            tmp4_local0 = tmp6_input.ar(tmp0_desc, 0);
            tmp3_bitMask0 = tmp3_bitMask0 | 1;
            break;
          case 1:
            tmp5_local1 = tmp6_input.hr(tmp0_desc, 1, ByteArraySerializer_getInstance(), tmp5_local1);
            tmp3_bitMask0 = tmp3_bitMask0 | 2;
            break;
          default:
            throw UnknownFieldException_init_$Create$(tmp2_index);
        }
      }
    tmp6_input.wq(tmp0_desc);
    return BitArray_init_$Create$(tmp3_bitMask0, tmp4_local0, tmp5_local1, null);
  };
  protoOf($serializer).ln = function () {
    return this.t1b_1;
  };
  protoOf($serializer).yy = function () {
    // Inline function 'kotlin.arrayOf' call
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    return [IntSerializer_getInstance(), ByteArraySerializer_getInstance()];
  };
  var $serializer_instance;
  function $serializer_getInstance() {
    if ($serializer_instance == null)
      new $serializer();
    return $serializer_instance;
  }
  function BitArray_init_$Init$(seen0, unusedBits, byteArray, serializationConstructorMarker, $this) {
    if (!(3 === (3 & seen0))) {
      throwMissingFieldException(seen0, 3, $serializer_getInstance().t1b_1);
    }
    $this.v1b_1 = unusedBits;
    $this.w1b_1 = byteArray;
    // Inline function 'kotlin.collections.isEmpty' call
    if ($this.w1b_1.length === 0) {
      // Inline function 'kotlin.check' call
      // Inline function 'kotlin.contracts.contract' call
      if (!($this.v1b_1 === 0)) {
        // Inline function 'dev.whyoleg.cryptography.serialization.asn1.BitArray.<init>.<anonymous>' call
        var message = "empty array couldn't have unused bits";
        throw IllegalStateException_init_$Create$(toString(message));
      }
    } else {
      // Inline function 'kotlin.check' call
      // Inline function 'kotlin.countTrailingZeroBits' call
      var this_0 = last($this.w1b_1);
      var tmp$ret$2 = countTrailingZeroBits(this_0 | 256);
      // Inline function 'kotlin.contracts.contract' call
      if (!($this.v1b_1 <= tmp$ret$2)) {
        // Inline function 'dev.whyoleg.cryptography.serialization.asn1.BitArray.<init>.<anonymous>' call
        var message_0 = 'At least ' + $this.v1b_1 + ' last bits should be unused';
        throw IllegalStateException_init_$Create$(toString(message_0));
      }
    }
    return $this;
  }
  function BitArray_init_$Create$(seen0, unusedBits, byteArray, serializationConstructorMarker) {
    return BitArray_init_$Init$(seen0, unusedBits, byteArray, serializationConstructorMarker, objectCreate(protoOf(BitArray)));
  }
  function BitArray(unusedBits, byteArray) {
    this.v1b_1 = unusedBits;
    this.w1b_1 = byteArray;
    // Inline function 'kotlin.collections.isEmpty' call
    if (this.w1b_1.length === 0) {
      // Inline function 'kotlin.check' call
      // Inline function 'kotlin.contracts.contract' call
      if (!(this.v1b_1 === 0)) {
        // Inline function 'dev.whyoleg.cryptography.serialization.asn1.BitArray.<anonymous>' call
        var message = "empty array couldn't have unused bits";
        throw IllegalStateException_init_$Create$(toString(message));
      }
    } else {
      // Inline function 'kotlin.check' call
      // Inline function 'kotlin.countTrailingZeroBits' call
      var this_0 = last(this.w1b_1);
      var tmp$ret$2 = countTrailingZeroBits(this_0 | 256);
      // Inline function 'kotlin.contracts.contract' call
      if (!(this.v1b_1 <= tmp$ret$2)) {
        // Inline function 'dev.whyoleg.cryptography.serialization.asn1.BitArray.<anonymous>' call
        var message_0 = 'At least ' + this.v1b_1 + ' last bits should be unused';
        throw IllegalStateException_init_$Create$(toString(message_0));
      }
    }
  }
  function Default() {
    Default_instance = this;
    Der.call(this, EmptySerializersModule());
  }
  var Default_instance;
  function Default_getInstance() {
    if (Default_instance == null)
      new Default();
    return Default_instance;
  }
  function Der(serializersModule) {
    Default_getInstance();
    this.x1b_1 = serializersModule;
  }
  protoOf(Der).wn = function () {
    return this.x1b_1;
  };
  protoOf(Der).y1b = function (serializer, value) {
    var output = new ByteArrayOutput();
    (new DerEncoder(this, output)).xn(serializer, value);
    return output.g1c();
  };
  protoOf(Der).z1b = function (deserializer, bytes) {
    var input = new ByteArrayInput(bytes);
    return (new DerDecoder(this, input)).ao(deserializer);
  };
  function _ObjectIdentifier___init__impl__7hutfr(value) {
    return value;
  }
  function _ObjectIdentifier___get_value__impl__wqzjw3($this) {
    return $this;
  }
  function Companion_0() {
  }
  protoOf(Companion_0).k19 = function () {
    return $serializer_getInstance_0();
  };
  var Companion_instance_0;
  function Companion_getInstance_1() {
    return Companion_instance_0;
  }
  function $serializer_0() {
    $serializer_instance_0 = this;
    var tmp0_serialDesc = new InlineClassDescriptor('dev.whyoleg.cryptography.serialization.asn1.ObjectIdentifier', this);
    tmp0_serialDesc.iy('value', false);
    this.l1c_1 = tmp0_serialDesc;
  }
  protoOf($serializer_0).m1c = function (encoder, value) {
    var tmp0_inlineEncoder = encoder.bs(this.l1c_1);
    if (tmp0_inlineEncoder == null)
      null;
    else {
      tmp0_inlineEncoder.zr(_ObjectIdentifier___get_value__impl__wqzjw3(value));
    }
  };
  protoOf($serializer_0).yn = function (encoder, value) {
    return this.m1c(encoder, value instanceof ObjectIdentifier ? value.n1c_1 : THROW_CCE());
  };
  protoOf($serializer_0).o1c = function (decoder) {
    return _ObjectIdentifier___init__impl__7hutfr(decoder.tq(this.l1c_1).rq());
  };
  protoOf($serializer_0).zn = function (decoder) {
    return new ObjectIdentifier(this.o1c(decoder));
  };
  protoOf($serializer_0).ln = function () {
    return this.l1c_1;
  };
  protoOf($serializer_0).yy = function () {
    // Inline function 'kotlin.arrayOf' call
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    return [StringSerializer_getInstance()];
  };
  var $serializer_instance_0;
  function $serializer_getInstance_0() {
    if ($serializer_instance_0 == null)
      new $serializer_0();
    return $serializer_instance_0;
  }
  function ObjectIdentifier__toString_impl_dtwsrd($this) {
    return 'ObjectIdentifier(value=' + $this + ')';
  }
  function ObjectIdentifier__hashCode_impl_7x941i($this) {
    return getStringHashCode($this);
  }
  function ObjectIdentifier__equals_impl_u5dbkq($this, other) {
    if (!(other instanceof ObjectIdentifier))
      return false;
    if (!($this === (other instanceof ObjectIdentifier ? other.n1c_1 : THROW_CCE())))
      return false;
    return true;
  }
  function ObjectIdentifier(value) {
    this.n1c_1 = value;
  }
  protoOf(ObjectIdentifier).toString = function () {
    return ObjectIdentifier__toString_impl_dtwsrd(this.n1c_1);
  };
  protoOf(ObjectIdentifier).hashCode = function () {
    return ObjectIdentifier__hashCode_impl_7x941i(this.n1c_1);
  };
  protoOf(ObjectIdentifier).equals = function (other) {
    return ObjectIdentifier__equals_impl_u5dbkq(this.n1c_1, other);
  };
  function get_emptyArray() {
    _init_properties_ByteArrayInput_kt__imea6f();
    return emptyArray;
  }
  var emptyArray;
  function _get_available__saq0tk($this) {
    return $this.q1c_1 - $this.r1c_1 | 0;
  }
  function ensureAvailableBytes($this, count) {
    // Inline function 'kotlin.check' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(_get_available__saq0tk($this) >= count)) {
      // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.ByteArrayInput.ensureAvailableBytes.<anonymous>' call
      var message = 'Unexpected EOF, available ' + _get_available__saq0tk($this) + ' bytes, requested: ' + count;
      throw IllegalStateException_init_$Create$(toString(message));
    }
  }
  function ByteArrayInput(array, startIndex, endIndex) {
    startIndex = startIndex === VOID ? 0 : startIndex;
    endIndex = endIndex === VOID ? array.length : endIndex;
    this.p1c_1 = array;
    this.q1c_1 = endIndex;
    this.r1c_1 = startIndex;
  }
  protoOf(ByteArrayInput).s1c = function () {
    return _get_available__saq0tk(this) === 0;
  };
  protoOf(ByteArrayInput).toString = function () {
    return 'ByteArrayInput(size=' + this.p1c_1.length + ', endIndex=' + this.q1c_1 + ', position=' + this.r1c_1 + ', available=' + _get_available__saq0tk(this) + ')';
  };
  protoOf(ByteArrayInput).t1c = function () {
    ensureAvailableBytes(this, 1);
    return this.p1c_1[this.r1c_1];
  };
  protoOf(ByteArrayInput).u1c = function () {
    ensureAvailableBytes(this, 1);
    var _unary__edvuaz = this.r1c_1;
    this.r1c_1 = _unary__edvuaz + 1 | 0;
    return this.p1c_1[_unary__edvuaz];
  };
  protoOf(ByteArrayInput).v1c = function (length) {
    ensureAvailableBytes(this, length);
    if (length === 0)
      return get_emptyArray();
    // Inline function 'kotlin.also' call
    var this_0 = copyOfRange(this.p1c_1, this.r1c_1, this.r1c_1 + length | 0);
    // Inline function 'kotlin.contracts.contract' call
    // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.ByteArrayInput.read.<anonymous>' call
    this.r1c_1 = this.r1c_1 + length | 0;
    return this_0;
  };
  protoOf(ByteArrayInput).w1c = function (length) {
    ensureAvailableBytes(this, length);
    if (length === 0)
      return new ByteArrayInput(get_emptyArray());
    // Inline function 'kotlin.also' call
    var this_0 = new ByteArrayInput(this.p1c_1, this.r1c_1, this.r1c_1 + length | 0);
    // Inline function 'kotlin.contracts.contract' call
    // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.ByteArrayInput.readSlice.<anonymous>' call
    this.r1c_1 = this.r1c_1 + length | 0;
    return this_0;
  };
  var properties_initialized_ByteArrayInput_kt_acgl39;
  function _init_properties_ByteArrayInput_kt__imea6f() {
    if (!properties_initialized_ByteArrayInput_kt_acgl39) {
      properties_initialized_ByteArrayInput_kt_acgl39 = true;
      emptyArray = new Int8Array(0);
    }
  }
  function ensureCapacity($this, elementsToAppend) {
    if (($this.f1c_1 + elementsToAppend | 0) <= $this.e1c_1.length)
      return Unit_instance;
    $this.e1c_1 = copyOf($this.e1c_1, takeHighestOneBit($this.f1c_1 + elementsToAppend | 0) << 1);
  }
  function ByteArrayOutput() {
    this.e1c_1 = new Int8Array(32);
    this.f1c_1 = 0;
  }
  protoOf(ByteArrayOutput).t = function () {
    return this.f1c_1;
  };
  protoOf(ByteArrayOutput).g1c = function () {
    return copyOf(this.e1c_1, this.f1c_1);
  };
  protoOf(ByteArrayOutput).x1c = function (byte) {
    ensureCapacity(this, 1);
    var tmp = this.e1c_1;
    var _unary__edvuaz = this.f1c_1;
    this.f1c_1 = _unary__edvuaz + 1 | 0;
    tmp[_unary__edvuaz] = byte;
  };
  protoOf(ByteArrayOutput).y1c = function (byte) {
    this.x1c(toByte(byte));
  };
  protoOf(ByteArrayOutput).z1c = function (bytes) {
    // Inline function 'kotlin.collections.isEmpty' call
    if (bytes.length === 0)
      return Unit_instance;
    ensureCapacity(this, bytes.length);
    // Inline function 'kotlin.collections.copyInto' call
    var destination = this.e1c_1;
    var destinationOffset = this.f1c_1;
    var endIndex = bytes.length;
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    var tmp = bytes;
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    arrayCopy(tmp, destination, destinationOffset, 0, endIndex);
    this.f1c_1 = this.f1c_1 + bytes.length | 0;
  };
  protoOf(ByteArrayOutput).a1d = function (output) {
    if (output.t() === 0)
      return Unit_instance;
    ensureCapacity(this, output.f1c_1);
    // Inline function 'kotlin.collections.copyInto' call
    var this_0 = output.e1c_1;
    var destination = this.e1c_1;
    var destinationOffset = this.f1c_1;
    var endIndex = output.f1c_1;
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    var tmp = this_0;
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    arrayCopy(tmp, destination, destinationOffset, 0, endIndex);
    this.f1c_1 = this.f1c_1 + output.f1c_1 | 0;
  };
  function getAndResetTagOverride($this) {
    var tag = $this.k1c_1;
    $this.k1c_1 = null;
    return tag;
  }
  function DerDecoder(der, byteArrayInput) {
    AbstractDecoder.call(this);
    this.h1c_1 = der;
    this.i1c_1 = new DerInput(byteArrayInput);
    this.j1c_1 = 0;
    this.k1c_1 = null;
  }
  protoOf(DerDecoder).wn = function () {
    return this.h1c_1.wn();
  };
  protoOf(DerDecoder).mr = function (descriptor) {
    if (this.i1c_1.s1c())
      return -1;
    var tag = this.i1c_1.c1d();
    $l$loop: while (true) {
      var index = this.j1c_1;
      this.k1c_1 = getElementContextSpecificTag(descriptor, index);
      if (descriptor.ip(index)) {
        var tmp$ret$1;
        $l$block: {
          // Inline function 'kotlin.checkNotNull' call
          var value = this.k1c_1;
          // Inline function 'kotlin.contracts.contract' call
          if (value == null) {
            // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerDecoder.decodeElementIndex.<anonymous>' call
            var message = 'Optional element ' + toString(descriptor) + '[' + index + '] must have context specific tag';
            throw IllegalStateException_init_$Create$(toString(message));
          } else {
            tmp$ret$1 = value;
            break $l$block;
          }
        }
        var requiredTag = tmp$ret$1;
        if (!(tag === get_tag(requiredTag))) {
          this.j1c_1 = this.j1c_1 + 1 | 0;
          continue $l$loop;
        }
      }
      var _unary__edvuaz = this.j1c_1;
      this.j1c_1 = _unary__edvuaz + 1 | 0;
      return _unary__edvuaz;
    }
  };
  protoOf(DerDecoder).hq = function () {
    return this.i1c_1.d1d();
  };
  protoOf(DerDecoder).iq = function () {
    return this.i1c_1.e1d();
  };
  protoOf(DerDecoder).kq = function () {
    return this.i1c_1.f1d(getAndResetTagOverride(this)).s3();
  };
  protoOf(DerDecoder).lq = function () {
    return this.i1c_1.f1d(getAndResetTagOverride(this)).t3();
  };
  protoOf(DerDecoder).mq = function () {
    return this.i1c_1.f1d(getAndResetTagOverride(this)).i1();
  };
  protoOf(DerDecoder).nq = function () {
    return this.i1c_1.f1d(getAndResetTagOverride(this)).m19();
  };
  protoOf(DerDecoder).ao = function (deserializer) {
    var tmp0_subject = deserializer.ln();
    var tmp;
    if (equals(tmp0_subject, ByteArraySerializer().ln())) {
      var tmp_0 = this.i1c_1.i1d(getAndResetTagOverride(this));
      tmp = !(tmp_0 == null) ? tmp_0 : THROW_CCE();
    } else if (equals(tmp0_subject, Companion_instance.k19().ln())) {
      var tmp_1 = this.i1c_1.h1d(getAndResetTagOverride(this));
      tmp = !(tmp_1 == null) ? tmp_1 : THROW_CCE();
    } else if (equals(tmp0_subject, Companion_instance_0.k19().ln())) {
      var tmp_2 = this.i1c_1.g1d(getAndResetTagOverride(this));
      tmp = !(new ObjectIdentifier(tmp_2) == null) ? new ObjectIdentifier(tmp_2) : THROW_CCE();
    } else if (equals(tmp0_subject, Companion_getInstance().k19().ln())) {
      var tmp_3 = this.i1c_1.f1d(getAndResetTagOverride(this));
      tmp = !(tmp_3 == null) ? tmp_3 : THROW_CCE();
    } else {
      tmp = deserializer.zn(this);
    }
    return tmp;
  };
  protoOf(DerDecoder).vq = function (descriptor) {
    var tmp0_subject = descriptor.bp();
    var tmp;
    var tmp_0;
    if (equals(tmp0_subject, CLASS_getInstance())) {
      tmp_0 = true;
    } else {
      tmp_0 = tmp0_subject instanceof PolymorphicKind;
    }
    if (tmp_0) {
      tmp = new DerDecoder(this.h1c_1, this.i1c_1.j1d(getAndResetTagOverride(this)));
    } else {
      throw SerializationException_init_$Create$('This serial kind is not supported as structure: ' + toString(descriptor));
    }
    return tmp;
  };
  protoOf(DerDecoder).tq = function (descriptor) {
    return this;
  };
  protoOf(DerDecoder).gr = function (descriptor, index) {
    return this;
  };
  protoOf(DerDecoder).sq = function (enumDescriptor) {
    var message = 'Enum decoding is not supported';
    throw IllegalStateException_init_$Create$(toString(message));
  };
  protoOf(DerDecoder).rq = function () {
    var message = 'String decoding is not supported';
    throw IllegalStateException_init_$Create$(toString(message));
  };
  protoOf(DerDecoder).qq = function () {
    var message = 'Char decoding is not supported';
    throw IllegalStateException_init_$Create$(toString(message));
  };
  protoOf(DerDecoder).jq = function () {
    var message = 'Boolean decoding is not supported';
    throw IllegalStateException_init_$Create$(toString(message));
  };
  protoOf(DerDecoder).pq = function () {
    var message = 'Double decoding is not supported';
    throw IllegalStateException_init_$Create$(toString(message));
  };
  protoOf(DerDecoder).oq = function () {
    var message = 'Float decoding is not supported';
    throw IllegalStateException_init_$Create$(toString(message));
  };
  protoOf(DerDecoder).gq = function () {
    var message = 'should not be called';
    throw IllegalStateException_init_$Create$(toString(message));
  };
  function getAndResetTagOverride_0($this) {
    var tag = $this.d1c_1;
    $this.d1c_1 = null;
    return tag;
  }
  function DerEncoder(der, byteArrayOutput, parentOutput) {
    parentOutput = parentOutput === VOID ? null : parentOutput;
    AbstractEncoder.call(this);
    this.a1c_1 = der;
    this.b1c_1 = parentOutput;
    this.c1c_1 = new DerOutput(byteArrayOutput);
    this.d1c_1 = null;
  }
  protoOf(DerEncoder).wn = function () {
    return this.a1c_1.wn();
  };
  protoOf(DerEncoder).rs = function (descriptor, index) {
    return false;
  };
  protoOf(DerEncoder).or = function (descriptor, index) {
    this.d1c_1 = getElementContextSpecificTag(descriptor, index);
    return true;
  };
  protoOf(DerEncoder).ps = function () {
  };
  protoOf(DerEncoder).qr = function () {
    return this.c1c_1.l1d();
  };
  protoOf(DerEncoder).sr = function (value) {
    return this.c1c_1.m1d(getAndResetTagOverride_0(this), toBigInt(value));
  };
  protoOf(DerEncoder).tr = function (value) {
    return this.c1c_1.m1d(getAndResetTagOverride_0(this), toBigInt_0(value));
  };
  protoOf(DerEncoder).ur = function (value) {
    return this.c1c_1.m1d(getAndResetTagOverride_0(this), toBigInt_1(value));
  };
  protoOf(DerEncoder).vr = function (value) {
    return this.c1c_1.m1d(getAndResetTagOverride_0(this), toBigInt_2(value));
  };
  protoOf(DerEncoder).xn = function (serializer, value) {
    var tmp0_subject = serializer.ln();
    var tmp;
    if (equals(tmp0_subject, ByteArraySerializer().ln())) {
      var tmp_0 = getAndResetTagOverride_0(this);
      this.c1c_1.p1d(tmp_0, (!(value == null) ? isByteArray(value) : false) ? value : THROW_CCE());
      tmp = Unit_instance;
    } else if (equals(tmp0_subject, Companion_instance.k19().ln())) {
      var tmp_1 = getAndResetTagOverride_0(this);
      this.c1c_1.o1d(tmp_1, value instanceof BitArray ? value : THROW_CCE());
      tmp = Unit_instance;
    } else if (equals(tmp0_subject, Companion_instance_0.k19().ln())) {
      var tmp_2 = getAndResetTagOverride_0(this);
      this.c1c_1.n1d(tmp_2, value instanceof ObjectIdentifier ? value.n1c_1 : THROW_CCE());
      tmp = Unit_instance;
    } else if (equals(tmp0_subject, Companion_getInstance().k19().ln())) {
      var tmp_3 = getAndResetTagOverride_0(this);
      this.c1c_1.m1d(tmp_3, value instanceof BigInt ? value : THROW_CCE());
      tmp = Unit_instance;
    } else {
      serializer.yn(this, value);
      tmp = Unit_instance;
    }
    return tmp;
  };
  protoOf(DerEncoder).vq = function (descriptor) {
    var tmp0_subject = descriptor.bp();
    var tmp;
    var tmp_0;
    if (equals(tmp0_subject, CLASS_getInstance())) {
      tmp_0 = true;
    } else {
      tmp_0 = tmp0_subject instanceof PolymorphicKind;
    }
    if (tmp_0) {
      tmp = new DerEncoder(this.a1c_1, new ByteArrayOutput(), this.c1c_1);
    } else {
      throw SerializationException_init_$Create$('This serial kind is not supported as structure: ' + toString(descriptor));
    }
    return tmp;
  };
  protoOf(DerEncoder).wq = function (descriptor) {
    var tmp$ret$1;
    $l$block: {
      // Inline function 'kotlin.checkNotNull' call
      var value = this.b1c_1;
      // Inline function 'kotlin.contracts.contract' call
      if (value == null) {
        // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerEncoder.endStructure.<anonymous>' call
        var message = 'Should be called after beginStructure';
        throw IllegalStateException_init_$Create$(toString(message));
      } else {
        tmp$ret$1 = value;
        break $l$block;
      }
    }
    tmp$ret$1.q1d(getAndResetTagOverride_0(this), this.c1c_1);
  };
  protoOf(DerEncoder).bs = function (descriptor) {
    return this;
  };
  protoOf(DerEncoder).r1d = function (enumDescriptor, index) {
    var message = 'Enum encoding is not supported';
    throw IllegalStateException_init_$Create$(toString(message));
  };
  protoOf(DerEncoder).as = function (enumDescriptor, index) {
    return this.r1d(enumDescriptor, index);
  };
  protoOf(DerEncoder).zr = function (value) {
    var message = 'String encoding is not supported';
    throw IllegalStateException_init_$Create$(toString(message));
  };
  protoOf(DerEncoder).yr = function (value) {
    var message = 'Char encoding is not supported';
    throw IllegalStateException_init_$Create$(toString(message));
  };
  protoOf(DerEncoder).rr = function (value) {
    var message = 'Boolean encoding is not supported';
    throw IllegalStateException_init_$Create$(toString(message));
  };
  protoOf(DerEncoder).wr = function (value) {
    var message = 'Float encoding is not supported';
    throw IllegalStateException_init_$Create$(toString(message));
  };
  protoOf(DerEncoder).xr = function (value) {
    var message = 'Double encoding is not supported';
    throw IllegalStateException_init_$Create$(toString(message));
  };
  protoOf(DerEncoder).pr = function (value) {
    var message = 'should not be called';
    throw IllegalStateException_init_$Create$(toString(message));
  };
  function DerInput(input) {
    this.b1d_1 = input;
  }
  protoOf(DerInput).s1c = function () {
    return this.b1d_1.s1c();
  };
  protoOf(DerInput).c1d = function () {
    return this.b1d_1.t1c();
  };
  protoOf(DerInput).d1d = function () {
    return !(this.c1d() === 5);
  };
  protoOf(DerInput).e1d = function () {
    readRequestedTag$accessor$1oqaisj(this.b1d_1, 5);
    var length = readLength$accessor$1oqaisj(this.b1d_1);
    // Inline function 'kotlin.check' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(length === 0)) {
      // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerInput.readNull.<anonymous>' call
      var message = 'NULL tag length should be zero, but was: ' + length;
      throw IllegalStateException_init_$Create$(toString(message));
    }
    return null;
  };
  protoOf(DerInput).f1d = function (tagOverride) {
    var tmp$ret$1;
    $l$block: {
      // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.readTagWithOverride' call
      var this_0 = this.b1d_1;
      if (tagOverride == null) {
        readRequestedTag$accessor$1oqaisj(this_0, 2);
        // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerInput.readInteger.<anonymous>' call
        var length = readLength$accessor$1oqaisj(this_0);
        var bytes = this_0.v1c(length);
        tmp$ret$1 = decodeToBigInt(bytes);
        break $l$block;
      }
      readRequestedTag$accessor$1oqaisj(this_0, get_tag(tagOverride));
      var tmp;
      switch (tagOverride.s1b_1.v2_1) {
        case 0:
          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerInput.readInteger.<anonymous>' call

          var length_0 = readLength$accessor$1oqaisj(this_0);
          var bytes_0 = this_0.v1c(length_0);
          tmp = decodeToBigInt(bytes_0);
          break;
        case 1:
          var length_1 = readLength$accessor$1oqaisj(this_0);
          var explicitInput = this_0.w1c(length_1);
          readRequestedTag$accessor$1oqaisj(explicitInput, 2);
          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerInput.readInteger.<anonymous>' call

          var length_2 = readLength$accessor$1oqaisj(explicitInput);
          var bytes_1 = explicitInput.v1c(length_2);
          tmp = decodeToBigInt(bytes_1);
          break;
        default:
          noWhenBranchMatchedException();
          break;
      }
      tmp$ret$1 = tmp;
    }
    return tmp$ret$1;
  };
  protoOf(DerInput).h1d = function (tagOverride) {
    var tmp$ret$5;
    $l$block: {
      // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.readTagWithOverride' call
      var this_0 = this.b1d_1;
      if (tagOverride == null) {
        readRequestedTag$accessor$1oqaisj(this_0, 3);
        // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerInput.readBitString.<anonymous>' call
        var length = readLength$accessor$1oqaisj(this_0);
        var unusedBits = this_0.u1c();
        var bytes = this_0.v1c(length - 1 | 0);
        // Inline function 'kotlin.collections.isEmpty' call
        if (bytes.length === 0) {
          // Inline function 'kotlin.check' call
          // Inline function 'kotlin.contracts.contract' call
          if (!(unusedBits === 0)) {
            // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerInput.readBitString.<anonymous>.<anonymous>' call
            var message = 'wrong number of unused bits, expected 0, received: ' + unusedBits;
            throw IllegalStateException_init_$Create$(toString(message));
          }
        } else {
          // Inline function 'kotlin.countTrailingZeroBits' call
          var this_1 = last(bytes);
          var trailingZeros = countTrailingZeroBits(this_1 | 256);
          // Inline function 'kotlin.check' call
          // Inline function 'kotlin.contracts.contract' call
          if (!(unusedBits <= trailingZeros)) {
            // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerInput.readBitString.<anonymous>.<anonymous>' call
            var message_0 = 'Not all unused bits are zeros, expected at least ' + unusedBits + ' trailing zeros, received ' + trailingZeros;
            throw IllegalStateException_init_$Create$(toString(message_0));
          }
        }
        tmp$ret$5 = new BitArray(unusedBits, bytes);
        break $l$block;
      }
      readRequestedTag$accessor$1oqaisj(this_0, get_tag(tagOverride));
      var tmp;
      switch (tagOverride.s1b_1.v2_1) {
        case 0:
          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerInput.readBitString.<anonymous>' call

          var length_0 = readLength$accessor$1oqaisj(this_0);
          var unusedBits_0 = this_0.u1c();
          var bytes_0 = this_0.v1c(length_0 - 1 | 0);
          // Inline function 'kotlin.collections.isEmpty' call

          if (bytes_0.length === 0) {
            // Inline function 'kotlin.check' call
            // Inline function 'kotlin.contracts.contract' call
            if (!(unusedBits_0 === 0)) {
              // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerInput.readBitString.<anonymous>.<anonymous>' call
              var message_1 = 'wrong number of unused bits, expected 0, received: ' + unusedBits_0;
              throw IllegalStateException_init_$Create$(toString(message_1));
            }
          } else {
            // Inline function 'kotlin.countTrailingZeroBits' call
            var this_2 = last(bytes_0);
            var trailingZeros_0 = countTrailingZeroBits(this_2 | 256);
            // Inline function 'kotlin.check' call
            // Inline function 'kotlin.contracts.contract' call
            if (!(unusedBits_0 <= trailingZeros_0)) {
              // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerInput.readBitString.<anonymous>.<anonymous>' call
              var message_2 = 'Not all unused bits are zeros, expected at least ' + unusedBits_0 + ' trailing zeros, received ' + trailingZeros_0;
              throw IllegalStateException_init_$Create$(toString(message_2));
            }
          }

          tmp = new BitArray(unusedBits_0, bytes_0);
          break;
        case 1:
          var length_1 = readLength$accessor$1oqaisj(this_0);
          var explicitInput = this_0.w1c(length_1);
          readRequestedTag$accessor$1oqaisj(explicitInput, 3);
          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerInput.readBitString.<anonymous>' call

          var length_2 = readLength$accessor$1oqaisj(explicitInput);
          var unusedBits_1 = explicitInput.u1c();
          var bytes_1 = explicitInput.v1c(length_2 - 1 | 0);
          // Inline function 'kotlin.collections.isEmpty' call

          if (bytes_1.length === 0) {
            // Inline function 'kotlin.check' call
            // Inline function 'kotlin.contracts.contract' call
            if (!(unusedBits_1 === 0)) {
              // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerInput.readBitString.<anonymous>.<anonymous>' call
              var message_3 = 'wrong number of unused bits, expected 0, received: ' + unusedBits_1;
              throw IllegalStateException_init_$Create$(toString(message_3));
            }
          } else {
            // Inline function 'kotlin.countTrailingZeroBits' call
            var this_3 = last(bytes_1);
            var trailingZeros_1 = countTrailingZeroBits(this_3 | 256);
            // Inline function 'kotlin.check' call
            // Inline function 'kotlin.contracts.contract' call
            if (!(unusedBits_1 <= trailingZeros_1)) {
              // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerInput.readBitString.<anonymous>.<anonymous>' call
              var message_4 = 'Not all unused bits are zeros, expected at least ' + unusedBits_1 + ' trailing zeros, received ' + trailingZeros_1;
              throw IllegalStateException_init_$Create$(toString(message_4));
            }
          }

          tmp = new BitArray(unusedBits_1, bytes_1);
          break;
        default:
          noWhenBranchMatchedException();
          break;
      }
      tmp$ret$5 = tmp;
    }
    return tmp$ret$5;
  };
  protoOf(DerInput).i1d = function (tagOverride) {
    var tmp$ret$1;
    $l$block: {
      // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.readTagWithOverride' call
      var this_0 = this.b1d_1;
      if (tagOverride == null) {
        readRequestedTag$accessor$1oqaisj(this_0, 4);
        // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerInput.readOctetString.<anonymous>' call
        var length = readLength$accessor$1oqaisj(this_0);
        tmp$ret$1 = this_0.v1c(length);
        break $l$block;
      }
      readRequestedTag$accessor$1oqaisj(this_0, get_tag(tagOverride));
      var tmp;
      switch (tagOverride.s1b_1.v2_1) {
        case 0:
          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerInput.readOctetString.<anonymous>' call

          var length_0 = readLength$accessor$1oqaisj(this_0);
          tmp = this_0.v1c(length_0);
          break;
        case 1:
          var length_1 = readLength$accessor$1oqaisj(this_0);
          var explicitInput = this_0.w1c(length_1);
          readRequestedTag$accessor$1oqaisj(explicitInput, 4);
          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerInput.readOctetString.<anonymous>' call

          var length_2 = readLength$accessor$1oqaisj(explicitInput);
          tmp = explicitInput.v1c(length_2);
          break;
        default:
          noWhenBranchMatchedException();
          break;
      }
      tmp$ret$1 = tmp;
    }
    return tmp$ret$1;
  };
  protoOf(DerInput).g1d = function (tagOverride) {
    var tmp$ret$1;
    $l$block: {
      // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.readTagWithOverride' call
      var this_0 = this.b1d_1;
      if (tagOverride == null) {
        readRequestedTag$accessor$1oqaisj(this_0, 6);
        // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerInput.readObjectIdentifier.<anonymous>' call
        var length = readLength$accessor$1oqaisj(this_0);
        var slice = this_0.w1c(length);
        tmp$ret$1 = _ObjectIdentifier___init__impl__7hutfr(readOidElements(slice));
        break $l$block;
      }
      readRequestedTag$accessor$1oqaisj(this_0, get_tag(tagOverride));
      var tmp;
      switch (tagOverride.s1b_1.v2_1) {
        case 0:
          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerInput.readObjectIdentifier.<anonymous>' call

          var length_0 = readLength$accessor$1oqaisj(this_0);
          var slice_0 = this_0.w1c(length_0);
          tmp = new ObjectIdentifier(_ObjectIdentifier___init__impl__7hutfr(readOidElements(slice_0)));
          break;
        case 1:
          var length_1 = readLength$accessor$1oqaisj(this_0);
          var explicitInput = this_0.w1c(length_1);
          readRequestedTag$accessor$1oqaisj(explicitInput, 6);
          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerInput.readObjectIdentifier.<anonymous>' call

          var length_2 = readLength$accessor$1oqaisj(explicitInput);
          var slice_1 = explicitInput.w1c(length_2);
          tmp = new ObjectIdentifier(_ObjectIdentifier___init__impl__7hutfr(readOidElements(slice_1)));
          break;
        default:
          noWhenBranchMatchedException();
          break;
      }
      tmp$ret$1 = tmp.n1c_1;
    }
    return tmp$ret$1;
  };
  protoOf(DerInput).j1d = function (tagOverride) {
    var tmp$ret$1;
    $l$block: {
      // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.readTagWithOverride' call
      var this_0 = this.b1d_1;
      if (tagOverride == null) {
        readRequestedTag$accessor$1oqaisj(this_0, 48);
        // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerInput.readSequence.<anonymous>' call
        var length = readLength$accessor$1oqaisj(this_0);
        tmp$ret$1 = this_0.w1c(length);
        break $l$block;
      }
      readRequestedTag$accessor$1oqaisj(this_0, get_tag(tagOverride));
      var tmp;
      switch (tagOverride.s1b_1.v2_1) {
        case 0:
          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerInput.readSequence.<anonymous>' call

          var length_0 = readLength$accessor$1oqaisj(this_0);
          tmp = this_0.w1c(length_0);
          break;
        case 1:
          var length_1 = readLength$accessor$1oqaisj(this_0);
          var explicitInput = this_0.w1c(length_1);
          readRequestedTag$accessor$1oqaisj(explicitInput, 48);
          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerInput.readSequence.<anonymous>' call

          var length_2 = readLength$accessor$1oqaisj(explicitInput);
          tmp = explicitInput.w1c(length_2);
          break;
        default:
          noWhenBranchMatchedException();
          break;
      }
      tmp$ret$1 = tmp;
    }
    return tmp$ret$1;
  };
  function readRequestedTag(_this__u8e3s4, tag) {
    var currentTag = _this__u8e3s4.u1c();
    // Inline function 'kotlin.check' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(currentTag === tag)) {
      // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.readRequestedTag.<anonymous>' call
      var message = "Requested tag '" + DerTag_name(tag) + "', received: '" + DerTag_name(currentTag) + "'";
      throw IllegalStateException_init_$Create$(toString(message));
    }
  }
  function readLength(_this__u8e3s4) {
    var first = _this__u8e3s4.u1c();
    if (first >= 0)
      return first;
    var numberOfLengthBytes = first & 127;
    // Inline function 'kotlin.check' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(numberOfLengthBytes <= 4)) {
      // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.readLength.<anonymous>' call
      var message = 'Supported number of bytes for tag length are in range 1..4, but was: ' + numberOfLengthBytes + ' ';
      throw IllegalStateException_init_$Create$(toString(message));
    }
    var length = 0;
    // Inline function 'kotlin.repeat' call
    // Inline function 'kotlin.contracts.contract' call
    var inductionVariable = 0;
    if (inductionVariable < numberOfLengthBytes)
      do {
        var index = inductionVariable;
        inductionVariable = inductionVariable + 1 | 0;
        // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.readLength.<anonymous>' call
        length = (length << 8) + (_this__u8e3s4.u1c() & 255) | 0;
      }
       while (inductionVariable < numberOfLengthBytes);
    // Inline function 'kotlin.check' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(length > 0)) {
      // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.readLength.<anonymous>' call
      var message_0 = 'length overflow: ' + length;
      throw IllegalStateException_init_$Create$(toString(message_0));
    }
    return length;
  }
  function readOidElements(_this__u8e3s4) {
    // Inline function 'kotlin.text.buildString' call
    // Inline function 'kotlin.contracts.contract' call
    // Inline function 'kotlin.apply' call
    var this_0 = StringBuilder_init_$Create$();
    // Inline function 'kotlin.contracts.contract' call
    // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.readOidElements.<anonymous>' call
    var first = readOidElement(_this__u8e3s4);
    if (first < 40)
      this_0.t8(_Char___init__impl__6a9atx(48)).t8(_Char___init__impl__6a9atx(46)).vc(first);
    else if (first < 80)
      this_0.t8(_Char___init__impl__6a9atx(49)).t8(_Char___init__impl__6a9atx(46)).vc(first - 40 | 0);
    else
      this_0.t8(_Char___init__impl__6a9atx(50)).t8(_Char___init__impl__6a9atx(46)).vc(first - 80 | 0);
    while (!_this__u8e3s4.s1c()) {
      this_0.t8(_Char___init__impl__6a9atx(46)).vc(readOidElement(_this__u8e3s4));
    }
    return this_0.toString();
  }
  function readOidElement(_this__u8e3s4) {
    var element = 0;
    do {
      var b = _this__u8e3s4.u1c();
      element = (element << 7) + (b & 127) | 0;
    }
     while ((b & 128) === 128);
    // Inline function 'kotlin.check' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(element >= 0)) {
      // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.readOidElement.<anonymous>' call
      var message = 'element overflow: ' + element;
      throw IllegalStateException_init_$Create$(toString(message));
    }
    return element;
  }
  function readLength$accessor$1oqaisj(_this__u8e3s4) {
    return readLength(_this__u8e3s4);
  }
  function readRequestedTag$accessor$1oqaisj(_this__u8e3s4, tag) {
    return readRequestedTag(_this__u8e3s4, tag);
  }
  function DerOutput(output) {
    this.k1d_1 = output;
  }
  protoOf(DerOutput).l1d = function () {
    this.k1d_1.x1c(5);
    writeLength(this.k1d_1, 0);
  };
  protoOf(DerOutput).m1d = function (tagOverride, value) {
    $l$block: {
      // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.writeTagWithOverride' call
      var this_0 = this.k1d_1;
      if (tagOverride == null) {
        this_0.x1c(2);
        // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerOutput.writeInteger.<anonymous>' call
        writeBytes(this_0, encodeToByteArray(value));
        break $l$block;
      }
      this_0.x1c(get_tag(tagOverride));
      switch (tagOverride.s1b_1.v2_1) {
        case 0:
          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerOutput.writeInteger.<anonymous>' call

          writeBytes(this_0, encodeToByteArray(value));
          break;
        case 1:
          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.writeBytes' call

          // Inline function 'kotlin.apply' call

          var this_1 = new ByteArrayOutput();
          // Inline function 'kotlin.contracts.contract' call

          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.writeTagWithOverride.<anonymous>' call

          this_1.x1c(2);
          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerOutput.writeInteger.<anonymous>' call

          writeBytes(this_1, encodeToByteArray(value));
          writeBytes$accessor$17lu55s(this_0, this_1);
          break;
        default:
          noWhenBranchMatchedException();
          break;
      }
    }
  };
  protoOf(DerOutput).o1d = function (tagOverride, bits) {
    $l$block: {
      // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.writeTagWithOverride' call
      var this_0 = this.k1d_1;
      if (tagOverride == null) {
        this_0.x1c(3);
        // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerOutput.writeBitString.<anonymous>' call
        writeLength(this_0, bits.w1b_1.length + 1 | 0);
        this_0.y1c(bits.v1b_1);
        this_0.z1c(bits.w1b_1);
        break $l$block;
      }
      this_0.x1c(get_tag(tagOverride));
      switch (tagOverride.s1b_1.v2_1) {
        case 0:
          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerOutput.writeBitString.<anonymous>' call

          writeLength(this_0, bits.w1b_1.length + 1 | 0);
          this_0.y1c(bits.v1b_1);
          this_0.z1c(bits.w1b_1);
          break;
        case 1:
          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.writeBytes' call

          // Inline function 'kotlin.apply' call

          var this_1 = new ByteArrayOutput();
          // Inline function 'kotlin.contracts.contract' call

          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.writeTagWithOverride.<anonymous>' call

          this_1.x1c(3);
          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerOutput.writeBitString.<anonymous>' call

          writeLength(this_1, bits.w1b_1.length + 1 | 0);
          this_1.y1c(bits.v1b_1);
          this_1.z1c(bits.w1b_1);
          writeBytes$accessor$17lu55s(this_0, this_1);
          break;
        default:
          noWhenBranchMatchedException();
          break;
      }
    }
  };
  protoOf(DerOutput).p1d = function (tagOverride, bytes) {
    $l$block: {
      // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.writeTagWithOverride' call
      var this_0 = this.k1d_1;
      if (tagOverride == null) {
        this_0.x1c(4);
        // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerOutput.writeOctetString.<anonymous>' call
        writeBytes(this_0, bytes);
        break $l$block;
      }
      this_0.x1c(get_tag(tagOverride));
      switch (tagOverride.s1b_1.v2_1) {
        case 0:
          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerOutput.writeOctetString.<anonymous>' call

          writeBytes(this_0, bytes);
          break;
        case 1:
          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.writeBytes' call

          // Inline function 'kotlin.apply' call

          var this_1 = new ByteArrayOutput();
          // Inline function 'kotlin.contracts.contract' call

          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.writeTagWithOverride.<anonymous>' call

          this_1.x1c(4);
          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerOutput.writeOctetString.<anonymous>' call

          writeBytes(this_1, bytes);
          writeBytes$accessor$17lu55s(this_0, this_1);
          break;
        default:
          noWhenBranchMatchedException();
          break;
      }
    }
  };
  protoOf(DerOutput).n1d = function (tagOverride, value) {
    $l$block: {
      // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.writeTagWithOverride' call
      var this_0 = this.k1d_1;
      if (tagOverride == null) {
        this_0.x1c(6);
        // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerOutput.writeObjectIdentifier.<anonymous>' call
        // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.writeBytes' call
        // Inline function 'kotlin.apply' call
        var this_1 = new ByteArrayOutput();
        // Inline function 'kotlin.contracts.contract' call
        // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerOutput.writeObjectIdentifier.<anonymous>.<anonymous>' call
        writeOidElements(this_1, split(_ObjectIdentifier___get_value__impl__wqzjw3(value), ['.']));
        writeBytes$accessor$17lu55s(this_0, this_1);
        break $l$block;
      }
      this_0.x1c(get_tag(tagOverride));
      switch (tagOverride.s1b_1.v2_1) {
        case 0:
          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerOutput.writeObjectIdentifier.<anonymous>' call

          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.writeBytes' call

          // Inline function 'kotlin.apply' call

          var this_2 = new ByteArrayOutput();
          // Inline function 'kotlin.contracts.contract' call

          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerOutput.writeObjectIdentifier.<anonymous>.<anonymous>' call

          writeOidElements(this_2, split(_ObjectIdentifier___get_value__impl__wqzjw3(value), ['.']));
          writeBytes$accessor$17lu55s(this_0, this_2);
          break;
        case 1:
          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.writeBytes' call

          // Inline function 'kotlin.apply' call

          var this_3 = new ByteArrayOutput();
          // Inline function 'kotlin.contracts.contract' call

          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.writeTagWithOverride.<anonymous>' call

          this_3.x1c(6);
          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerOutput.writeObjectIdentifier.<anonymous>' call

          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.writeBytes' call

          // Inline function 'kotlin.apply' call

          var this_4 = new ByteArrayOutput();
          // Inline function 'kotlin.contracts.contract' call

          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerOutput.writeObjectIdentifier.<anonymous>.<anonymous>' call

          writeOidElements(this_4, split(_ObjectIdentifier___get_value__impl__wqzjw3(value), ['.']));
          writeBytes$accessor$17lu55s(this_3, this_4);
          writeBytes$accessor$17lu55s(this_0, this_3);
          break;
        default:
          noWhenBranchMatchedException();
          break;
      }
    }
  };
  protoOf(DerOutput).q1d = function (tagOverride, bytes) {
    $l$block: {
      // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.writeTagWithOverride' call
      var this_0 = this.k1d_1;
      if (tagOverride == null) {
        this_0.x1c(48);
        // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerOutput.writeSequence.<anonymous>' call
        writeBytes$accessor$17lu55s(this_0, bytes.k1d_1);
        break $l$block;
      }
      this_0.x1c(get_tag(tagOverride));
      switch (tagOverride.s1b_1.v2_1) {
        case 0:
          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerOutput.writeSequence.<anonymous>' call

          writeBytes$accessor$17lu55s(this_0, bytes.k1d_1);
          break;
        case 1:
          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.writeBytes' call

          // Inline function 'kotlin.apply' call

          var this_1 = new ByteArrayOutput();
          // Inline function 'kotlin.contracts.contract' call

          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.writeTagWithOverride.<anonymous>' call

          this_1.x1c(48);
          // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.DerOutput.writeSequence.<anonymous>' call

          writeBytes$accessor$17lu55s(this_1, bytes.k1d_1);
          writeBytes$accessor$17lu55s(this_0, this_1);
          break;
        default:
          noWhenBranchMatchedException();
          break;
      }
    }
  };
  function writeLength(_this__u8e3s4, length) {
    if (length < 128)
      return _this__u8e3s4.y1c(length);
    // Inline function 'kotlin.countLeadingZeroBits' call
    var numberOfLengthBytes = 4 - (clz32(length) / 8 | 0) | 0;
    _this__u8e3s4.y1c(numberOfLengthBytes | 128);
    // Inline function 'kotlin.repeat' call
    // Inline function 'kotlin.contracts.contract' call
    var inductionVariable = 0;
    if (inductionVariable < numberOfLengthBytes)
      do {
        var index = inductionVariable;
        inductionVariable = inductionVariable + 1 | 0;
        // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.writeLength.<anonymous>' call
        _this__u8e3s4.y1c(length >>> imul(8, (numberOfLengthBytes - 1 | 0) - index | 0) | 0);
      }
       while (inductionVariable < numberOfLengthBytes);
  }
  function writeBytes(_this__u8e3s4, bytes) {
    writeLength(_this__u8e3s4, bytes.length);
    _this__u8e3s4.z1c(bytes);
  }
  function writeOidElements(_this__u8e3s4, elements) {
    // Inline function 'kotlin.check' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(elements.t() >= 2)) {
      // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.writeOidElements.<anonymous>' call
      var message = 'at least 2 components expected but was ' + elements.t();
      throw IllegalStateException_init_$Create$(toString(message));
    }
    writeOidElement(_this__u8e3s4, imul(writeOidElements$element(elements, 0), 40) + writeOidElements$element(elements, 1) | 0);
    // Inline function 'kotlin.repeat' call
    var times = elements.t() - 2 | 0;
    // Inline function 'kotlin.contracts.contract' call
    var inductionVariable = 0;
    if (inductionVariable < times)
      do {
        var index = inductionVariable;
        inductionVariable = inductionVariable + 1 | 0;
        // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.writeOidElements.<anonymous>' call
        writeOidElement(_this__u8e3s4, writeOidElements$element(elements, index + 2 | 0));
      }
       while (inductionVariable < times);
  }
  function writeBytes_0(_this__u8e3s4, bytes) {
    writeLength(_this__u8e3s4, bytes.t());
    _this__u8e3s4.a1d(bytes);
  }
  function writeOidElement(_this__u8e3s4, element) {
    if (element < 128)
      return _this__u8e3s4.y1c(element);
    // Inline function 'kotlin.countLeadingZeroBits' call
    var l = (32 - clz32(element) | 0) / 7 | 0;
    // Inline function 'kotlin.repeat' call
    // Inline function 'kotlin.contracts.contract' call
    var inductionVariable = 0;
    if (inductionVariable < l)
      do {
        var index = inductionVariable;
        inductionVariable = inductionVariable + 1 | 0;
        // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.writeOidElement.<anonymous>' call
        var value = element >>> imul(l - index | 0, 7) | 0;
        if (!(value === 0)) {
          _this__u8e3s4.y1c(value & 127 | 128);
        }
      }
       while (inductionVariable < l);
    _this__u8e3s4.y1c(element & 127);
  }
  function writeBytes$accessor$17lu55s(_this__u8e3s4, bytes) {
    return writeBytes_0(_this__u8e3s4, bytes);
  }
  function writeOidElements$element($elements, index) {
    return toInt($elements.u(index));
  }
  function getElementContextSpecificTag(_this__u8e3s4, index) {
    var tmp$ret$1;
    $l$block: {
      // Inline function 'kotlin.collections.firstOrNull' call
      var tmp0_iterator = _this__u8e3s4.gp(index).q();
      while (tmp0_iterator.r()) {
        var element = tmp0_iterator.s();
        // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.getElementContextSpecificTag.<anonymous>' call
        if (element instanceof ContextSpecificTag) {
          tmp$ret$1 = element;
          break $l$block;
        }
      }
      tmp$ret$1 = null;
    }
    var tmp = tmp$ret$1;
    var tmp0_elvis_lhs = tmp instanceof ContextSpecificTag ? tmp : null;
    var tmp_0;
    if (tmp0_elvis_lhs == null) {
      return null;
    } else {
      tmp_0 = tmp0_elvis_lhs;
    }
    var annotation = tmp_0;
    // Inline function 'kotlin.check' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(annotation.r1b_1 < 31)) {
      // Inline function 'dev.whyoleg.cryptography.serialization.asn1.internal.getElementContextSpecificTag.<anonymous>' call
      var message = 'Context specific tag class must be less than 31, but was ' + annotation.r1b_1;
      throw IllegalStateException_init_$Create$(toString(message));
    }
    return annotation;
  }
  function get_tag(_this__u8e3s4) {
    var contextSpecificTag = _this__u8e3s4.r1b_1 | 128;
    var tmp;
    switch (_this__u8e3s4.s1b_1.v2_1) {
      case 0:
        tmp = toByte(contextSpecificTag);
        break;
      case 1:
        tmp = toByte(contextSpecificTag | 32);
        break;
      default:
        noWhenBranchMatchedException();
        break;
    }
    return tmp;
  }
  function DerTag_name(tag) {
    var tmp;
    if (tag === 2) {
      tmp = 'INTEGER';
    } else if (tag === 3) {
      tmp = 'BIT_STRING';
    } else if (tag === 4) {
      tmp = 'OCTET_STRING';
    } else if (tag === 5) {
      tmp = 'NULL';
    } else if (tag === 6) {
      tmp = 'OID';
    } else if (tag === 48) {
      tmp = 'SEQUENCE';
    } else {
      var intTag = tag;
      var tmp_0;
      if (hasFlag(intTag, 128)) {
        var tmp_1;
        if (hasFlag(intTag, 32)) {
          // Inline function 'kotlin.toUByte' call
          var this_0 = (intTag - 128 | 0) - 32 | 0;
          var tmp$ret$0 = _UByte___init__impl__g9hnc4(toByte(this_0));
          tmp_1 = 'CONTEXT_SPECIFIC:EXPLICIT[' + new UByte(tmp$ret$0) + ']';
        } else {
          // Inline function 'kotlin.toUByte' call
          var this_1 = intTag - 128 | 0;
          var tmp$ret$1 = _UByte___init__impl__g9hnc4(toByte(this_1));
          tmp_1 = 'CONTEXT_SPECIFIC:IMPLICIT[' + new UByte(tmp$ret$1) + ']';
        }
        tmp_0 = tmp_1;
      } else {
        // Inline function 'kotlin.toUByte' call
        var tmp$ret$2 = _UByte___init__impl__g9hnc4(tag);
        tmp_0 = 'UNKNOWN[0x' + toString_0(tmp$ret$2, 16) + ']';
      }
      tmp = tmp_0;
    }
    return tmp;
  }
  function hasFlag(_this__u8e3s4, flag) {
    return (_this__u8e3s4 & flag) === flag;
  }
  //region block: post-declaration
  protoOf($serializer).zy = typeParametersSerializers;
  protoOf($serializer_0).zy = typeParametersSerializers;
  //endregion
  //region block: init
  Companion_instance = new Companion();
  Companion_instance_0 = new Companion_0();
  //endregion
  //region block: exports
  _.$_$ = _.$_$ || {};
  _.$_$.a = BitArray;
  _.$_$.b = ContextSpecificTag;
  _.$_$.c = ObjectIdentifier;
  _.$_$.d = TagType_EXPLICIT_getInstance;
  _.$_$.e = _ObjectIdentifier___init__impl__7hutfr;
  _.$_$.f = ObjectIdentifier__hashCode_impl_7x941i;
  _.$_$.g = ObjectIdentifier__toString_impl_dtwsrd;
  _.$_$.h = $serializer_getInstance;
  _.$_$.i = Default_getInstance;
  _.$_$.j = $serializer_getInstance_0;
  _.$_$.k = Companion_instance_0;
  //endregion
  return _;
}));
