// @sphereon/kmp-mdoc-core/cryptography-kotlin-cryptography-bigint.js
(function (factory) {
  if (typeof define === 'function' && define.amd)
    define(['exports', './kotlinx-serialization-kotlinx-serialization-core.js', './kotlin-kotlin-stdlib.js'], factory);
  else if (typeof exports === 'object')
    factory(module.exports, require('./kotlinx-serialization-kotlinx-serialization-core.js'), require('./kotlin-kotlin-stdlib.js'));
  else {
    if (typeof globalThis['kotlinx-serialization-kotlinx-serialization-core'] === 'undefined') {
      throw new Error("Error loading module 'cryptography-kotlin-cryptography-bigint'. Its dependency 'kotlinx-serialization-kotlinx-serialization-core' was not found. Please, check whether 'kotlinx-serialization-kotlinx-serialization-core' is loaded prior to 'cryptography-kotlin-cryptography-bigint'.");
    }
    if (typeof globalThis['kotlin-kotlin-stdlib'] === 'undefined') {
      throw new Error("Error loading module 'cryptography-kotlin-cryptography-bigint'. Its dependency 'kotlin-kotlin-stdlib' was not found. Please, check whether 'kotlin-kotlin-stdlib' is loaded prior to 'cryptography-kotlin-cryptography-bigint'.");
    }
    globalThis['cryptography-kotlin-cryptography-bigint'] = factory(typeof globalThis['cryptography-kotlin-cryptography-bigint'] === 'undefined' ? {} : globalThis['cryptography-kotlin-cryptography-bigint'], globalThis['kotlinx-serialization-kotlinx-serialization-core'], globalThis['kotlin-kotlin-stdlib']);
  }
}(function (_, kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core, kotlin_kotlin) {
  'use strict';
  //region block: imports
  var STRING_getInstance = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.j;
  var PrimitiveSerialDescriptor = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.h1;
  var protoOf = kotlin_kotlin.$_$.xa;
  var THROW_CCE = kotlin_kotlin.$_$.ge;
  var KSerializer = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.r2;
  var initMetadataForObject = kotlin_kotlin.$_$.ba;
  var VOID = kotlin_kotlin.$_$.f;
  var arrayCopy = kotlin_kotlin.$_$.f5;
  var Unit_instance = kotlin_kotlin.$_$.q4;
  var initMetadataForCompanion = kotlin_kotlin.$_$.x9;
  var Number_0 = kotlin_kotlin.$_$.ce;
  var toByte = kotlin_kotlin.$_$.uc;
  var toShort = kotlin_kotlin.$_$.dd;
  var toInt = kotlin_kotlin.$_$.ad;
  var toLong = kotlin_kotlin.$_$.cd;
  var getStringHashCode = kotlin_kotlin.$_$.u9;
  var Comparable = kotlin_kotlin.$_$.qd;
  var initMetadataForClass = kotlin_kotlin.$_$.w9;
  var toString = kotlin_kotlin.$_$.bb;
  var IllegalArgumentException_init_$Create$ = kotlin_kotlin.$_$.n1;
  var toHexString = kotlin_kotlin.$_$.yc;
  var Long = kotlin_kotlin.$_$.ae;
  var toByte_0 = kotlin_kotlin.$_$.ya;
  var isBlank = kotlin_kotlin.$_$.dc;
  var IllegalStateException_init_$Create$ = kotlin_kotlin.$_$.q1;
  var hexToByteArray = kotlin_kotlin.$_$.ac;
  //endregion
  //region block: pre-declaration
  initMetadataForObject(BigIntAsStringSerializer, 'BigIntAsStringSerializer', VOID, VOID, [KSerializer]);
  initMetadataForCompanion(Companion);
  initMetadataForClass(BigInt_0, 'BigInt', VOID, Number_0, [Number_0, Comparable], VOID, VOID, {0: BigIntAsStringSerializer_getInstance});
  //endregion
  function BigIntAsStringSerializer() {
    BigIntAsStringSerializer_instance = this;
    this.f19_1 = PrimitiveSerialDescriptor('BigInt', STRING_getInstance());
  }
  protoOf(BigIntAsStringSerializer).ln = function () {
    return this.f19_1;
  };
  protoOf(BigIntAsStringSerializer).g19 = function (encoder, value) {
    encoder.zr(value.toString());
  };
  protoOf(BigIntAsStringSerializer).yn = function (encoder, value) {
    return this.g19(encoder, value instanceof BigInt_0 ? value : THROW_CCE());
  };
  protoOf(BigIntAsStringSerializer).zn = function (decoder) {
    return toBigInt_3(decoder.rq());
  };
  var BigIntAsStringSerializer_instance;
  function BigIntAsStringSerializer_getInstance() {
    if (BigIntAsStringSerializer_instance == null)
      new BigIntAsStringSerializer();
    return BigIntAsStringSerializer_instance;
  }
  function encodeToByteArray(_this__u8e3s4) {
    var positive = _this__u8e3s4.i19(0) >= 0;
    var bytes = positive ? encodeToByteArray$decodeFromHex(_this__u8e3s4.h19_1.toString(16)) : invertTwoComplement(encodeToByteArray$decodeFromHex(jsBigIntNegate(_this__u8e3s4.h19_1).toString(16)));
    var firstBytePositive = bytes[0] >= 0;
    if (positive === firstBytePositive)
      return bytes;
    // Inline function 'kotlin.also' call
    var this_0 = new Int8Array(bytes.length + 1 | 0);
    // Inline function 'kotlin.contracts.contract' call
    // Inline function 'dev.whyoleg.cryptography.bigint.encodeToByteArray.<anonymous>' call
    this_0[0] = positive ? 0 : -1;
    // Inline function 'kotlin.collections.copyInto' call
    var endIndex = bytes.length;
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    var tmp = bytes;
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    arrayCopy(tmp, this_0, 1, 0, endIndex);
    return this_0;
  }
  function Companion() {
    Companion_instance = this;
    this.j19_1 = new BigInt_0(BigInt(0));
  }
  protoOf(Companion).k19 = function () {
    return BigIntAsStringSerializer_getInstance();
  };
  var Companion_instance;
  function Companion_getInstance() {
    if (Companion_instance == null)
      new Companion();
    return Companion_instance;
  }
  function BigInt_0(jsBigInt) {
    Companion_getInstance();
    Number_0.call(this);
    this.h19_1 = jsBigInt;
  }
  protoOf(BigInt_0).i19 = function (other) {
    return this.l19(toBigInt(other));
  };
  protoOf(BigInt_0).l19 = function (other) {
    return jsBigIntCompareTo(this.h19_1, other.h19_1);
  };
  protoOf(BigInt_0).d = function (other) {
    return this.l19(other instanceof BigInt_0 ? other : THROW_CCE());
  };
  protoOf(BigInt_0).s3 = function () {
    return toByte(jsBigIntAsInt(this.h19_1, 8).toString());
  };
  protoOf(BigInt_0).t3 = function () {
    return toShort(jsBigIntAsInt(this.h19_1, 16).toString());
  };
  protoOf(BigInt_0).i1 = function () {
    return toInt(jsBigIntAsInt(this.h19_1, 32).toString());
  };
  protoOf(BigInt_0).m19 = function () {
    return toLong(jsBigIntAsInt(this.h19_1, 64).toString());
  };
  protoOf(BigInt_0).toString = function () {
    return this.h19_1.toString();
  };
  protoOf(BigInt_0).hashCode = function () {
    return getStringHashCode(this.h19_1.toString(36));
  };
  protoOf(BigInt_0).equals = function (other) {
    if (!(other instanceof BigInt_0))
      return false;
    return jsBigIntEquals(this.h19_1, other.h19_1);
  };
  function decodeToBigInt(_this__u8e3s4) {
    // Inline function 'kotlin.require' call
    // Inline function 'kotlin.collections.isNotEmpty' call
    // Inline function 'kotlin.collections.isEmpty' call
    // Inline function 'kotlin.contracts.contract' call
    if (!!(_this__u8e3s4.length === 0)) {
      // Inline function 'dev.whyoleg.cryptography.bigint.decodeToBigInt.<anonymous>' call
      var message = 'empty array';
      throw IllegalArgumentException_init_$Create$(toString(message));
    }
    if (_this__u8e3s4.length === 1 && _this__u8e3s4[0] === 0)
      return Companion_getInstance().j19_1;
    var positive = _this__u8e3s4[0] >= 0;
    var tmp;
    if (positive) {
      tmp = BigInt('0x' + toHexString(_this__u8e3s4));
    } else {
      // Inline function 'kotlin.collections.copyOf' call
      // Inline function 'kotlin.js.asDynamic' call
      var tmp$ret$4 = _this__u8e3s4.slice();
      tmp = jsBigIntNegate(BigInt('0x' + toHexString(invertTwoComplement(tmp$ret$4))));
    }
    var jsBigInt = tmp;
    return new BigInt_0(jsBigInt);
  }
  function toBigInt(_this__u8e3s4) {
    return _this__u8e3s4 === 0 ? Companion_getInstance().j19_1 : new BigInt_0(BigInt(_this__u8e3s4));
  }
  function toBigInt_0(_this__u8e3s4) {
    return toBigInt(_this__u8e3s4);
  }
  function toBigInt_1(_this__u8e3s4) {
    return toBigInt(_this__u8e3s4);
  }
  function toBigInt_2(_this__u8e3s4) {
    return _this__u8e3s4.equals(new Long(0, 0)) ? Companion_getInstance().j19_1 : new BigInt_0(BigInt(_this__u8e3s4.toString()));
  }
  function invertTwoComplement(_this__u8e3s4) {
    var tmp$ret$1;
    $l$block: {
      // Inline function 'kotlin.collections.indexOfLast' call
      var inductionVariable = _this__u8e3s4.length - 1 | 0;
      if (0 <= inductionVariable)
        do {
          var index = inductionVariable;
          inductionVariable = inductionVariable + -1 | 0;
          // Inline function 'dev.whyoleg.cryptography.bigint.invertTwoComplement.<anonymous>' call
          if (!(_this__u8e3s4[index] === 0)) {
            tmp$ret$1 = index;
            break $l$block;
          }
        }
         while (0 <= inductionVariable);
      tmp$ret$1 = -1;
    }
    var firstNonZeroFromEnd = tmp$ret$1;
    // Inline function 'kotlin.repeat' call
    var times = _this__u8e3s4.length;
    // Inline function 'kotlin.contracts.contract' call
    var inductionVariable_0 = 0;
    if (inductionVariable_0 < times)
      do {
        var index_0 = inductionVariable_0;
        inductionVariable_0 = inductionVariable_0 + 1 | 0;
        // Inline function 'dev.whyoleg.cryptography.bigint.invertTwoComplement.<anonymous>' call
        var negated = ~_this__u8e3s4[index_0];
        _this__u8e3s4[index_0] = toByte_0(index_0 < firstNonZeroFromEnd ? negated : negated + 1 | 0);
      }
       while (inductionVariable_0 < times);
    return _this__u8e3s4;
  }
  function toBigInt_3(_this__u8e3s4) {
    // Inline function 'kotlin.check' call
    // Inline function 'kotlin.text.isNotBlank' call
    // Inline function 'kotlin.contracts.contract' call
    if (!!isBlank(_this__u8e3s4)) {
      // Inline function 'dev.whyoleg.cryptography.bigint.toBigInt.<anonymous>' call
      var message = 'empty or blank string';
      throw IllegalStateException_init_$Create$(toString(message));
    }
    return new BigInt_0(BigInt(_this__u8e3s4));
  }
  function encodeToByteArray$decodeFromHex(_this__u8e3s4) {
    return hexToByteArray((_this__u8e3s4.length % 2 | 0) === 0 ? _this__u8e3s4 : '0' + _this__u8e3s4);
  }
  function jsBigIntNegate(value) {
    // Inline function 'kotlin.js.unsafeCast' call
    return -value;
  }
  function jsBigIntCompareTo(a, b) {
    // Inline function 'kotlin.js.unsafeCast' call
    return function () {
      if (a < b)
        return -1;
      if (a > b)
        return 1;
      return 0;
    }();
  }
  function jsBigIntAsInt(value, bits) {
    // Inline function 'kotlin.js.unsafeCast' call
    return BigInt.asIntN(bits, value);
  }
  function jsBigIntEquals(a, b) {
    // Inline function 'kotlin.js.unsafeCast' call
    return a === b;
  }
  //region block: exports
  _.$_$ = _.$_$ || {};
  _.$_$.a = BigInt_0;
  _.$_$.b = decodeToBigInt;
  _.$_$.c = encodeToByteArray;
  _.$_$.d = toBigInt;
  _.$_$.e = toBigInt_0;
  _.$_$.f = toBigInt_2;
  _.$_$.g = toBigInt_1;
  _.$_$.h = Companion_getInstance;
  //endregion
  return _;
}));
