// @sphereon/kmp-mdoc-core/cryptography-kotlin-cryptography-core.js
(function (factory) {
  if (typeof define === 'function' && define.amd)
    define(['exports', './kotlin-kotlin-stdlib.js'], factory);
  else if (typeof exports === 'object')
    factory(module.exports, require('./kotlin-kotlin-stdlib.js'));
  else {
    if (typeof globalThis['kotlin-kotlin-stdlib'] === 'undefined') {
      throw new Error("Error loading module 'cryptography-kotlin-cryptography-core'. Its dependency 'kotlin-kotlin-stdlib' was not found. Please, check whether 'kotlin-kotlin-stdlib' is loaded prior to 'cryptography-kotlin-cryptography-core'.");
    }
    globalThis['cryptography-kotlin-cryptography-core'] = factory(typeof globalThis['cryptography-kotlin-cryptography-core'] === 'undefined' ? {} : globalThis['cryptography-kotlin-cryptography-core'], globalThis['kotlin-kotlin-stdlib']);
  }
}(function (_, kotlin_kotlin) {
  'use strict';
  //region block: imports
  var imul = Math.imul;
  var toString = kotlin_kotlin.$_$.bb;
  var IllegalArgumentException_init_$Create$ = kotlin_kotlin.$_$.n1;
  var protoOf = kotlin_kotlin.$_$.xa;
  var initMetadataForCompanion = kotlin_kotlin.$_$.x9;
  var initMetadataForInterface = kotlin_kotlin.$_$.z9;
  var initMetadataForClass = kotlin_kotlin.$_$.w9;
  var firstOrNull = kotlin_kotlin.$_$.rb;
  var IllegalStateException_init_$Create$ = kotlin_kotlin.$_$.q1;
  var Unit_instance = kotlin_kotlin.$_$.q4;
  var toMutableList = kotlin_kotlin.$_$.b8;
  var toList = kotlin_kotlin.$_$.x7;
  var asSequence = kotlin_kotlin.$_$.i5;
  var map = kotlin_kotlin.$_$.sb;
  var initMetadataForObject = kotlin_kotlin.$_$.ba;
  var lazy = kotlin_kotlin.$_$.ef;
  var KProperty1 = kotlin_kotlin.$_$.pb;
  var getPropertyCallableRef = kotlin_kotlin.$_$.t9;
  var Enum = kotlin_kotlin.$_$.ud;
  var VOID = kotlin_kotlin.$_$.f;
  var THROW_CCE = kotlin_kotlin.$_$.ge;
  var closeFinally = kotlin_kotlin.$_$.ve;
  var emptyList = kotlin_kotlin.$_$.h6;
  //endregion
  //region block: pre-declaration
  initMetadataForCompanion(Companion);
  initMetadataForInterface(CryptographyAlgorithm, 'CryptographyAlgorithm');
  initMetadataForClass(CryptographyAlgorithmId, 'CryptographyAlgorithmId');
  initMetadataForObject(Registry, 'Registry');
  initMetadataForCompanion(Companion_0);
  initMetadataForClass(CryptographyProvider, 'CryptographyProvider');
  initMetadataForClass(Format, 'Format', VOID, Enum);
  initMetadataForCompanion(Companion_1, CryptographyAlgorithmId);
  initMetadataForCompanion(Companion_2, CryptographyAlgorithmId);
  initMetadataForCompanion(Companion_3, CryptographyAlgorithmId);
  function encryptWithIv(iv, plaintext, associatedData, $completion) {
    return this.m1a(iv, plaintext, associatedData);
  }
  initMetadataForInterface(IvAuthenticatedEncryptor, 'IvAuthenticatedEncryptor', VOID, VOID, VOID, [2, 3, 1]);
  function decryptWithIv(iv, ciphertext, associatedData, $completion) {
    return this.o1a(iv, ciphertext, associatedData);
  }
  initMetadataForInterface(IvAuthenticatedDecryptor, 'IvAuthenticatedDecryptor', VOID, VOID, VOID, [2, 3, 1]);
  initMetadataForObject(SHA1, 'SHA1', VOID, CryptographyAlgorithmId);
  initMetadataForObject(SHA256, 'SHA256', VOID, CryptographyAlgorithmId);
  initMetadataForObject(SHA384, 'SHA384', VOID, CryptographyAlgorithmId);
  initMetadataForObject(SHA512, 'SHA512', VOID, CryptographyAlgorithmId);
  initMetadataForCompanion(Companion_4);
  initMetadataForClass(Format_0, 'Format');
  initMetadataForObject(JWK, 'JWK', VOID, Format_0);
  initMetadataForObject(RAW, 'RAW', VOID, Format_0);
  initMetadataForObject(DER, 'DER', VOID, Format_0);
  initMetadataForObject(PEM, 'PEM', VOID, Format_0);
  initMetadataForClass(Format_1, 'Format');
  initMetadataForClass(DER_0, 'DER', VOID, Format_1);
  initMetadataForObject(Generic, 'Generic', VOID, DER_0);
  initMetadataForObject(SEC1, 'SEC1', VOID, DER_0);
  initMetadataForClass(PEM_0, 'PEM', VOID, Format_1);
  initMetadataForObject(Generic_0, 'Generic', VOID, PEM_0);
  initMetadataForObject(SEC1_0, 'SEC1', VOID, PEM_0);
  initMetadataForObject(JWK_0, 'JWK', VOID, Format_1);
  initMetadataForObject(RAW_0, 'RAW', VOID, Format_1);
  initMetadataForCompanion(Companion_5, CryptographyAlgorithmId);
  initMetadataForInterface(PublicKey, 'PublicKey', VOID, VOID, VOID, [1]);
  initMetadataForInterface(PrivateKey, 'PrivateKey', VOID, VOID, VOID, [1]);
  initMetadataForCompanion(Companion_6, CryptographyAlgorithmId);
  initMetadataForCompanion(Companion_7, CryptographyAlgorithmId);
  initMetadataForClass(Format_2, 'Format', VOID, Enum);
  initMetadataForCompanion(Companion_8, CryptographyAlgorithmId);
  initMetadataForCompanion(Companion_9, CryptographyAlgorithmId);
  initMetadataForClass(Format_3, 'Format');
  initMetadataForClass(DER_1, 'DER', VOID, Format_3);
  initMetadataForObject(Generic_1, 'Generic', VOID, DER_1);
  initMetadataForObject(PKCS1, 'PKCS1', VOID, DER_1);
  initMetadataForClass(PEM_1, 'PEM', VOID, Format_3);
  initMetadataForObject(Generic_2, 'Generic', VOID, PEM_1);
  initMetadataForObject(PKCS1_0, 'PKCS1', VOID, PEM_1);
  initMetadataForObject(JWK_1, 'JWK', VOID, Format_3);
  initMetadataForClass(Format_4, 'Format');
  initMetadataForClass(DER_2, 'DER', VOID, Format_4);
  initMetadataForObject(Generic_3, 'Generic', VOID, DER_2);
  initMetadataForObject(PKCS1_1, 'PKCS1', VOID, DER_2);
  initMetadataForClass(PEM_2, 'PEM', VOID, Format_4);
  initMetadataForObject(Generic_4, 'Generic', VOID, PEM_2);
  initMetadataForObject(PKCS1_2, 'PKCS1', VOID, PEM_2);
  initMetadataForObject(JWK_2, 'JWK', VOID, Format_4);
  initMetadataForCompanion(Companion_10, CryptographyAlgorithmId);
  initMetadataForCompanion(Companion_11, CryptographyAlgorithmId);
  initMetadataForCompanion(Companion_12, CryptographyAlgorithmId);
  function decodeFromByteArray(format, bytes, $completion) {
    return this.u1a(format, bytes);
  }
  initMetadataForInterface(KeyDecoder, 'KeyDecoder', VOID, VOID, VOID, [2]);
  function deriveSecretToByteArray(input, $completion) {
    return this.w1a(input);
  }
  initMetadataForInterface(SecretDerivation, 'SecretDerivation', VOID, VOID, VOID, [1]);
  function generateSharedSecretToByteArray(other, $completion) {
    return this.y1a(other);
  }
  initMetadataForInterface(SharedSecretGenerator, 'SharedSecretGenerator', VOID, VOID, VOID, [1]);
  function generateSignature(data, $completion) {
    return this.b1b(data);
  }
  function generateSignatureBlocking(data) {
    var tmp$ret$1;
    $l$block: {
      // Inline function 'kotlin.use' call
      var this_0 = this.z1a();
      // Inline function 'kotlin.contracts.contract' call
      var exception = null;
      try {
        // Inline function 'dev.whyoleg.cryptography.operations.SignatureGenerator.generateSignatureBlocking.<anonymous>' call
        this_0.c1b(data);
        tmp$ret$1 = this_0.d1b();
        break $l$block;
      } catch ($p) {
        if ($p instanceof Error) {
          var e = $p;
          exception = e;
          throw e;
        } else {
          throw $p;
        }
      }
      finally {
        closeFinally(this_0, exception);
      }
    }
    return tmp$ret$1;
  }
  initMetadataForInterface(SignatureGenerator, 'SignatureGenerator', VOID, VOID, VOID, [1]);
  function update$default(source, startIndex, endIndex, $super) {
    startIndex = startIndex === VOID ? 0 : startIndex;
    endIndex = endIndex === VOID ? source.length : endIndex;
    var tmp;
    if ($super === VOID) {
      this.e1b(source, startIndex, endIndex);
      tmp = Unit_instance;
    } else {
      tmp = $super.e1b.call(this, source, startIndex, endIndex);
    }
    return tmp;
  }
  initMetadataForInterface(UpdateFunction, 'UpdateFunction');
  //endregion
  function _BinarySize___init__impl__j238f5(bits) {
    // Inline function 'kotlin.require' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(_get_bits__d4bv2d(bits) >= 0)) {
      // Inline function 'dev.whyoleg.cryptography.BinarySize.<anonymous>' call
      var message = 'Value must be greater than or equal to 0';
      throw IllegalArgumentException_init_$Create$(toString(message));
    }
    // Inline function 'kotlin.require' call
    // Inline function 'kotlin.contracts.contract' call
    if (!((_get_bits__d4bv2d(bits) % 8 | 0) === 0)) {
      // Inline function 'dev.whyoleg.cryptography.BinarySize.<anonymous>' call
      var message_0 = 'Value must be a multiple of 8';
      throw IllegalArgumentException_init_$Create$(toString(message_0));
    }
    return bits;
  }
  function _get_bits__d4bv2d($this) {
    return $this;
  }
  function _BinarySize___get_inBits__impl__p3mjjl($this) {
    return _get_bits__d4bv2d($this);
  }
  function Companion() {
  }
  protoOf(Companion).b1a = function (_this__u8e3s4) {
    return _BinarySize___init__impl__j238f5(_this__u8e3s4);
  };
  protoOf(Companion).c1a = function (_this__u8e3s4) {
    return _BinarySize___init__impl__j238f5(imul(_this__u8e3s4, 8));
  };
  var Companion_instance;
  function Companion_getInstance() {
    return Companion_instance;
  }
  function CryptographyAlgorithm() {
  }
  function CryptographyAlgorithmId(name) {
    this.d1a_1 = name;
  }
  function CryptographyProvider$Companion$Default$delegate$lambda() {
    var tmp$ret$1;
    $l$block: {
      // Inline function 'kotlin.checkNotNull' call
      var value = firstOrNull(Registry_getInstance().f1a());
      // Inline function 'kotlin.contracts.contract' call
      if (value == null) {
        // Inline function 'dev.whyoleg.cryptography.Companion.Default$delegate.<anonymous>.<anonymous>' call
        var message = 'No providers registered. Please provide a dependency or register provider explicitly';
        throw IllegalStateException_init_$Create$(toString(message));
      } else {
        tmp$ret$1 = value;
        break $l$block;
      }
    }
    return tmp$ret$1;
  }
  function Registry() {
    Registry_instance = this;
    this.e1a_1 = toMutableList(initProviders());
  }
  protoOf(Registry).f1a = function () {
    return map(asSequence(toList(this.e1a_1)), value$factory());
  };
  protoOf(Registry).g1a = function (provider) {
    this.e1a_1.l(provider);
  };
  var Registry_instance;
  function Registry_getInstance() {
    if (Registry_instance == null)
      new Registry();
    return Registry_instance;
  }
  function Companion_0() {
    Companion_instance_0 = this;
    var tmp = this;
    tmp.h1a_1 = lazy(CryptographyProvider$Companion$Default$delegate$lambda);
  }
  protoOf(Companion_0).i1a = function () {
    // Inline function 'kotlin.getValue' call
    var this_0 = this.h1a_1;
    Default$factory();
    return this_0.d2();
  };
  var Companion_instance_0;
  function Companion_getInstance_0() {
    if (Companion_instance_0 == null)
      new Companion_0();
    return Companion_instance_0;
  }
  function CryptographyProvider() {
    Companion_getInstance_0();
  }
  protoOf(CryptographyProvider).k1a = function (identifier) {
    var tmp0_elvis_lhs = this.j1a(identifier);
    var tmp;
    if (tmp0_elvis_lhs == null) {
      throw IllegalStateException_init_$Create$('Algorithm not found: ' + toString(identifier));
    } else {
      tmp = tmp0_elvis_lhs;
    }
    return tmp;
  };
  function value$factory() {
    return getPropertyCallableRef('value', 1, KProperty1, function (receiver) {
      return receiver.d2();
    }, null);
  }
  function Default$factory() {
    return getPropertyCallableRef('Default', 1, KProperty1, function (receiver) {
      return receiver.i1a();
    }, null);
  }
  var Format_RAW_instance;
  var Format_JWK_instance;
  var Format_entriesInitialized;
  function Format_initEntries() {
    if (Format_entriesInitialized)
      return Unit_instance;
    Format_entriesInitialized = true;
    Format_RAW_instance = new Format('RAW', 0);
    Format_JWK_instance = new Format('JWK', 1);
  }
  function Format(name, ordinal) {
    Enum.call(this, name, ordinal);
  }
  function Format_RAW_getInstance() {
    Format_initEntries();
    return Format_RAW_instance;
  }
  function Companion_1() {
    Companion_instance_1 = this;
    CryptographyAlgorithmId.call(this, 'AES-CBC');
  }
  var Companion_instance_1;
  function Companion_getInstance_1() {
    if (Companion_instance_1 == null)
      new Companion_1();
    return Companion_instance_1;
  }
  function Companion_2() {
    Companion_instance_2 = this;
    CryptographyAlgorithmId.call(this, 'AES-CTR');
  }
  var Companion_instance_2;
  function Companion_getInstance_2() {
    if (Companion_instance_2 == null)
      new Companion_2();
    return Companion_instance_2;
  }
  function Companion_3() {
    Companion_instance_3 = this;
    CryptographyAlgorithmId.call(this, 'AES-GCM');
  }
  var Companion_instance_3;
  function Companion_getInstance_3() {
    if (Companion_instance_3 == null)
      new Companion_3();
    return Companion_instance_3;
  }
  function IvAuthenticatedEncryptor() {
  }
  function IvAuthenticatedDecryptor() {
  }
  function SHA1() {
    SHA1_instance = this;
    CryptographyAlgorithmId.call(this, 'SHA-1');
  }
  var SHA1_instance;
  function SHA1_getInstance() {
    if (SHA1_instance == null)
      new SHA1();
    return SHA1_instance;
  }
  function SHA256() {
    SHA256_instance = this;
    CryptographyAlgorithmId.call(this, 'SHA-256');
  }
  var SHA256_instance;
  function SHA256_getInstance() {
    if (SHA256_instance == null)
      new SHA256();
    return SHA256_instance;
  }
  function SHA384() {
    SHA384_instance = this;
    CryptographyAlgorithmId.call(this, 'SHA-384');
  }
  var SHA384_instance;
  function SHA384_getInstance() {
    if (SHA384_instance == null)
      new SHA384();
    return SHA384_instance;
  }
  function SHA512() {
    SHA512_instance = this;
    CryptographyAlgorithmId.call(this, 'SHA-512');
  }
  var SHA512_instance;
  function SHA512_getInstance() {
    if (SHA512_instance == null)
      new SHA512();
    return SHA512_instance;
  }
  function _Curve___init__impl__ygix12(name) {
    return name;
  }
  function _Curve___get_name__impl__klrusc($this) {
    return $this;
  }
  function Companion_4() {
  }
  protoOf(Companion_4).p1a = function () {
    return _Curve___init__impl__ygix12('P-256');
  };
  protoOf(Companion_4).q1a = function () {
    return _Curve___init__impl__ygix12('P-384');
  };
  protoOf(Companion_4).r1a = function () {
    return _Curve___init__impl__ygix12('P-521');
  };
  var Companion_instance_4;
  function Companion_getInstance_4() {
    return Companion_instance_4;
  }
  function JWK() {
    JWK_instance = this;
    Format_0.call(this);
  }
  protoOf(JWK).w2 = function () {
    return 'JWK';
  };
  protoOf(JWK).hashCode = function () {
    return -164862489;
  };
  protoOf(JWK).equals = function (other) {
    if (this === other)
      return true;
    if (!(other instanceof JWK))
      return false;
    other instanceof JWK || THROW_CCE();
    return true;
  };
  var JWK_instance;
  function JWK_getInstance() {
    if (JWK_instance == null)
      new JWK();
    return JWK_instance;
  }
  function RAW() {
    RAW_instance = this;
    Format_0.call(this);
  }
  protoOf(RAW).w2 = function () {
    return 'RAW';
  };
  protoOf(RAW).hashCode = function () {
    return -164855471;
  };
  protoOf(RAW).equals = function (other) {
    if (this === other)
      return true;
    if (!(other instanceof RAW))
      return false;
    other instanceof RAW || THROW_CCE();
    return true;
  };
  var RAW_instance;
  function RAW_getInstance() {
    if (RAW_instance == null)
      new RAW();
    return RAW_instance;
  }
  function DER() {
    DER_instance = this;
    Format_0.call(this);
  }
  protoOf(DER).w2 = function () {
    return 'DER';
  };
  protoOf(DER).hashCode = function () {
    return -164868806;
  };
  protoOf(DER).equals = function (other) {
    if (this === other)
      return true;
    if (!(other instanceof DER))
      return false;
    other instanceof DER || THROW_CCE();
    return true;
  };
  var DER_instance;
  function DER_getInstance() {
    if (DER_instance == null)
      new DER();
    return DER_instance;
  }
  function PEM() {
    PEM_instance = this;
    Format_0.call(this);
  }
  protoOf(PEM).w2 = function () {
    return 'PEM';
  };
  protoOf(PEM).hashCode = function () {
    return -164857279;
  };
  protoOf(PEM).equals = function (other) {
    if (this === other)
      return true;
    if (!(other instanceof PEM))
      return false;
    other instanceof PEM || THROW_CCE();
    return true;
  };
  var PEM_instance;
  function PEM_getInstance() {
    if (PEM_instance == null)
      new PEM();
    return PEM_instance;
  }
  function Format_0() {
  }
  protoOf(Format_0).toString = function () {
    return this.w2();
  };
  function Generic() {
    Generic_instance = this;
    DER_0.call(this);
  }
  protoOf(Generic).w2 = function () {
    return 'DER';
  };
  var Generic_instance;
  function Generic_getInstance() {
    if (Generic_instance == null)
      new Generic();
    return Generic_instance;
  }
  function SEC1() {
    SEC1_instance = this;
    DER_0.call(this);
  }
  protoOf(SEC1).w2 = function () {
    return 'DER/SEC1';
  };
  protoOf(SEC1).hashCode = function () {
    return -137777718;
  };
  protoOf(SEC1).equals = function (other) {
    if (this === other)
      return true;
    if (!(other instanceof SEC1))
      return false;
    other instanceof SEC1 || THROW_CCE();
    return true;
  };
  var SEC1_instance;
  function SEC1_getInstance() {
    if (SEC1_instance == null)
      new SEC1();
    return SEC1_instance;
  }
  function Generic_0() {
    Generic_instance_0 = this;
    PEM_0.call(this);
  }
  protoOf(Generic_0).w2 = function () {
    return 'PEM';
  };
  var Generic_instance_0;
  function Generic_getInstance_0() {
    if (Generic_instance_0 == null)
      new Generic_0();
    return Generic_instance_0;
  }
  function SEC1_0() {
    SEC1_instance_0 = this;
    PEM_0.call(this);
  }
  protoOf(SEC1_0).w2 = function () {
    return 'PEM/SEC1';
  };
  protoOf(SEC1_0).hashCode = function () {
    return -842035933;
  };
  protoOf(SEC1_0).equals = function (other) {
    if (this === other)
      return true;
    if (!(other instanceof SEC1_0))
      return false;
    other instanceof SEC1_0 || THROW_CCE();
    return true;
  };
  var SEC1_instance_0;
  function SEC1_getInstance_0() {
    if (SEC1_instance_0 == null)
      new SEC1_0();
    return SEC1_instance_0;
  }
  function JWK_0() {
    JWK_instance_0 = this;
    Format_1.call(this);
  }
  protoOf(JWK_0).w2 = function () {
    return 'JWK';
  };
  protoOf(JWK_0).hashCode = function () {
    return 1029262513;
  };
  protoOf(JWK_0).equals = function (other) {
    if (this === other)
      return true;
    if (!(other instanceof JWK_0))
      return false;
    other instanceof JWK_0 || THROW_CCE();
    return true;
  };
  var JWK_instance_0;
  function JWK_getInstance_0() {
    if (JWK_instance_0 == null)
      new JWK_0();
    return JWK_instance_0;
  }
  function RAW_0() {
    RAW_instance_0 = this;
    Format_1.call(this);
  }
  protoOf(RAW_0).w2 = function () {
    return 'RAW';
  };
  protoOf(RAW_0).hashCode = function () {
    return 1029269531;
  };
  protoOf(RAW_0).equals = function (other) {
    if (this === other)
      return true;
    if (!(other instanceof RAW_0))
      return false;
    other instanceof RAW_0 || THROW_CCE();
    return true;
  };
  var RAW_instance_0;
  function RAW_getInstance_0() {
    if (RAW_instance_0 == null)
      new RAW_0();
    return RAW_instance_0;
  }
  function DER_0() {
    Generic_getInstance();
    Format_1.call(this);
  }
  function PEM_0() {
    Generic_getInstance_0();
    Format_1.call(this);
  }
  function Format_1() {
  }
  protoOf(Format_1).toString = function () {
    return this.w2();
  };
  function Companion_5() {
    Companion_instance_5 = this;
    CryptographyAlgorithmId.call(this, 'ECDH');
  }
  var Companion_instance_5;
  function Companion_getInstance_5() {
    if (Companion_instance_5 == null)
      new Companion_5();
    return Companion_instance_5;
  }
  function PublicKey() {
  }
  function PrivateKey() {
  }
  function Companion_6() {
    Companion_instance_6 = this;
    CryptographyAlgorithmId.call(this, 'ECDSA');
  }
  var Companion_instance_6;
  function Companion_getInstance_6() {
    if (Companion_instance_6 == null)
      new Companion_6();
    return Companion_instance_6;
  }
  function Companion_7() {
    Companion_instance_7 = this;
    CryptographyAlgorithmId.call(this, 'HKDF');
  }
  var Companion_instance_7;
  function Companion_getInstance_7() {
    if (Companion_instance_7 == null)
      new Companion_7();
    return Companion_instance_7;
  }
  var Format_RAW_instance_0;
  var Format_JWK_instance_0;
  var Format_entriesInitialized_0;
  function Format_initEntries_0() {
    if (Format_entriesInitialized_0)
      return Unit_instance;
    Format_entriesInitialized_0 = true;
    Format_RAW_instance_0 = new Format_2('RAW', 0);
    Format_JWK_instance_0 = new Format_2('JWK', 1);
  }
  function Format_2(name, ordinal) {
    Enum.call(this, name, ordinal);
  }
  function Format_RAW_getInstance_0() {
    Format_initEntries_0();
    return Format_RAW_instance_0;
  }
  function Companion_8() {
    Companion_instance_8 = this;
    CryptographyAlgorithmId.call(this, 'HMAC');
  }
  var Companion_instance_8;
  function Companion_getInstance_8() {
    if (Companion_instance_8 == null)
      new Companion_8();
    return Companion_instance_8;
  }
  function Companion_9() {
    Companion_instance_9 = this;
    CryptographyAlgorithmId.call(this, 'PBKDF2');
  }
  var Companion_instance_9;
  function Companion_getInstance_9() {
    if (Companion_instance_9 == null)
      new Companion_9();
    return Companion_instance_9;
  }
  function Generic_1() {
    Generic_instance_1 = this;
    DER_1.call(this);
  }
  protoOf(Generic_1).w2 = function () {
    return 'DER';
  };
  var Generic_instance_1;
  function Generic_getInstance_1() {
    if (Generic_instance_1 == null)
      new Generic_1();
    return Generic_instance_1;
  }
  function PKCS1() {
    PKCS1_instance = this;
    DER_1.call(this);
  }
  protoOf(PKCS1).w2 = function () {
    return 'DER/PKCS#1';
  };
  protoOf(PKCS1).hashCode = function () {
    return 1950278528;
  };
  protoOf(PKCS1).equals = function (other) {
    if (this === other)
      return true;
    if (!(other instanceof PKCS1))
      return false;
    other instanceof PKCS1 || THROW_CCE();
    return true;
  };
  var PKCS1_instance;
  function PKCS1_getInstance() {
    if (PKCS1_instance == null)
      new PKCS1();
    return PKCS1_instance;
  }
  function Generic_2() {
    Generic_instance_2 = this;
    PEM_1.call(this);
  }
  protoOf(Generic_2).w2 = function () {
    return 'PEM';
  };
  var Generic_instance_2;
  function Generic_getInstance_2() {
    if (Generic_instance_2 == null)
      new Generic_2();
    return Generic_instance_2;
  }
  function PKCS1_0() {
    PKCS1_instance_0 = this;
    PEM_1.call(this);
  }
  protoOf(PKCS1_0).w2 = function () {
    return 'PEM/PKCS#1';
  };
  protoOf(PKCS1_0).hashCode = function () {
    return 1593110343;
  };
  protoOf(PKCS1_0).equals = function (other) {
    if (this === other)
      return true;
    if (!(other instanceof PKCS1_0))
      return false;
    other instanceof PKCS1_0 || THROW_CCE();
    return true;
  };
  var PKCS1_instance_0;
  function PKCS1_getInstance_0() {
    if (PKCS1_instance_0 == null)
      new PKCS1_0();
    return PKCS1_instance_0;
  }
  function JWK_1() {
    JWK_instance_1 = this;
    Format_3.call(this);
  }
  protoOf(JWK_1).w2 = function () {
    return 'JWK';
  };
  protoOf(JWK_1).hashCode = function () {
    return -1819476043;
  };
  protoOf(JWK_1).equals = function (other) {
    if (this === other)
      return true;
    if (!(other instanceof JWK_1))
      return false;
    other instanceof JWK_1 || THROW_CCE();
    return true;
  };
  var JWK_instance_1;
  function JWK_getInstance_1() {
    if (JWK_instance_1 == null)
      new JWK_1();
    return JWK_instance_1;
  }
  function DER_1() {
    Generic_getInstance_1();
    Format_3.call(this);
  }
  function PEM_1() {
    Generic_getInstance_2();
    Format_3.call(this);
  }
  function Format_3() {
  }
  protoOf(Format_3).toString = function () {
    return this.w2();
  };
  function Generic_3() {
    Generic_instance_3 = this;
    DER_2.call(this);
  }
  protoOf(Generic_3).w2 = function () {
    return 'DER';
  };
  var Generic_instance_3;
  function Generic_getInstance_3() {
    if (Generic_instance_3 == null)
      new Generic_3();
    return Generic_instance_3;
  }
  function PKCS1_1() {
    PKCS1_instance_1 = this;
    DER_2.call(this);
  }
  protoOf(PKCS1_1).w2 = function () {
    return 'DER/PKCS#1';
  };
  protoOf(PKCS1_1).hashCode = function () {
    return -1024989714;
  };
  protoOf(PKCS1_1).equals = function (other) {
    if (this === other)
      return true;
    if (!(other instanceof PKCS1_1))
      return false;
    other instanceof PKCS1_1 || THROW_CCE();
    return true;
  };
  var PKCS1_instance_1;
  function PKCS1_getInstance_1() {
    if (PKCS1_instance_1 == null)
      new PKCS1_1();
    return PKCS1_instance_1;
  }
  function Generic_4() {
    Generic_instance_4 = this;
    PEM_2.call(this);
  }
  protoOf(Generic_4).w2 = function () {
    return 'PEM';
  };
  var Generic_instance_4;
  function Generic_getInstance_4() {
    if (Generic_instance_4 == null)
      new Generic_4();
    return Generic_instance_4;
  }
  function PKCS1_2() {
    PKCS1_instance_2 = this;
    PEM_2.call(this);
  }
  protoOf(PKCS1_2).w2 = function () {
    return 'PEM/PKCS#1';
  };
  protoOf(PKCS1_2).hashCode = function () {
    return -1382157899;
  };
  protoOf(PKCS1_2).equals = function (other) {
    if (this === other)
      return true;
    if (!(other instanceof PKCS1_2))
      return false;
    other instanceof PKCS1_2 || THROW_CCE();
    return true;
  };
  var PKCS1_instance_2;
  function PKCS1_getInstance_2() {
    if (PKCS1_instance_2 == null)
      new PKCS1_2();
    return PKCS1_instance_2;
  }
  function JWK_2() {
    JWK_instance_2 = this;
    Format_4.call(this);
  }
  protoOf(JWK_2).w2 = function () {
    return 'JWK';
  };
  protoOf(JWK_2).hashCode = function () {
    return 1275849891;
  };
  protoOf(JWK_2).equals = function (other) {
    if (this === other)
      return true;
    if (!(other instanceof JWK_2))
      return false;
    other instanceof JWK_2 || THROW_CCE();
    return true;
  };
  var JWK_instance_2;
  function JWK_getInstance_2() {
    if (JWK_instance_2 == null)
      new JWK_2();
    return JWK_instance_2;
  }
  function DER_2() {
    Generic_getInstance_3();
    Format_4.call(this);
  }
  function PEM_2() {
    Generic_getInstance_4();
    Format_4.call(this);
  }
  function Format_4() {
  }
  protoOf(Format_4).toString = function () {
    return this.w2();
  };
  function Companion_10() {
    Companion_instance_10 = this;
    CryptographyAlgorithmId.call(this, 'RSA-OAEP');
  }
  var Companion_instance_10;
  function Companion_getInstance_10() {
    if (Companion_instance_10 == null)
      new Companion_10();
    return Companion_instance_10;
  }
  function Companion_11() {
    Companion_instance_11 = this;
    CryptographyAlgorithmId.call(this, 'RSA-PSS');
  }
  var Companion_instance_11;
  function Companion_getInstance_11() {
    if (Companion_instance_11 == null)
      new Companion_11();
    return Companion_instance_11;
  }
  function Companion_12() {
    Companion_instance_12 = this;
    CryptographyAlgorithmId.call(this, 'RSA-PKCS1-V1.5');
  }
  var Companion_instance_12;
  function Companion_getInstance_12() {
    if (Companion_instance_12 == null)
      new Companion_12();
    return Companion_instance_12;
  }
  function KeyDecoder() {
  }
  function SecretDerivation() {
  }
  function SharedSecretGenerator() {
  }
  function SignatureGenerator() {
  }
  function UpdateFunction() {
  }
  function initProviders() {
    return emptyList();
  }
  //region block: init
  Companion_instance = new Companion();
  Companion_instance_4 = new Companion_4();
  //endregion
  //region block: exports
  _.$_$ = _.$_$ || {};
  _.$_$.a = IvAuthenticatedDecryptor;
  _.$_$.b = IvAuthenticatedEncryptor;
  _.$_$.c = Format;
  _.$_$.d = Format_1;
  _.$_$.e = Format_0;
  _.$_$.f = PrivateKey;
  _.$_$.g = PublicKey;
  _.$_$.h = Format_2;
  _.$_$.i = Format_4;
  _.$_$.j = Format_3;
  _.$_$.k = KeyDecoder;
  _.$_$.l = SecretDerivation;
  _.$_$.m = SharedSecretGenerator;
  _.$_$.n = SignatureGenerator;
  _.$_$.o = CryptographyAlgorithm;
  _.$_$.p = CryptographyProvider;
  _.$_$.q = Format_RAW_getInstance;
  _.$_$.r = Format_RAW_getInstance_0;
  _.$_$.s = _Curve___get_name__impl__klrusc;
  _.$_$.t = _BinarySize___get_inBits__impl__p3mjjl;
  _.$_$.u = Companion_getInstance_1;
  _.$_$.v = Companion_getInstance_2;
  _.$_$.w = Companion_getInstance_3;
  _.$_$.x = Companion_instance_4;
  _.$_$.y = Generic_getInstance;
  _.$_$.z = SEC1_getInstance;
  _.$_$.a1 = JWK_getInstance_0;
  _.$_$.b1 = Generic_getInstance_0;
  _.$_$.c1 = SEC1_getInstance_0;
  _.$_$.d1 = RAW_getInstance_0;
  _.$_$.e1 = DER_getInstance;
  _.$_$.f1 = JWK_getInstance;
  _.$_$.g1 = PEM_getInstance;
  _.$_$.h1 = RAW_getInstance;
  _.$_$.i1 = Companion_getInstance_5;
  _.$_$.j1 = Companion_getInstance_6;
  _.$_$.k1 = Companion_getInstance_7;
  _.$_$.l1 = Companion_getInstance_8;
  _.$_$.m1 = Companion_getInstance_9;
  _.$_$.n1 = Companion_getInstance_10;
  _.$_$.o1 = Companion_getInstance_12;
  _.$_$.p1 = Companion_getInstance_11;
  _.$_$.q1 = Generic_getInstance_3;
  _.$_$.r1 = PKCS1_getInstance_1;
  _.$_$.s1 = JWK_getInstance_2;
  _.$_$.t1 = Generic_getInstance_4;
  _.$_$.u1 = PKCS1_getInstance_2;
  _.$_$.v1 = Generic_getInstance_1;
  _.$_$.w1 = PKCS1_getInstance;
  _.$_$.x1 = JWK_getInstance_1;
  _.$_$.y1 = Generic_getInstance_2;
  _.$_$.z1 = PKCS1_getInstance_0;
  _.$_$.a2 = SHA1_getInstance;
  _.$_$.b2 = SHA256_getInstance;
  _.$_$.c2 = SHA384_getInstance;
  _.$_$.d2 = SHA512_getInstance;
  _.$_$.e2 = Companion_instance;
  _.$_$.f2 = Companion_getInstance_0;
  _.$_$.g2 = Registry_getInstance;
  //endregion
  return _;
}));
