// @sphereon/kmp-mdoc-core/hash-library-sha2.js
(function (factory) {
  if (typeof define === 'function' && define.amd)
    define(['exports', './core-library-digest.js', './kotlin-kotlin-stdlib.js'], factory);
  else if (typeof exports === 'object')
    factory(module.exports, require('./core-library-digest.js'), require('./kotlin-kotlin-stdlib.js'));
  else {
    if (typeof globalThis['core-library-digest'] === 'undefined') {
      throw new Error("Error loading module 'hash-library-sha2'. Its dependency 'core-library-digest' was not found. Please, check whether 'core-library-digest' is loaded prior to 'hash-library-sha2'.");
    }
    if (typeof globalThis['kotlin-kotlin-stdlib'] === 'undefined') {
      throw new Error("Error loading module 'hash-library-sha2'. Its dependency 'kotlin-kotlin-stdlib' was not found. Please, check whether 'kotlin-kotlin-stdlib' is loaded prior to 'hash-library-sha2'.");
    }
    globalThis['hash-library-sha2'] = factory(typeof globalThis['hash-library-sha2'] === 'undefined' ? {} : globalThis['hash-library-sha2'], globalThis['core-library-digest'], globalThis['kotlin-kotlin-stdlib']);
  }
}(function (_, kotlin_org_kotlincrypto_core_digest, kotlin_kotlin) {
  'use strict';
  //region block: imports
  var Digest_init_$Init$ = kotlin_org_kotlincrypto_core_digest.$_$.a;
  var protoOf = kotlin_kotlin.$_$.xa;
  var initMetadataForCompanion = kotlin_kotlin.$_$.x9;
  var Digest = kotlin_org_kotlincrypto_core_digest.$_$.b;
  var Unit_instance = kotlin_kotlin.$_$.q4;
  var fill = kotlin_kotlin.$_$.m6;
  var fill_0 = kotlin_kotlin.$_$.k6;
  var initMetadataForClass = kotlin_kotlin.$_$.w9;
  var VOID = kotlin_kotlin.$_$.f;
  var toString = kotlin_kotlin.$_$.bb;
  var IllegalArgumentException_init_$Create$ = kotlin_kotlin.$_$.n1;
  var longArray = kotlin_kotlin.$_$.qa;
  var longArrayOf = kotlin_kotlin.$_$.pa;
  var Long = kotlin_kotlin.$_$.ae;
  var toLong = kotlin_kotlin.$_$.za;
  var fill_1 = kotlin_kotlin.$_$.l6;
  var objectCreate = kotlin_kotlin.$_$.wa;
  var toByte = kotlin_kotlin.$_$.ya;
  //endregion
  //region block: pre-declaration
  initMetadataForCompanion(Companion);
  initMetadataForClass(Bit32Digest, 'Bit32Digest', VOID, Digest);
  initMetadataForCompanion(Companion_0);
  initMetadataForClass(Bit64Digest, 'Bit64Digest', VOID, Digest);
  initMetadataForClass(SHA256, 'SHA256', SHA256_init_$Create$, Bit32Digest);
  initMetadataForClass(SHA384, 'SHA384', SHA384_init_$Create$, Bit64Digest);
  initMetadataForClass(SHA512, 'SHA512', SHA512_init_$Create$, Bit64Digest);
  //endregion
  function Bit32Digest_init_$Init$(d, h0, h1, h2, h3, h4, h5, h6, h7, $this) {
    Digest_init_$Init$('SHA-' + d, 64, d / 8 | 0, $this);
    Bit32Digest.call($this);
    $this.e1n_1 = h0;
    $this.f1n_1 = h1;
    $this.g1n_1 = h2;
    $this.h1n_1 = h3;
    $this.i1n_1 = h4;
    $this.j1n_1 = h5;
    $this.k1n_1 = h6;
    $this.l1n_1 = h7;
    $this.m1n_1 = new Int32Array(64);
    var tmp = $this;
    // Inline function 'kotlin.intArrayOf' call
    tmp.n1n_1 = new Int32Array([h0, h1, h2, h3, h4, h5, h6, h7]);
    return $this;
  }
  function Companion() {
    Companion_instance = this;
    var tmp = this;
    // Inline function 'kotlin.intArrayOf' call
    tmp.o1n_1 = new Int32Array([1116352408, 1899447441, -1245643825, -373957723, 961987163, 1508970993, -1841331548, -1424204075, -670586216, 310598401, 607225278, 1426881987, 1925078388, -2132889090, -1680079193, -1046744716, -459576895, -272742522, 264347078, 604807628, 770255983, 1249150122, 1555081692, 1996064986, -1740746414, -1473132947, -1341970488, -1084653625, -958395405, -710438585, 113926993, 338241895, 666307205, 773529912, 1294757372, 1396182291, 1695183700, 1986661051, -2117940946, -1838011259, -1564481375, -1474664885, -1035236496, -949202525, -778901479, -694614492, -200395387, 275423344, 430227734, 506948616, 659060556, 883997877, 958139571, 1322822218, 1537002063, 1747873779, 1955562222, 2024104815, -2067236844, -1933114872, -1866530822, -1538233109, -1090935817, -965641998]);
  }
  var Companion_instance;
  function Companion_getInstance() {
    if (Companion_instance == null)
      new Companion();
    return Companion_instance;
  }
  protoOf(Bit32Digest).w1m = function (input, offset) {
    var x = this.m1n_1;
    var j = offset;
    var inductionVariable = 0;
    if (inductionVariable < 16)
      do {
        var i = inductionVariable;
        inductionVariable = inductionVariable + 1 | 0;
        var tmp4 = j;
        j = tmp4 + 1 | 0;
        var tmp = (input[tmp4] & 255) << 24;
        var tmp3 = j;
        j = tmp3 + 1 | 0;
        var tmp_0 = tmp | (input[tmp3] & 255) << 16;
        var tmp2 = j;
        j = tmp2 + 1 | 0;
        var tmp_1 = tmp_0 | (input[tmp2] & 255) << 8;
        var tmp1 = j;
        j = tmp1 + 1 | 0;
        x[i] = tmp_1 | input[tmp1] & 255;
      }
       while (inductionVariable < 16);
    var inductionVariable_0 = 16;
    if (inductionVariable_0 < 64)
      do {
        var i_0 = inductionVariable_0;
        inductionVariable_0 = inductionVariable_0 + 1 | 0;
        var x15 = x[i_0 - 15 | 0];
        var s0 = (x15 >>> 7 | 0 | x15 << 25) ^ (x15 >>> 18 | 0 | x15 << 14) ^ (x15 >>> 3 | 0);
        var x2 = x[i_0 - 2 | 0];
        var s1 = (x2 >>> 17 | 0 | x2 << 15) ^ (x2 >>> 19 | 0 | x2 << 13) ^ (x2 >>> 10 | 0);
        var x16 = x[i_0 - 16 | 0];
        var x7 = x[i_0 - 7 | 0];
        x[i_0] = ((x16 + s0 | 0) + x7 | 0) + s1 | 0;
      }
       while (inductionVariable_0 < 64);
    var k = Companion_getInstance().o1n_1;
    var a = this.n1n_1[0];
    var b = this.n1n_1[1];
    var c = this.n1n_1[2];
    var d = this.n1n_1[3];
    var e = this.n1n_1[4];
    var f = this.n1n_1[5];
    var g = this.n1n_1[6];
    var h = this.n1n_1[7];
    var inductionVariable_1 = 0;
    if (inductionVariable_1 < 64)
      do {
        var i_1 = inductionVariable_1;
        inductionVariable_1 = inductionVariable_1 + 1 | 0;
        var s0_0 = (a >>> 2 | 0 | a << 30) ^ (a >>> 13 | 0 | a << 19) ^ (a >>> 22 | 0 | a << 10);
        var s1_0 = (e >>> 6 | 0 | e << 26) ^ (e >>> 11 | 0 | e << 21) ^ (e >>> 25 | 0 | e << 7);
        var ch = e & f ^ ~e & g;
        var maj = a & b ^ a & c ^ b & c;
        var t1 = (((h + s1_0 | 0) + ch | 0) + k[i_1] | 0) + x[i_1] | 0;
        var t2 = s0_0 + maj | 0;
        h = g;
        g = f;
        f = e;
        e = d + t1 | 0;
        d = c;
        c = b;
        b = a;
        a = t1 + t2 | 0;
      }
       while (inductionVariable_1 < 64);
    var tmp7_array = this.n1n_1;
    tmp7_array[0] = tmp7_array[0] + a | 0;
    var tmp9_array = this.n1n_1;
    tmp9_array[1] = tmp9_array[1] + b | 0;
    var tmp11_array = this.n1n_1;
    tmp11_array[2] = tmp11_array[2] + c | 0;
    var tmp13_array = this.n1n_1;
    tmp13_array[3] = tmp13_array[3] + d | 0;
    var tmp15_array = this.n1n_1;
    tmp15_array[4] = tmp15_array[4] + e | 0;
    var tmp17_array = this.n1n_1;
    tmp17_array[5] = tmp17_array[5] + f | 0;
    var tmp19_array = this.n1n_1;
    tmp19_array[6] = tmp19_array[6] + g | 0;
    var tmp21_array = this.n1n_1;
    tmp21_array[7] = tmp21_array[7] + h | 0;
  };
  protoOf(Bit32Digest).x1m = function (bitLength, bufferOffset, buffer) {
    buffer[bufferOffset] = -128;
    var size = bufferOffset + 1 | 0;
    if (size > 56) {
      fill(buffer, 0, size, 64);
      this.w1m(buffer, 0);
      fill(buffer, 0, 0, size);
    } else {
      fill(buffer, 0, size, 56);
    }
    buffer[56] = bitLength.o3(56).s3();
    buffer[57] = bitLength.o3(48).s3();
    buffer[58] = bitLength.o3(40).s3();
    buffer[59] = bitLength.o3(32).s3();
    buffer[60] = bitLength.o3(24).s3();
    buffer[61] = bitLength.o3(16).s3();
    buffer[62] = bitLength.o3(8).s3();
    buffer[63] = bitLength.s3();
    this.w1m(buffer, 0);
    return this.p1n(this.n1n_1[0], this.n1n_1[1], this.n1n_1[2], this.n1n_1[3], this.n1n_1[4], this.n1n_1[5], this.n1n_1[6], this.n1n_1[7]);
  };
  protoOf(Bit32Digest).y1m = function () {
    fill_0(this.m1n_1, 0);
    this.n1n_1[0] = this.e1n_1;
    this.n1n_1[1] = this.f1n_1;
    this.n1n_1[2] = this.g1n_1;
    this.n1n_1[3] = this.h1n_1;
    this.n1n_1[4] = this.i1n_1;
    this.n1n_1[5] = this.j1n_1;
    this.n1n_1[6] = this.k1n_1;
    this.n1n_1[7] = this.l1n_1;
  };
  function Bit32Digest() {
    Companion_getInstance();
  }
  function Bit64Digest_init_$Init$(d, t, h0, h1, h2, h3, h4, h5, h6, h7, $this) {
    var tmp = 'SHA-' + d;
    var tmp_0;
    if (t == null) {
      tmp_0 = null;
    } else {
      // Inline function 'kotlin.let' call
      // Inline function 'kotlin.contracts.contract' call
      // Inline function 'org.kotlincrypto.hash.sha2.Bit64Digest.<init>.<anonymous>' call
      tmp_0 = '/' + t;
    }
    var tmp1_elvis_lhs = tmp_0;
    var tmp_1 = tmp + (tmp1_elvis_lhs == null ? '' : tmp1_elvis_lhs);
    Digest_init_$Init$(tmp_1, 128, (t == null ? d : t) / 8 | 0, $this);
    Bit64Digest.call($this);
    if (!(t == null)) {
      // Inline function 'kotlin.require' call
      // Inline function 'kotlin.contracts.contract' call
      if (!(d === 512)) {
        // Inline function 'org.kotlincrypto.hash.sha2.Bit64Digest.<init>.<anonymous>' call
        var message = 't can only be expressed for SHA-512';
        throw IllegalArgumentException_init_$Create$(toString(message));
      }
      // Inline function 'kotlin.require' call
      // Inline function 'kotlin.contracts.contract' call
      if (!((t % 8 | 0) === 0)) {
        // Inline function 'org.kotlincrypto.hash.sha2.Bit64Digest.<init>.<anonymous>' call
        var message_0 = 't must be a factor of 8';
        throw IllegalArgumentException_init_$Create$(toString(message_0));
      }
      // Inline function 'kotlin.require' call
      // Inline function 'kotlin.contracts.contract' call
      if (!(t < 512)) {
        // Inline function 'org.kotlincrypto.hash.sha2.Bit64Digest.<init>.<anonymous>' call
        var message_1 = 't must be less than 512';
        throw IllegalArgumentException_init_$Create$(toString(message_1));
      }
      // Inline function 'kotlin.require' call
      // Inline function 'kotlin.contracts.contract' call
      if (!!(t === 384)) {
        // Inline function 'org.kotlincrypto.hash.sha2.Bit64Digest.<init>.<anonymous>' call
        var message_2 = 't cannot be 384';
        throw IllegalArgumentException_init_$Create$(toString(message_2));
      }
    }
    $this.r1n_1 = h0;
    $this.s1n_1 = h1;
    $this.t1n_1 = h2;
    $this.u1n_1 = h3;
    $this.v1n_1 = h4;
    $this.w1n_1 = h5;
    $this.x1n_1 = h6;
    $this.y1n_1 = h7;
    $this.z1n_1 = longArray(80);
    var tmp_2 = $this;
    // Inline function 'kotlin.longArrayOf' call
    tmp_2.a1o_1 = longArrayOf([h0, h1, h2, h3, h4, h5, h6, h7]);
    return $this;
  }
  function Companion_0() {
    Companion_instance_0 = this;
    var tmp = this;
    // Inline function 'kotlin.longArrayOf' call
    tmp.b1o_1 = longArrayOf([new Long(-685199838, 1116352408), new Long(602891725, 1899447441), new Long(-330482897, -1245643825), new Long(-2121671748, -373957723), new Long(-213338824, 961987163), new Long(-1241133031, 1508970993), new Long(-1357295717, -1841331548), new Long(-630357736, -1424204075), new Long(-1560083902, -670586216), new Long(1164996542, 310598401), new Long(1323610764, 607225278), new Long(-704662302, 1426881987), new Long(-226784913, 1925078388), new Long(991336113, -2132889090), new Long(633803317, -1680079193), new Long(-815192428, -1046744716), new Long(-1628353838, -459576895), new Long(944711139, -272742522), new Long(-1953704523, 264347078), new Long(2007800933, 604807628), new Long(1495990901, 770255983), new Long(1856431235, 1249150122), new Long(-1119749164, 1555081692), new Long(-2096016459, 1996064986), new Long(-295247957, -1740746414), new Long(766784016, -1473132947), new Long(-1728372417, -1341970488), new Long(-1091629340, -1084653625), new Long(1034457026, -958395405), new Long(-1828018395, -710438585), new Long(-536640913, 113926993), new Long(168717936, 338241895), new Long(1188179964, 666307205), new Long(1546045734, 773529912), new Long(1522805485, 1294757372), new Long(-1651133473, 1396182291), new Long(-1951439906, 1695183700), new Long(1014477480, 1986661051), new Long(1206759142, -2117940946), new Long(344077627, -1838011259), new Long(1290863460, -1564481375), new Long(-1136513023, -1474664885), new Long(-789014639, -1035236496), new Long(106217008, -949202525), new Long(-688958952, -778901479), new Long(1432725776, -694614492), new Long(1467031594, -200395387), new Long(851169720, 275423344), new Long(-1194143544, 430227734), new Long(1363258195, 506948616), new Long(-544281703, 659060556), new Long(-509917016, 883997877), new Long(-976659869, 958139571), new Long(-482243893, 1322822218), new Long(2003034995, 1537002063), new Long(-692930397, 1747873779), new Long(1575990012, 1955562222), new Long(1125592928, 2024104815), new Long(-1578062990, -2067236844), new Long(442776044, -1933114872), new Long(593698344, -1866530822), new Long(-561857047, -1538233109), new Long(-1295615723, -1090935817), new Long(-479046869, -965641998), new Long(-366583396, -903397682), new Long(566280711, -779700025), new Long(-840897762, -354779690), new Long(-294727304, -176337025), new Long(1914138554, 116418474), new Long(-1563912026, 174292421), new Long(-1090974290, 289380356), new Long(320620315, 460393269), new Long(587496836, 685471733), new Long(1086792851, 852142971), new Long(365543100, 1017036298), new Long(-1676669620, 1126000580), new Long(-885112138, 1288033470), new Long(-60457430, 1501505948), new Long(987167468, 1607167915), new Long(1246189591, 1816402316)]);
  }
  var Companion_instance_0;
  function Companion_getInstance_0() {
    if (Companion_instance_0 == null)
      new Companion_0();
    return Companion_instance_0;
  }
  protoOf(Bit64Digest).w1m = function (input, offset) {
    var x = this.z1n_1;
    var j = offset;
    var inductionVariable = 0;
    if (inductionVariable < 16)
      do {
        var i = inductionVariable;
        inductionVariable = inductionVariable + 1 | 0;
        var tmp8 = j;
        j = tmp8 + 1 | 0;
        var tmp = toLong(input[tmp8]).p3(new Long(255, 0)).m3(56);
        var tmp7 = j;
        j = tmp7 + 1 | 0;
        var tmp_0 = tmp.q3(toLong(input[tmp7]).p3(new Long(255, 0)).m3(48));
        var tmp6 = j;
        j = tmp6 + 1 | 0;
        var tmp_1 = tmp_0.q3(toLong(input[tmp6]).p3(new Long(255, 0)).m3(40));
        var tmp5 = j;
        j = tmp5 + 1 | 0;
        var tmp_2 = tmp_1.q3(toLong(input[tmp5]).p3(new Long(255, 0)).m3(32));
        var tmp4 = j;
        j = tmp4 + 1 | 0;
        var tmp_3 = tmp_2.q3(toLong(input[tmp4]).p3(new Long(255, 0)).m3(24));
        var tmp3 = j;
        j = tmp3 + 1 | 0;
        var tmp_4 = tmp_3.q3(toLong(input[tmp3]).p3(new Long(255, 0)).m3(16));
        var tmp2 = j;
        j = tmp2 + 1 | 0;
        var tmp_5 = tmp_4.q3(toLong(input[tmp2]).p3(new Long(255, 0)).m3(8));
        var tmp1 = j;
        j = tmp1 + 1 | 0;
        x[i] = tmp_5.q3(toLong(input[tmp1]).p3(new Long(255, 0)));
      }
       while (inductionVariable < 16);
    var inductionVariable_0 = 16;
    if (inductionVariable_0 < 80)
      do {
        var i_0 = inductionVariable_0;
        inductionVariable_0 = inductionVariable_0 + 1 | 0;
        var x15 = x[i_0 - 15 | 0];
        // Inline function 'org.kotlincrypto.hash.sha2.Bit64Digest.rotateRight' call
        var tmp_6 = x15.o3(1).q3(x15.m3(64 - 1 | 0));
        // Inline function 'org.kotlincrypto.hash.sha2.Bit64Digest.rotateRight' call
        var tmp$ret$1 = x15.o3(8).q3(x15.m3(64 - 8 | 0));
        var s0 = tmp_6.r3(tmp$ret$1).r3(x15.o3(7));
        var x2 = x[i_0 - 2 | 0];
        // Inline function 'org.kotlincrypto.hash.sha2.Bit64Digest.rotateRight' call
        var tmp_7 = x2.o3(19).q3(x2.m3(64 - 19 | 0));
        // Inline function 'org.kotlincrypto.hash.sha2.Bit64Digest.rotateRight' call
        var tmp$ret$3 = x2.o3(61).q3(x2.m3(64 - 61 | 0));
        var s1 = tmp_7.r3(tmp$ret$3).r3(x2.o3(6));
        var x16 = x[i_0 - 16 | 0];
        var x7 = x[i_0 - 7 | 0];
        x[i_0] = x16.d3(s0).d3(x7).d3(s1);
      }
       while (inductionVariable_0 < 80);
    var k = Companion_getInstance_0().b1o_1;
    var a = this.a1o_1[0];
    var b = this.a1o_1[1];
    var c = this.a1o_1[2];
    var d = this.a1o_1[3];
    var e = this.a1o_1[4];
    var f = this.a1o_1[5];
    var g = this.a1o_1[6];
    var h = this.a1o_1[7];
    var inductionVariable_1 = 0;
    if (inductionVariable_1 < 80)
      do {
        var i_1 = inductionVariable_1;
        inductionVariable_1 = inductionVariable_1 + 1 | 0;
        // Inline function 'org.kotlincrypto.hash.sha2.Bit64Digest.rotateRight' call
        var this_0 = a;
        var tmp_8 = this_0.o3(28).q3(this_0.m3(64 - 28 | 0));
        // Inline function 'org.kotlincrypto.hash.sha2.Bit64Digest.rotateRight' call
        var this_1 = a;
        var tmp$ret$5 = this_1.o3(34).q3(this_1.m3(64 - 34 | 0));
        var tmp_9 = tmp_8.r3(tmp$ret$5);
        // Inline function 'org.kotlincrypto.hash.sha2.Bit64Digest.rotateRight' call
        var this_2 = a;
        var tmp$ret$6 = this_2.o3(39).q3(this_2.m3(64 - 39 | 0));
        var s0_0 = tmp_9.r3(tmp$ret$6);
        // Inline function 'org.kotlincrypto.hash.sha2.Bit64Digest.rotateRight' call
        var this_3 = e;
        var tmp_10 = this_3.o3(14).q3(this_3.m3(64 - 14 | 0));
        // Inline function 'org.kotlincrypto.hash.sha2.Bit64Digest.rotateRight' call
        var this_4 = e;
        var tmp$ret$8 = this_4.o3(18).q3(this_4.m3(64 - 18 | 0));
        var tmp_11 = tmp_10.r3(tmp$ret$8);
        // Inline function 'org.kotlincrypto.hash.sha2.Bit64Digest.rotateRight' call
        var this_5 = e;
        var tmp$ret$9 = this_5.o3(41).q3(this_5.m3(64 - 41 | 0));
        var s1_0 = tmp_11.r3(tmp$ret$9);
        var ch = e.p3(f).r3(e.l3().p3(g));
        var maj = a.p3(b).r3(a.p3(c)).r3(b.p3(c));
        var t1 = h.d3(s1_0).d3(ch).d3(k[i_1]).d3(x[i_1]);
        var t2 = s0_0.d3(maj);
        h = g;
        g = f;
        f = e;
        e = d.d3(t1);
        d = c;
        c = b;
        b = a;
        a = t1.d3(t2);
      }
       while (inductionVariable_1 < 80);
    var tmp11_array = this.a1o_1;
    tmp11_array[0] = tmp11_array[0].d3(a);
    var tmp13_array = this.a1o_1;
    tmp13_array[1] = tmp13_array[1].d3(b);
    var tmp15_array = this.a1o_1;
    tmp15_array[2] = tmp15_array[2].d3(c);
    var tmp17_array = this.a1o_1;
    tmp17_array[3] = tmp17_array[3].d3(d);
    var tmp19_array = this.a1o_1;
    tmp19_array[4] = tmp19_array[4].d3(e);
    var tmp21_array = this.a1o_1;
    tmp21_array[5] = tmp21_array[5].d3(f);
    var tmp23_array = this.a1o_1;
    tmp23_array[6] = tmp23_array[6].d3(g);
    var tmp25_array = this.a1o_1;
    tmp25_array[7] = tmp25_array[7].d3(h);
  };
  protoOf(Bit64Digest).x1m = function (bitLength, bufferOffset, buffer) {
    buffer[bufferOffset] = -128;
    var size = bufferOffset + 1 | 0;
    if (size > 112) {
      fill(buffer, 0, size, 128);
      this.w1m(buffer, 0);
      fill(buffer, 0, 0, size);
    } else {
      fill(buffer, 0, size, 120);
    }
    buffer[120] = bitLength.o3(56).s3();
    buffer[121] = bitLength.o3(48).s3();
    buffer[122] = bitLength.o3(40).s3();
    buffer[123] = bitLength.o3(32).s3();
    buffer[124] = bitLength.o3(24).s3();
    buffer[125] = bitLength.o3(16).s3();
    buffer[126] = bitLength.o3(8).s3();
    buffer[127] = bitLength.s3();
    this.w1m(buffer, 0);
    return this.c1o(this.a1o_1[0], this.a1o_1[1], this.a1o_1[2], this.a1o_1[3], this.a1o_1[4], this.a1o_1[5], this.a1o_1[6], this.a1o_1[7]);
  };
  protoOf(Bit64Digest).y1m = function () {
    fill_1(this.z1n_1, new Long(0, 0));
    this.a1o_1[0] = this.r1n_1;
    this.a1o_1[1] = this.s1n_1;
    this.a1o_1[2] = this.t1n_1;
    this.a1o_1[3] = this.u1n_1;
    this.a1o_1[4] = this.v1n_1;
    this.a1o_1[5] = this.w1n_1;
    this.a1o_1[6] = this.x1n_1;
    this.a1o_1[7] = this.y1n_1;
  };
  function Bit64Digest() {
    Companion_getInstance_0();
  }
  function SHA256_init_$Init$($this) {
    Bit32Digest_init_$Init$(256, 1779033703, -1150833019, 1013904242, -1521486534, 1359893119, -1694144372, 528734635, 1541459225, $this);
    SHA256.call($this);
    return $this;
  }
  function SHA256_init_$Create$() {
    return SHA256_init_$Init$(objectCreate(protoOf(SHA256)));
  }
  protoOf(SHA256).p1n = function (a, b, c, d, e, f, g, h) {
    // Inline function 'kotlin.byteArrayOf' call
    return new Int8Array([toByte(a >> 24), toByte(a >> 16), toByte(a >> 8), toByte(a), toByte(b >> 24), toByte(b >> 16), toByte(b >> 8), toByte(b), toByte(c >> 24), toByte(c >> 16), toByte(c >> 8), toByte(c), toByte(d >> 24), toByte(d >> 16), toByte(d >> 8), toByte(d), toByte(e >> 24), toByte(e >> 16), toByte(e >> 8), toByte(e), toByte(f >> 24), toByte(f >> 16), toByte(f >> 8), toByte(f), toByte(g >> 24), toByte(g >> 16), toByte(g >> 8), toByte(g), toByte(h >> 24), toByte(h >> 16), toByte(h >> 8), toByte(h)]);
  };
  function SHA256() {
  }
  function SHA384_init_$Init$($this) {
    Bit64Digest_init_$Init$(384, null, new Long(-1056596264, -876896931), new Long(914150663, 1654270250), new Long(812702999, -1856437926), new Long(-150054599, 355462360), new Long(-4191439, 1731405415), new Long(1750603025, -1900787065), new Long(1694076839, -619958771), new Long(-1090891868, 1203062813), $this);
    SHA384.call($this);
    return $this;
  }
  function SHA384_init_$Create$() {
    return SHA384_init_$Init$(objectCreate(protoOf(SHA384)));
  }
  protoOf(SHA384).c1o = function (a, b, c, d, e, f, g, h) {
    // Inline function 'kotlin.byteArrayOf' call
    return new Int8Array([a.n3(56).s3(), a.n3(48).s3(), a.n3(40).s3(), a.n3(32).s3(), a.n3(24).s3(), a.n3(16).s3(), a.n3(8).s3(), a.s3(), b.n3(56).s3(), b.n3(48).s3(), b.n3(40).s3(), b.n3(32).s3(), b.n3(24).s3(), b.n3(16).s3(), b.n3(8).s3(), b.s3(), c.n3(56).s3(), c.n3(48).s3(), c.n3(40).s3(), c.n3(32).s3(), c.n3(24).s3(), c.n3(16).s3(), c.n3(8).s3(), c.s3(), d.n3(56).s3(), d.n3(48).s3(), d.n3(40).s3(), d.n3(32).s3(), d.n3(24).s3(), d.n3(16).s3(), d.n3(8).s3(), d.s3(), e.n3(56).s3(), e.n3(48).s3(), e.n3(40).s3(), e.n3(32).s3(), e.n3(24).s3(), e.n3(16).s3(), e.n3(8).s3(), e.s3(), f.n3(56).s3(), f.n3(48).s3(), f.n3(40).s3(), f.n3(32).s3(), f.n3(24).s3(), f.n3(16).s3(), f.n3(8).s3(), f.s3()]);
  };
  function SHA384() {
  }
  function SHA512_init_$Init$($this) {
    Bit64Digest_init_$Init$(512, null, new Long(-205731576, 1779033703), new Long(-2067093701, -1150833019), new Long(-23791573, 1013904242), new Long(1595750129, -1521486534), new Long(-1377402159, 1359893119), new Long(725511199, -1694144372), new Long(-79577749, 528734635), new Long(327033209, 1541459225), $this);
    SHA512.call($this);
    return $this;
  }
  function SHA512_init_$Create$() {
    return SHA512_init_$Init$(objectCreate(protoOf(SHA512)));
  }
  protoOf(SHA512).c1o = function (a, b, c, d, e, f, g, h) {
    // Inline function 'kotlin.byteArrayOf' call
    return new Int8Array([a.n3(56).s3(), a.n3(48).s3(), a.n3(40).s3(), a.n3(32).s3(), a.n3(24).s3(), a.n3(16).s3(), a.n3(8).s3(), a.s3(), b.n3(56).s3(), b.n3(48).s3(), b.n3(40).s3(), b.n3(32).s3(), b.n3(24).s3(), b.n3(16).s3(), b.n3(8).s3(), b.s3(), c.n3(56).s3(), c.n3(48).s3(), c.n3(40).s3(), c.n3(32).s3(), c.n3(24).s3(), c.n3(16).s3(), c.n3(8).s3(), c.s3(), d.n3(56).s3(), d.n3(48).s3(), d.n3(40).s3(), d.n3(32).s3(), d.n3(24).s3(), d.n3(16).s3(), d.n3(8).s3(), d.s3(), e.n3(56).s3(), e.n3(48).s3(), e.n3(40).s3(), e.n3(32).s3(), e.n3(24).s3(), e.n3(16).s3(), e.n3(8).s3(), e.s3(), f.n3(56).s3(), f.n3(48).s3(), f.n3(40).s3(), f.n3(32).s3(), f.n3(24).s3(), f.n3(16).s3(), f.n3(8).s3(), f.s3(), g.n3(56).s3(), g.n3(48).s3(), g.n3(40).s3(), g.n3(32).s3(), g.n3(24).s3(), g.n3(16).s3(), g.n3(8).s3(), g.s3(), h.n3(56).s3(), h.n3(48).s3(), h.n3(40).s3(), h.n3(32).s3(), h.n3(24).s3(), h.n3(16).s3(), h.n3(8).s3(), h.s3()]);
  };
  function SHA512() {
  }
  //region block: exports
  _.$_$ = _.$_$ || {};
  _.$_$.a = SHA256_init_$Create$;
  _.$_$.b = SHA384_init_$Create$;
  _.$_$.c = SHA512_init_$Create$;
  //endregion
  return _;
}));
