// @sphereon/kmp-mdoc-core/kotlinx-io-kotlinx-io-bytestring.js
(function (factory) {
  if (typeof define === 'function' && define.amd)
    define(['exports', './kotlin-kotlin-stdlib.js'], factory);
  else if (typeof exports === 'object')
    factory(module.exports, require('./kotlin-kotlin-stdlib.js'));
  else {
    if (typeof globalThis['kotlin-kotlin-stdlib'] === 'undefined') {
      throw new Error("Error loading module 'kotlinx-io-kotlinx-io-bytestring'. Its dependency 'kotlin-kotlin-stdlib' was not found. Please, check whether 'kotlin-kotlin-stdlib' is loaded prior to 'kotlinx-io-kotlinx-io-bytestring'.");
    }
    globalThis['kotlinx-io-kotlinx-io-bytestring'] = factory(typeof globalThis['kotlinx-io-kotlinx-io-bytestring'] === 'undefined' ? {} : globalThis['kotlinx-io-kotlinx-io-bytestring'], globalThis['kotlin-kotlin-stdlib']);
  }
}(function (_, kotlin_kotlin) {
  'use strict';
  //region block: imports
  var imul = Math.imul;
  var VOID = kotlin_kotlin.$_$.f;
  var copyOfRange = kotlin_kotlin.$_$.u5;
  var protoOf = kotlin_kotlin.$_$.xa;
  var objectCreate = kotlin_kotlin.$_$.wa;
  var toCharArray = kotlin_kotlin.$_$.vc;
  var initMetadataForCompanion = kotlin_kotlin.$_$.x9;
  var getKClassFromExpression = kotlin_kotlin.$_$.d;
  var THROW_CCE = kotlin_kotlin.$_$.ge;
  var Unit_instance = kotlin_kotlin.$_$.q4;
  var contentEquals = kotlin_kotlin.$_$.p5;
  var contentHashCode = kotlin_kotlin.$_$.r5;
  var toString = kotlin_kotlin.$_$.bb;
  var IllegalArgumentException_init_$Create$ = kotlin_kotlin.$_$.n1;
  var _UByte___init__impl__g9hnc4 = kotlin_kotlin.$_$.k2;
  var _UByte___get_data__impl__jof9qr = kotlin_kotlin.$_$.l2;
  var compareTo = kotlin_kotlin.$_$.m9;
  var StringBuilder_init_$Create$ = kotlin_kotlin.$_$.b1;
  var _Char___init__impl__6a9atx = kotlin_kotlin.$_$.c2;
  var Comparable = kotlin_kotlin.$_$.qd;
  var initMetadataForClass = kotlin_kotlin.$_$.w9;
  var numberToInt = kotlin_kotlin.$_$.ua;
  var arrayCopy = kotlin_kotlin.$_$.f5;
  var IndexOutOfBoundsException_init_$Create$ = kotlin_kotlin.$_$.s1;
  //endregion
  //region block: pre-declaration
  initMetadataForCompanion(Companion);
  initMetadataForClass(ByteString, 'ByteString', VOID, VOID, [Comparable]);
  initMetadataForClass(ByteStringBuilder, 'ByteStringBuilder', ByteStringBuilder);
  //endregion
  function ByteString_init_$Init$(data, startIndex, endIndex, $this) {
    startIndex = startIndex === VOID ? 0 : startIndex;
    endIndex = endIndex === VOID ? data.length : endIndex;
    ByteString.call($this, copyOfRange(data, startIndex, endIndex), null);
    return $this;
  }
  function ByteString_init_$Create$(data, startIndex, endIndex) {
    return ByteString_init_$Init$(data, startIndex, endIndex, objectCreate(protoOf(ByteString)));
  }
  function Companion() {
    Companion_instance = this;
    this.n19_1 = new ByteString(new Int8Array(0), null);
    this.o19_1 = toCharArray('0123456789abcdef');
  }
  protoOf(Companion).p19 = function (byteArray) {
    return new ByteString(byteArray, null);
  };
  var Companion_instance;
  function Companion_getInstance() {
    if (Companion_instance == null)
      new Companion();
    return Companion_instance;
  }
  function ByteString(data, dummy) {
    Companion_getInstance();
    this.q19_1 = data;
    this.r19_1 = 0;
  }
  protoOf(ByteString).t = function () {
    return this.q19_1.length;
  };
  protoOf(ByteString).equals = function (other) {
    if (this === other)
      return true;
    if (other == null || !getKClassFromExpression(this).equals(getKClassFromExpression(other)))
      return false;
    if (!(other instanceof ByteString))
      THROW_CCE();
    if (!(other.q19_1.length === this.q19_1.length))
      return false;
    if (!(other.r19_1 === 0) && !(this.r19_1 === 0) && !(other.r19_1 === this.r19_1))
      return false;
    return contentEquals(this.q19_1, other.q19_1);
  };
  protoOf(ByteString).hashCode = function () {
    var hc = this.r19_1;
    if (hc === 0) {
      hc = contentHashCode(this.q19_1);
      this.r19_1 = hc;
    }
    return hc;
  };
  protoOf(ByteString).s19 = function (startIndex, endIndex) {
    // Inline function 'kotlin.require' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(startIndex <= endIndex)) {
      // Inline function 'kotlinx.io.bytestring.ByteString.toByteArray.<anonymous>' call
      var message = 'startIndex (' + startIndex + ') > endIndex (' + endIndex + ')';
      throw IllegalArgumentException_init_$Create$(toString(message));
    }
    return copyOfRange(this.q19_1, startIndex, endIndex);
  };
  protoOf(ByteString).t19 = function (startIndex, endIndex, $super) {
    startIndex = startIndex === VOID ? 0 : startIndex;
    endIndex = endIndex === VOID ? this.t() : endIndex;
    return $super === VOID ? this.s19(startIndex, endIndex) : $super.s19.call(this, startIndex, endIndex);
  };
  protoOf(ByteString).u19 = function (other) {
    if (other === this)
      return 0;
    var localData = this.q19_1;
    var otherData = other.q19_1;
    var inductionVariable = 0;
    // Inline function 'kotlin.math.min' call
    var a = this.t();
    var b = other.t();
    var last = Math.min(a, b);
    if (inductionVariable < last)
      do {
        var i = inductionVariable;
        inductionVariable = inductionVariable + 1 | 0;
        // Inline function 'kotlin.UByte.compareTo' call
        // Inline function 'kotlin.toUByte' call
        var this_0 = localData[i];
        var this_1 = _UByte___init__impl__g9hnc4(this_0);
        // Inline function 'kotlin.toUByte' call
        var this_2 = otherData[i];
        var other_0 = _UByte___init__impl__g9hnc4(this_2);
        // Inline function 'kotlin.UByte.toInt' call
        var tmp = _UByte___get_data__impl__jof9qr(this_1) & 255;
        // Inline function 'kotlin.UByte.toInt' call
        var tmp$ret$4 = _UByte___get_data__impl__jof9qr(other_0) & 255;
        var cmp = compareTo(tmp, tmp$ret$4);
        if (!(cmp === 0))
          return cmp;
      }
       while (inductionVariable < last);
    return compareTo(this.t(), other.t());
  };
  protoOf(ByteString).d = function (other) {
    return this.u19(other instanceof ByteString ? other : THROW_CCE());
  };
  protoOf(ByteString).toString = function () {
    if (isEmpty(this)) {
      return 'ByteString(size=0)';
    }
    var sizeStr = this.t().toString();
    var len = (22 + sizeStr.length | 0) + imul(this.t(), 2) | 0;
    // Inline function 'kotlin.with' call
    // Inline function 'kotlin.contracts.contract' call
    // Inline function 'kotlinx.io.bytestring.ByteString.toString.<anonymous>' call
    var $this$with = StringBuilder_init_$Create$(len);
    $this$with.s8('ByteString(size=');
    $this$with.s8(sizeStr);
    $this$with.s8(' hex=');
    var localData = this.q19_1;
    var inductionVariable = 0;
    var last = this.t();
    if (inductionVariable < last)
      do {
        var i = inductionVariable;
        inductionVariable = inductionVariable + 1 | 0;
        var b = localData[i];
        $this$with.t8(Companion_getInstance().o19_1[(b >>> 4 | 0) & 15]);
        $this$with.t8(Companion_getInstance().o19_1[b & 15]);
      }
       while (inductionVariable < last);
    return $this$with.t8(_Char___init__impl__6a9atx(41)).toString();
  };
  function isEmpty(_this__u8e3s4) {
    return _this__u8e3s4.t() === 0;
  }
  function EmptyByteString() {
    return Companion_getInstance().n19_1;
  }
  function ensureCapacity($this, requiredCapacity) {
    if ($this.v19_1.length >= requiredCapacity) {
      return Unit_instance;
    }
    var tmp;
    // Inline function 'kotlin.collections.isEmpty' call
    if ($this.v19_1.length === 0) {
      tmp = 16;
    } else {
      tmp = numberToInt($this.v19_1.length * 1.5);
    }
    var desiredSize = tmp;
    // Inline function 'kotlin.math.max' call
    var a = desiredSize;
    desiredSize = Math.max(a, requiredCapacity);
    var newBuffer = new Int8Array(desiredSize);
    // Inline function 'kotlin.collections.copyInto' call
    var this_0 = $this.v19_1;
    var endIndex = this_0.length;
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    var tmp_0 = this_0;
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    arrayCopy(tmp_0, newBuffer, 0, 0, endIndex);
    $this.v19_1 = newBuffer;
  }
  function ByteStringBuilder(initialCapacity) {
    initialCapacity = initialCapacity === VOID ? 0 : initialCapacity;
    this.v19_1 = new Int8Array(initialCapacity);
    this.w19_1 = 0;
  }
  protoOf(ByteStringBuilder).t = function () {
    return this.w19_1;
  };
  protoOf(ByteStringBuilder).x19 = function () {
    if (this.t() === 0) {
      return EmptyByteString();
    }
    if (this.v19_1.length === this.t()) {
      return Companion_getInstance().p19(this.v19_1);
    }
    return ByteString_init_$Create$(this.v19_1, 0, this.t());
  };
  protoOf(ByteStringBuilder).y19 = function (byte) {
    ensureCapacity(this, this.t() + 1 | 0);
    var tmp = this.v19_1;
    var _unary__edvuaz = this.w19_1;
    this.w19_1 = _unary__edvuaz + 1 | 0;
    tmp[_unary__edvuaz] = byte;
  };
  protoOf(ByteStringBuilder).z19 = function (array, startIndex, endIndex) {
    // Inline function 'kotlin.require' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(startIndex <= endIndex)) {
      // Inline function 'kotlinx.io.bytestring.ByteStringBuilder.append.<anonymous>' call
      var message = 'startIndex (' + startIndex + ') > endIndex (' + endIndex + ')';
      throw IllegalArgumentException_init_$Create$(toString(message));
    }
    if (startIndex < 0 || endIndex > array.length) {
      throw IndexOutOfBoundsException_init_$Create$('startIndex (' + startIndex + ') and endIndex (' + endIndex + ') represents ' + ("an interval out of array's bounds [0.." + array.length + ').'));
    }
    ensureCapacity(this, (this.w19_1 + endIndex | 0) - startIndex | 0);
    // Inline function 'kotlin.collections.copyInto' call
    var destination = this.v19_1;
    var destinationOffset = this.w19_1;
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    var tmp = array;
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    arrayCopy(tmp, destination, destinationOffset, startIndex, endIndex);
    this.w19_1 = this.w19_1 + (endIndex - startIndex | 0) | 0;
  };
  protoOf(ByteStringBuilder).a1a = function (array, startIndex, endIndex, $super) {
    startIndex = startIndex === VOID ? 0 : startIndex;
    endIndex = endIndex === VOID ? array.length : endIndex;
    var tmp;
    if ($super === VOID) {
      this.z19(array, startIndex, endIndex);
      tmp = Unit_instance;
    } else {
      tmp = $super.z19.call(this, array, startIndex, endIndex);
    }
    return tmp;
  };
  //region block: exports
  _.$_$ = _.$_$ || {};
  _.$_$.a = ByteStringBuilder;
  //endregion
  return _;
}));
